#from distutils.core import setup
from setuptools import find_packages, setup

setup(
    name                = 'NeatSeq-Flow-modules',
    version             = '1.1.0',
    author              = 'Menachem Sklarz',
    author_email        = 'sklarz@bgu.ac.il',
    maintainer          = 'Menachem Sklarz',
    maintainer_email    = 'sklarz@bgu.ac.il',
    url                 = 'http://neatseq-flow.readthedocs.io/projects/neatseq-flow-modules/en/latest/',
    description         = 'Shared modules and workflows for use with NeatSeq-Flow',
    license             = 'Creative Commons Attribution-Noncommercial-Share Alike license',
    long_description    =  open('README').read(),
    download_url        = 'https://github.com/bioinfo-core-BGU/neatseq-flow-modules',
    platforms           = ["POSIX","Windows"],
    packages            = [ 'neatseq_flow_modules',
                            'neatseq_flow_modules.Annotation',
                            'neatseq_flow_modules.Assembly',
                            'neatseq_flow_modules.ChIP_seq',
                            'neatseq_flow_modules.clustering',
                            'neatseq_flow_modules.Liron.cgMLST_and_MLST_typing_module',
                            'neatseq_flow_modules.Liron.Collect_results_module',
                            'neatseq_flow_modules.Liron.Cutadapt_module',
                            'neatseq_flow_modules.Liron.Gassst_module',
                            'neatseq_flow_modules.Liron.Prokka_module',
                            'neatseq_flow_modules.Liron.Roary_module',
                            'neatseq_flow_modules.Liron.RSEM_module',
                            'neatseq_flow_modules.Liron.Snippy_module',
                            'neatseq_flow_modules.Liron.Trim_Galore_module',
                            'neatseq_flow_modules.Liron',
                            'neatseq_flow_modules.main_NSF_classes.Assembly',
                            'neatseq_flow_modules.main_NSF_classes.ChIP_seq',
                            'neatseq_flow_modules.main_NSF_classes.Generic',
                            'neatseq_flow_modules.main_NSF_classes.IGVtools',
                            'neatseq_flow_modules.main_NSF_classes.mapping',
                            'neatseq_flow_modules.main_NSF_classes.preparing',
                            'neatseq_flow_modules.main_NSF_classes.Reports',
                            'neatseq_flow_modules.main_NSF_classes.RNA_seq',
                            'neatseq_flow_modules.main_NSF_classes.searching',
                            'neatseq_flow_modules.main_NSF_classes.UCSCtools',
                            'neatseq_flow_modules.main_NSF_classes.variants',
                            'neatseq_flow_modules.main_NSF_classes',
                            'neatseq_flow_modules.mapping',
                            'neatseq_flow_modules.metagenomics',
                            'neatseq_flow_modules.microbiology',
                            'neatseq_flow_modules.miscellaneous',
                            'neatseq_flow_modules.QIIME',
                            'neatseq_flow_modules.Reports',
                            'neatseq_flow_modules.searching',
                            'neatseq_flow_modules.Transcriptomics',
                            'neatseq_flow_modules.UCSCtools',
                            'neatseq_flow_modules.variants'],
    data_files          = [('Workflows',['Workflows/Assembly_Indexing_mapping.yaml',
                                        'Workflows/BLAST_db.yaml',
                                        'Workflows/BLAST_fasta.yaml',
                                        'Workflows/ChIP_seq_bowtie2.yaml',
                                        'Workflows/Clustering.yaml',
                                        'Workflows/Metagenomics.yaml',
                                        'Workflows/parameter_set_QIIME/qiime1_mapping.txt',
                                        'Workflows/parameter_set_QIIME/qiime1_params.txt',
                                        'Workflows/parameter_set_QIIME/QIIME_workflow.yaml',
                                        'Workflows/parameter_set_QIIME/sample_file_paired_small.nsfs',
                                        'Workflows/RNA_seq_reference.yaml',
                                        'Workflows/RNA_seq_Trinity.yaml',
                                        'Workflows/variant_calling.yaml']),
                            ('Workflows/Sample_sets',['Workflows/sample_sets/ChIP_tabular.nsfs',
                                                        'Workflows/sample_sets/Fasta.nsfs',
                                                        'Workflows/sample_sets/PE.nsfs',
                                                        'Workflows/sample_sets/PE_ChIP.nsfs',
                                                        'Workflows/sample_sets/PE_tabular.nsfs',
                                                        'Workflows/sample_sets/SE.nsfs',])],
    # install_requires    = [],
    classifiers         = [
                          'Development Status :: 4 - Beta',
                          'Environment :: Console',
                          'Intended Audience :: End Users',
                          'Intended Audience :: Developers',
                          'License :: OSI Approved :: Python Software Foundation License',
                          'Operating System :: Microsoft :: Windows',
                          'Operating System :: POSIX',
                          'Programming Language :: Python',
                          ],
    )
    

    
