

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>QIIME (version 1.9) &mdash; NeatSeq_Flow Module Repository 1.1.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="NeatSeq_Flow Module Repository 1.1.0 documentation" href="../index.html"/>
        <link rel="next" title="Sequence Annotation" href="SequenceAnnotation.html"/>
        <link rel="prev" title="Preparation and QC" href="PreparationAndQC.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> NeatSeq_Flow Module Repository
          

          
            
            <img src="../_static/logo_Hor.png" class="logo" />
          
          </a>

          
            
            
              <div class="version">
                1.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <p class="caption"><span class="caption-text">Modules</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="BAMConversion.html">BAM Conversion to Other Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="ChIPseq.html">ChIP-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="GenericModule.html">Generic Module</a></li>
<li class="toctree-l1"><a class="reference internal" href="GenomeAssembly.html">Genome Assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="Mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="Metagenomics.html">Metagenomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="Microbiology.html">Microbiology</a></li>
<li class="toctree-l1"><a class="reference internal" href="PreparationAndQC.html">Preparation and QC</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">QIIME (version 1.9)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#qiime-prep"><code class="docutils literal"><span class="pre">qiime_prep</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#requires">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parameters-that-can-be-set">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#lines-for-parameter-file">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-demult"><code class="docutils literal"><span class="pre">qiime_demult</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-chimera"><code class="docutils literal"><span class="pre">qiime_chimera</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id5">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id6">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id7">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id8">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id9">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-pick-otus"><code class="docutils literal"><span class="pre">qiime_pick_otus</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id10">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id11">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id13">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id14">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-pick-rep-set"><code class="docutils literal"><span class="pre">qiime_pick_rep_set</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id15">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id16">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id17">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id18">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id19">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-align-seqs"><code class="docutils literal"><span class="pre">qiime_align_seqs</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id20">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id21">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id22">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id24">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-filter-alignment"><code class="docutils literal"><span class="pre">qiime_filter_alignment</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id25">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id26">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id27">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id28">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id29">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-assign-taxonomy"><code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id30">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id31">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id32">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id33">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id34">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-make-phylogeny"><code class="docutils literal"><span class="pre">qiime_make_phylogeny</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id35">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id36">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id37">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id38">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id39">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-make-otu-table"><code class="docutils literal"><span class="pre">qiime_make_otu_table</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id40">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#optional">Optional</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id41">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id42">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id43">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id44">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-filter-samples-from-otu-table"><code class="docutils literal"><span class="pre">qiime_filter_samples_from_otu_table</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id45">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id46">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id47">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id48">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id49">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-filter-otus"><code class="docutils literal"><span class="pre">qiime_filter_otus</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id50">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id51">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id52">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id53">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id54">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-sort-otu-table"><code class="docutils literal"><span class="pre">qiime_sort_otu_table</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id55">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id56">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id57">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id58">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id59">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-divers"><code class="docutils literal"><span class="pre">qiime_divers</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id60">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id61">Optional</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id62">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id63">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id64">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id65">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="SequenceAnnotation.html">Sequence Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="SequenceClustering.html">Sequence Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="SequenceSearching.html">Sequence-Searching Related Tasks</a></li>
<li class="toctree-l1"><a class="reference internal" href="TranscriptomeAssembly.html">Transcriptome Assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="Variants.html">Variant-related Tasks</a></li>
<li class="toctree-l1"><a class="reference internal" href="VariousReportingPrograms.html">Various Reporting Programs</a></li>
</ul>
<p class="caption"><span class="caption-text">Workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/Assembly_Indexing_mapping.html">Assembly and Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/BLAST_db.html">BLAST on <cite>fasta</cite> files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/BLAST_fasta.html">Assembly and Searching</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/ChIP_seq_bowtie2.html">ChIP-seq analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/Metagenomics.html">Metagenomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/RNA_seq_Trinity.html">Non-Model Transcriptomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/RNA_seq_reference.html">Reference-based Transcriptomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/variant_calling.html">Variant calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/Clustering.html">Read clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/QIIME_workflow.html">QIIME analysis</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">NeatSeq_Flow Module Repository</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
    <li>QIIME (version 1.9)</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/Module_docs/QIIME.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="qiime-version-1-9">
<h1>QIIME (version 1.9)<a class="headerlink" href="#qiime-version-1-9" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="modules-included-in-this-section">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#qiime-prep" id="id66"><code class="docutils literal"><span class="pre">qiime_prep</span></code></a></li>
<li><a class="reference internal" href="#qiime-demult" id="id67"><code class="docutils literal"><span class="pre">qiime_demult</span></code></a></li>
<li><a class="reference internal" href="#qiime-chimera" id="id68"><code class="docutils literal"><span class="pre">qiime_chimera</span></code></a></li>
<li><a class="reference internal" href="#qiime-pick-otus" id="id69"><code class="docutils literal"><span class="pre">qiime_pick_otus</span></code></a></li>
<li><a class="reference internal" href="#qiime-pick-rep-set" id="id70"><code class="docutils literal"><span class="pre">qiime_pick_rep_set</span></code></a></li>
<li><a class="reference internal" href="#qiime-align-seqs" id="id71"><code class="docutils literal"><span class="pre">qiime_align_seqs</span></code></a></li>
<li><a class="reference internal" href="#qiime-filter-alignment" id="id72"><code class="docutils literal"><span class="pre">qiime_filter_alignment</span></code></a></li>
<li><a class="reference internal" href="#qiime-assign-taxonomy" id="id73"><code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code></a></li>
<li><a class="reference internal" href="#qiime-make-phylogeny" id="id74"><code class="docutils literal"><span class="pre">qiime_make_phylogeny</span></code></a></li>
<li><a class="reference internal" href="#qiime-make-otu-table" id="id75"><code class="docutils literal"><span class="pre">qiime_make_otu_table</span></code></a></li>
<li><a class="reference internal" href="#qiime-filter-samples-from-otu-table" id="id76"><code class="docutils literal"><span class="pre">qiime_filter_samples_from_otu_table</span></code></a></li>
<li><a class="reference internal" href="#qiime-filter-otus" id="id77"><code class="docutils literal"><span class="pre">qiime_filter_otus</span></code></a></li>
<li><a class="reference internal" href="#qiime-sort-otu-table" id="id78"><code class="docutils literal"><span class="pre">qiime_sort_otu_table</span></code></a></li>
<li><a class="reference internal" href="#qiime-divers" id="id79"><code class="docutils literal"><span class="pre">qiime_divers</span></code></a></li>
</ul>
</div>
<span class="target" id="module-QIIME.qiime_prep"></span><div class="section" id="qiime-prep">
<h2><a class="toc-backref" href="#id66"><code class="docutils literal"><span class="pre">qiime_prep</span></code></a><a class="headerlink" href="#qiime-prep" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for preparing fastq reads for analysis with QIIME (1.9):</p>
<p>The reads stored in each sample are optinally joined and then set it a directory in such a way the downstream, QIIME’s demult can concatenate the sequences while saving the sample of origin.</p>
<p>The directory will contain symbolic links to the files to be used by demult in the following step.</p>
<div class="section" id="requires">
<h3>Requires<a class="headerlink" href="#requires" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts directory of links to files to use with QIIME:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;qiime.prep_links_dir&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If join is performed:</p>
<blockquote>
<div><ul>
<li><p class="first">puts the new joined reads in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fastq.J&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">puts the unjoined forward reads in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">puts the unjoined reverse reads in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="parameters-that-can-be-set">
<h3>Parameters that can be set<a class="headerlink" href="#parameters-that-can-be-set" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>join</td>
<td>none, join (or join_cat - not implemented)</td>
<td>Wheather to join paired reads.</td>
</tr>
<tr class="row-odd"><td>unjoined</td>
<td>forward, reverse, both or none</td>
<td>What to do with unjoined sequences? Use only forward, only reverse, both or none. If join is none, use this parameter to indicate which reads to take for analysis.</td>
</tr>
<tr class="row-even"><td>join_algo</td>
<td>forward, reverse, both or none</td>
<td>What to do with unjoined sequences?</td>
</tr>
<tr class="row-odd"><td>parameters</td>
<td>&#160;</td>
<td>Path to QIIME parameter file to be used downstream</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="lines-for-parameter-file">
<h3>Lines for parameter file<a class="headerlink" href="#lines-for-parameter-file" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_prep_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_prep</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">merge1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">join_paired_ends</span><span class="o">.</span><span class="n">py</span>
    <span class="n">join</span><span class="p">:</span> <span class="n">join</span>
    <span class="n">unjoined</span><span class="p">:</span> <span class="n">forward</span>
    <span class="n">parameters</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">qiime_params</span><span class="o">.</span><span class="n">txt</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">pe_join_method</span><span class="p">:</span> <span class="n">fastq</span><span class="o">-</span><span class="n">join</span>
</pre></div>
</div>
</div>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_demult"></span><div class="section" id="qiime-demult">
<h2><a class="toc-backref" href="#id67"><code class="docutils literal"><span class="pre">qiime_demult</span></code></a><a class="headerlink" href="#qiime-demult" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s multiple_split_libraries_fastq.py:</p>
<p>The reads from step qiime_prep are combined into one seqs.fna file.</p>
<p>..Note:: The module has not been tested on other types of data, such as undemultiplexed reads. It should work but there will probably be unexpected problems.</p>
<div class="section" id="id1">
<h3>Requires<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A directory of read files with smaple names coded in the file names, such as the directory produced by qiime_prep:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;qiime.prep_links_dir&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id2">
<h3>Output<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting <code class="docutils literal"><span class="pre">seqs.fna</span></code> file in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;qiime.demult_seqs&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;qiime.fasta&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id3">
<h3>Lines for parameter file<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_demult_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_demult</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_prep_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;/path/to/multiple_split_libraries_fastq.py&#39;</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">demultiplexing_method</span><span class="p">:</span> <span class="n">sampleid_by_file</span>
        <span class="o">--</span><span class="n">include_input_dir_path</span><span class="p">:</span> <span class="n">null</span>
        <span class="o">--</span><span class="n">parameter_fp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">qiime_params</span>
        <span class="o">--</span><span class="n">remove_filepath_in_name</span><span class="p">:</span> <span class="n">null</span>
</pre></div>
</div>
</div>
<div class="section" id="id4">
<h3>References<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_chimera"></span><div class="section" id="qiime-chimera">
<h2><a class="toc-backref" href="#id68"><code class="docutils literal"><span class="pre">qiime_chimera</span></code></a><a class="headerlink" href="#qiime-chimera" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s identify_chimeric_seqs.py:</p>
<p>The module can operate on the raw <code class="docutils literal"><span class="pre">seqs.fna</span></code> or on an aligned version. The latter is used for ChimeraSlayer and the former for usearch61</p>
<div class="section" id="id5">
<h3>Requires<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;qiime.fasta&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Alternatively, an aligned fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.aligned&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id6">
<h3>Output<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting list of chimeras in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;chimeras&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the filtered fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.chimera_removed&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>..Note:: When using <code class="docutils literal"><span class="pre">parallel_identify_chimeric_seqs.py</span></code>, the module tries to build the scripts appropriately. <strong>It is wise to check the parallel scripts before running them…</strong></p>
</div>
<div class="section" id="id7">
<h3>Parameters that can be set<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>method</td>
<td>usearch61 or ChimeraSlayer</td>
<td>Method to use for the analysis (passed to the –chimera_detection_method of <code class="docutils literal"><span class="pre">identify_chimeric_seqs.py</span></code></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id8">
<h3>Lines for parameter file<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_chimera_usrch</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_chimera</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_demult_1</span>
    <span class="c1"># script_path: &#39;{Vars.qiime_path}/parallel_identify_chimeric_seqs.py&#39;</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/identify_chimeric_seqs.py&#39;</span>
    <span class="n">method</span><span class="p">:</span>         <span class="n">usearch61</span> <span class="c1"># Or ChimeraSlayer. Will guess depending on existing files.</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="c1"># --jobs_to_start:              20</span>
        <span class="o">--</span><span class="n">aligned_reference_seqs_fp</span><span class="p">:</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">reference_files</span><span class="o">.</span><span class="n">otus_aligned</span>
        <span class="o">--</span><span class="n">reference_seqs_fp</span><span class="p">:</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">reference_files</span><span class="o">.</span><span class="n">otus</span>
</pre></div>
</div>
</div>
<div class="section" id="id9">
<h3>References<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_pick_otus"></span><div class="section" id="qiime-pick-otus">
<h2><a class="toc-backref" href="#id69"><code class="docutils literal"><span class="pre">qiime_pick_otus</span></code></a><a class="headerlink" href="#qiime-pick-otus" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">pick_otus.py</span></code></p>
<div class="section" id="id10">
<h3>Requires<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id11">
<h3>Output<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting OTU table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;otu_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id12">
<h3>Parameters that can be set<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id13">
<h3>Lines for parameter file<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_pick_otu_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_pick_otus</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_chimera_usrch</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/pick_otus.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id14">
<h3>References<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_pick_rep_set"></span><div class="section" id="qiime-pick-rep-set">
<h2><a class="toc-backref" href="#id70"><code class="docutils literal"><span class="pre">qiime_pick_rep_set</span></code></a><a class="headerlink" href="#qiime-pick-rep-set" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">pick_rep_set.py</span></code></p>
<div class="section" id="id15">
<h3>Requires<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">An OTU table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;otu_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id16">
<h3>Output<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Saves the original fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;qiime.full_fasta&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id17">
<h3>Parameters that can be set<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id18">
<h3>Lines for parameter file<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_rep_set_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_pick_rep_set</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_pick_otu_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/pick_rep_set.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id19">
<h3>References<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_align_seqs"></span><div class="section" id="qiime-align-seqs">
<h2><a class="toc-backref" href="#id71"><code class="docutils literal"><span class="pre">qiime_align_seqs</span></code></a><a class="headerlink" href="#qiime-align-seqs" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running <code class="docutils literal"><span class="pre">QIIME's</span> <span class="pre">align_seqs.py</span></code>:</p>
<p>Can be used for the parallel versions thereof: <code class="docutils literal"><span class="pre">parallel_align_seqs_pynast.py</span></code></p>
<div class="section" id="id20">
<h3>Requires<a class="headerlink" href="#id20" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id21">
<h3>Output<a class="headerlink" href="#id21" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting aligned fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.aligned&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Stores the old, unaligned version in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.unaligned&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>..Note:: When using <code class="docutils literal"><span class="pre">parallel_align_seqs_pynast.py</span></code>, the module tries to build the scripts appropriately. <strong>It is wise to check the parallel scripts before running them…</strong></p>
</div>
<div class="section" id="id22">
<h3>Parameters that can be set<a class="headerlink" href="#id22" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id23">
<h3>Lines for parameter file<a class="headerlink" href="#id23" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_align_para</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_align_seqs</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_rep_set_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/parallel_align_seqs_pynast.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">jobs_to_start</span><span class="p">:</span> <span class="mi">5</span>
        <span class="o">--</span><span class="n">retain_temp_files</span><span class="p">:</span> 
</pre></div>
</div>
</div>
<div class="section" id="id24">
<h3>References<a class="headerlink" href="#id24" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_filter_alignment"></span><div class="section" id="qiime-filter-alignment">
<h2><a class="toc-backref" href="#id72"><code class="docutils literal"><span class="pre">qiime_filter_alignment</span></code></a><a class="headerlink" href="#qiime-filter-alignment" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">filter_alignment.py</span></code></p>
<div class="section" id="id25">
<h3>Requires<a class="headerlink" href="#id25" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id26">
<h3>Output<a class="headerlink" href="#id26" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting aligned fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Saves the original unaligned fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.aligned_unfiltered&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id27">
<h3>Parameters that can be set<a class="headerlink" href="#id27" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id28">
<h3>Lines for parameter file<a class="headerlink" href="#id28" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_filt_align_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_filter_alignment</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_align_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/filter_alignment.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id29">
<h3>References<a class="headerlink" href="#id29" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_assign_taxonomy"></span><div class="section" id="qiime-assign-taxonomy">
<h2><a class="toc-backref" href="#id73"><code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code></a><a class="headerlink" href="#qiime-assign-taxonomy" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">assign_taxonomy.py</span></code></p>
<p>Can also be used to run the parallel versions of the program:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">parallel_assign_taxonomy_blast.py</span></code></li>
<li><code class="docutils literal"><span class="pre">parallel_assign_taxonomy_rdp.py</span></code></li>
<li><code class="docutils literal"><span class="pre">parallel_assign_taxonomy_uclust.py</span></code></li>
</ul>
</div></blockquote>
<div class="section" id="id30">
<h3>Requires<a class="headerlink" href="#id30" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id31">
<h3>Output<a class="headerlink" href="#id31" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting list of chimeras in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;taxonomy&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>..Note:: When using the parallel version, the module tries to build the scripts appropriately. <strong>It is wise to check the parallel scripts before running them…</strong></p>
</div>
<div class="section" id="id32">
<h3>Parameters that can be set<a class="headerlink" href="#id32" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id33">
<h3>Lines for parameter file<a class="headerlink" href="#id33" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_tax_asn_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_assign_taxonomy</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_rep_set_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/parallel_assign_taxonomy_rdp.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">confidence</span><span class="p">:</span> <span class="mf">0.5</span>
        <span class="o">--</span><span class="n">id_to_taxonomy_fp</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">reference_files</span><span class="o">.</span><span class="n">id_to_taxonomy</span><span class="p">}</span>
        <span class="o">--</span><span class="n">jobs_to_start</span><span class="p">:</span> <span class="mi">20</span>
        <span class="o">--</span><span class="n">rdp_max_memory</span><span class="p">:</span> <span class="mi">50000</span>
        <span class="o">--</span><span class="n">reference_seqs_fp</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">reference_files</span><span class="o">.</span><span class="n">otus</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id34">
<h3>References<a class="headerlink" href="#id34" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_make_phylogeny"></span><div class="section" id="qiime-make-phylogeny">
<h2><a class="toc-backref" href="#id74"><code class="docutils literal"><span class="pre">qiime_make_phylogeny</span></code></a><a class="headerlink" href="#qiime-make-phylogeny" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">make_phylogeny.py</span></code></p>
<div class="section" id="id35">
<h3>Requires<a class="headerlink" href="#id35" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id36">
<h3>Output<a class="headerlink" href="#id36" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting OTU table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;phylotree&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id37">
<h3>Parameters that can be set<a class="headerlink" href="#id37" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id38">
<h3>Lines for parameter file<a class="headerlink" href="#id38" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_phylo_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_make_phylogeny</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_filt_align_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/make_phylogeny.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id39">
<h3>References<a class="headerlink" href="#id39" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_make_otu_table"></span><div class="section" id="qiime-make-otu-table">
<h2><a class="toc-backref" href="#id75"><code class="docutils literal"><span class="pre">qiime_make_otu_table</span></code></a><a class="headerlink" href="#qiime-make-otu-table" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">make_otu_table.py</span></code>:</p>
<p>The module creates a BIOM table based on the OTU table and a taxonomy assignment if avaliable (will be available if the <code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code> is in the branch).</p>
<p>If chimera checking has been performed, the suspected chimeric sequences will be removed from the BIOM table.</p>
<p>The module also adds code for creating a summary of the BIOM table and a tab-delimited version thereof.</p>
<div class="section" id="id40">
<h3>Requires<a class="headerlink" href="#id40" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">An OTU table:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;otu_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="optional">
<h3>Optional<a class="headerlink" href="#optional" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A taxonomy assignment of the sequences:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;taxonomy&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id41">
<h3>Output<a class="headerlink" href="#id41" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the BIOM table in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table summary in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table in tab-delimited format in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If a fasta.chimera_removed file exists, will put the unfiltered BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;unfiltered_biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id42">
<h3>Parameters that can be set<a class="headerlink" href="#id42" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>skip_summary</td>
<td>&#160;</td>
<td>If passed, will not create the BIOM table summary.</td>
</tr>
<tr class="row-odd"><td>skip_tsv</td>
<td>&#160;</td>
<td>If passed, will not create the tsv version of the BIOM table.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id43">
<h3>Lines for parameter file<a class="headerlink" href="#id43" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_mk_otu_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_make_otu_table</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_phylo_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/make_otu_table.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="c1"># skip_summary:</span>
    <span class="c1"># skip_tsv:</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">mapping_fp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">qiime1_mapping</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
</div>
<div class="section" id="id44">
<h3>References<a class="headerlink" href="#id44" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_filter_samples_from_otu_table"></span><div class="section" id="qiime-filter-samples-from-otu-table">
<h2><a class="toc-backref" href="#id76"><code class="docutils literal"><span class="pre">qiime_filter_samples_from_otu_table</span></code></a><a class="headerlink" href="#qiime-filter-samples-from-otu-table" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">filter_samples_from_otu_table.py</span></code></p>
<div class="section" id="id45">
<h3>Requires<a class="headerlink" href="#id45" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id46">
<h3>Output<a class="headerlink" href="#id46" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table summary in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table in tab-delimited format in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the unfiltered BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;prefilter_biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id47">
<h3>Parameters that can be set<a class="headerlink" href="#id47" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>skip_summary</td>
<td>&#160;</td>
<td>If passed, will not create the BIOM table summary.</td>
</tr>
<tr class="row-odd"><td>skip_tsv</td>
<td>&#160;</td>
<td>If passed, will not create the tsv version of the BIOM table.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id48">
<h3>Lines for parameter file<a class="headerlink" href="#id48" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">filt_samp_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_filter_samples_from_otu_table</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_mk_otu_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/filter_samples_from_otu_table.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">mapping_fp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">mapping</span><span class="o">.</span><span class="n">txt</span>
        <span class="o">--</span><span class="n">min_count</span><span class="p">:</span> <span class="mi">100000</span>
</pre></div>
</div>
</div>
<div class="section" id="id49">
<h3>References<a class="headerlink" href="#id49" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_filter_otus"></span><div class="section" id="qiime-filter-otus">
<h2><a class="toc-backref" href="#id77"><code class="docutils literal"><span class="pre">qiime_filter_otus</span></code></a><a class="headerlink" href="#qiime-filter-otus" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">filter_otus_from_otu_table.py</span></code></p>
<div class="section" id="id50">
<h3>Requires<a class="headerlink" href="#id50" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id51">
<h3>Output<a class="headerlink" href="#id51" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table summary in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table in tab-delimited format in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the unfiltered BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;prefilter_biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id52">
<h3>Parameters that can be set<a class="headerlink" href="#id52" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>skip_summary</td>
<td>&#160;</td>
<td>If passed, will not create the BIOM table summary.</td>
</tr>
<tr class="row-odd"><td>skip_tsv</td>
<td>&#160;</td>
<td>If passed, will not create the tsv version of the BIOM table.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id53">
<h3>Lines for parameter file<a class="headerlink" href="#id53" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_filt_otus_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_filter_otus</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">filt_samp_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/filter_otus_from_otu_table.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">min_count_fraction</span><span class="p">:</span> <span class="mf">0.00005</span>
        <span class="o">--</span><span class="n">min_samples</span><span class="p">:</span> <span class="mi">10</span>
</pre></div>
</div>
</div>
<div class="section" id="id54">
<h3>References<a class="headerlink" href="#id54" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_sort_otu_table"></span><div class="section" id="qiime-sort-otu-table">
<h2><a class="toc-backref" href="#id78"><code class="docutils literal"><span class="pre">qiime_sort_otu_table</span></code></a><a class="headerlink" href="#qiime-sort-otu-table" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">sort_otu_table.py</span></code></p>
<div class="section" id="id55">
<h3>Requires<a class="headerlink" href="#id55" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id56">
<h3>Output<a class="headerlink" href="#id56" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table summary in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table in tab-delimited format in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id57">
<h3>Parameters that can be set<a class="headerlink" href="#id57" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>skip_summary</td>
<td>&#160;</td>
<td>If passed, will not create the BIOM table summary.</td>
</tr>
<tr class="row-odd"><td>skip_tsv</td>
<td>&#160;</td>
<td>If passed, will not create the tsv version of the BIOM table.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id58">
<h3>Lines for parameter file<a class="headerlink" href="#id58" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_sort_otus_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_sort_otu_table</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">filt_samp_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/sort_otu_table.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">sort_field</span><span class="p">:</span>   <span class="n">XXX</span>
</pre></div>
</div>
</div>
<div class="section" id="id59">
<h3>References<a class="headerlink" href="#id59" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_divers"></span><div class="section" id="qiime-divers">
<h2><a class="toc-backref" href="#id79"><code class="docutils literal"><span class="pre">qiime_divers</span></code></a><a class="headerlink" href="#qiime-divers" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">core_diversity_analyses.py</span></code>:</p>
<p>The module creates a BIOM table based on the OTU table and a taxonomy assignment if avaliable (will be available if the <code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code> is in the branch).</p>
<p>If chimera checking has been performed, the suspected chimeric sequences will be removed from the BIOM table.</p>
<p>The module also adds code for creating a summary of the BIOM table and a tab-delimited version thereof.</p>
<div class="section" id="id60">
<h3>Requires<a class="headerlink" href="#id60" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BIOM table:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id61">
<h3>Optional<a class="headerlink" href="#id61" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A phylogenetic tree:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;phylotree&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id62">
<h3>Output<a class="headerlink" href="#id62" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the core diversity directory name in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;diversity&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id63">
<h3>Parameters that can be set<a class="headerlink" href="#id63" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>–mapping_fp</td>
<td>&#160;</td>
<td>A path to the qiime mapping file (if not set, will use the mapping file passed in <code class="docutils literal"><span class="pre">qiime_prep</span></code>.</td>
</tr>
<tr class="row-odd"><td>–parameter_fp</td>
<td>&#160;</td>
<td>A path to a qiime parameter file.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id64">
<h3>Lines for parameter file<a class="headerlink" href="#id64" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_divers_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_divers</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_filt_otus_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">QIIME</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">core_diversity_analyses</span><span class="o">.</span><span class="n">py</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">sampling_depth</span><span class="p">:</span> <span class="mi">109897</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">categories</span><span class="p">:</span> <span class="n">Disease</span><span class="p">,</span><span class="n">sex</span>
        <span class="o">--</span><span class="n">parameter_fp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">parameter_file</span>
</pre></div>
</div>
</div>
<div class="section" id="id65">
<h3>References<a class="headerlink" href="#id65" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="SequenceAnnotation.html" class="btn btn-neutral float-right" title="Sequence Annotation" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="PreparationAndQC.html" class="btn btn-neutral" title="Preparation and QC" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Menachem Sklarz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>