

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Transcriptome Assembly &mdash; NeatSeq_Flow Module Repository 1.1.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="NeatSeq_Flow Module Repository 1.1.0 documentation" href="../index.html"/>
        <link rel="next" title="Variant-related Tasks" href="Variants.html"/>
        <link rel="prev" title="Sequence-Searching Related Tasks" href="SequenceSearching.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> NeatSeq_Flow Module Repository
          

          
            
            <img src="../_static/logo_Hor.png" class="logo" />
          
          </a>

          
            
            
              <div class="version">
                1.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <p class="caption"><span class="caption-text">Modules</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="BAMConversion.html">BAM Conversion to Other Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="ChIPseq.html">ChIP-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="GenericModule.html">Generic Module</a></li>
<li class="toctree-l1"><a class="reference internal" href="GenomeAssembly.html">Genome Assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="Mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="Metagenomics.html">Metagenomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="Microbiology.html">Microbiology</a></li>
<li class="toctree-l1"><a class="reference internal" href="PreparationAndQC.html">Preparation and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="QIIME.html">QIIME (version 1.9)</a></li>
<li class="toctree-l1"><a class="reference internal" href="SequenceAnnotation.html">Sequence Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="SequenceClustering.html">Sequence Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="SequenceSearching.html">Sequence-Searching Related Tasks</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Transcriptome Assembly</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#trinity"><code class="docutils literal"><span class="pre">trinity</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#requires">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parameters-that-can-be-set">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#lines-for-parameter-file">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#add-trinity-tags"><code class="docutils literal"><span class="pre">add_trinity_tags</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Lines for parameter file</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#trinity-mapping"><code class="docutils literal"><span class="pre">trinity_mapping</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id4">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id6">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id7">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id8">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id9"><code class="docutils literal"><span class="pre">trinity</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id11">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id13">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id14">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id15">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-rsem">Module <code class="docutils literal"><span class="pre">RSEM</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id16">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id17">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id18">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#comments">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id19">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id20">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#quast"><code class="docutils literal"><span class="pre">quast</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id21">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id22">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id24">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id25">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Variants.html">Variant-related Tasks</a></li>
<li class="toctree-l1"><a class="reference internal" href="VariousReportingPrograms.html">Various Reporting Programs</a></li>
</ul>
<p class="caption"><span class="caption-text">Workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/Assembly_Indexing_mapping.html">Assembly and Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/BLAST_db.html">BLAST on <cite>fasta</cite> files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/BLAST_fasta.html">Assembly and Searching</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/ChIP_seq_bowtie2.html">ChIP-seq analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/Metagenomics.html">Metagenomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/RNA_seq_Trinity.html">Non-Model Transcriptomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/RNA_seq_reference.html">Reference-based Transcriptomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/variant_calling.html">Variant calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/Clustering.html">Read clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Workflow_docs/QIIME_workflow.html">QIIME analysis</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">NeatSeq_Flow Module Repository</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
    <li>Transcriptome Assembly</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/Module_docs/TranscriptomeAssembly.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="transcriptome-assembly">
<h1>Transcriptome Assembly<a class="headerlink" href="#transcriptome-assembly" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="modules-included-in-this-section">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#trinity" id="id26"><code class="docutils literal"><span class="pre">trinity</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#add-trinity-tags" id="id27"><code class="docutils literal"><span class="pre">add_trinity_tags</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#trinity-mapping" id="id28"><code class="docutils literal"><span class="pre">trinity_mapping</span></code></a></li>
<li><a class="reference internal" href="#id9" id="id29"><code class="docutils literal"><span class="pre">trinity</span></code></a></li>
<li><a class="reference internal" href="#module-rsem" id="id30">Module <code class="docutils literal"><span class="pre">RSEM</span></code></a></li>
<li><a class="reference internal" href="#quast" id="id31"><code class="docutils literal"><span class="pre">quast</span></code> <sup>*</sup></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.RNA_seq.trinity"></span><div class="section" id="trinity">
<h2><a class="toc-backref" href="#id26"><code class="docutils literal"><span class="pre">trinity</span></code> <sup>*</sup></a><a class="headerlink" href="#trinity" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for RNA_seq assembly using the <a class="reference external" href="https://github.com/trinityrnaseq/trinityrnaseq/wiki">Trinity assembler</a>.</p>
<div class="section" id="requires">
<h3>Requires<a class="headerlink" href="#requires" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">fastq</span></code> files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="output">
<h3>Output:<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">puts <code class="docutils literal"><span class="pre">fasta</span></code> output files in the following slots:</p>
<blockquote>
<div><ul>
<li><p class="first">for sample-wise assembly:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;Trinity.contigs&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">for project-wise assembly:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;Trinity.contigs&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="parameters-that-can-be-set">
<h3>Parameters that can be set<a class="headerlink" href="#parameters-that-can-be-set" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample|project</td>
<td>Set if project-wide fasta slot should be used</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="lines-for-parameter-file">
<h3>Lines for parameter file<a class="headerlink" href="#lines-for-parameter-file" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">trinity1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>     <span class="n">trinity</span>
    <span class="n">base</span><span class="p">:</span>       <span class="n">trin_tags1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">Trinity</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="n">node</span><span class="p">:</span>      <span class="n">sge213</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>       <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">grid_conf</span><span class="p">:</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SGE_Trinity_conf</span><span class="o">.</span><span class="n">txt</span>
        <span class="o">--</span><span class="n">CPU</span><span class="p">:</span>              <span class="mi">20</span>
        <span class="o">--</span><span class="n">seqType</span><span class="p">:</span>          <span class="n">fq</span>
        <span class="o">--</span><span class="n">JM</span><span class="p">:</span>               <span class="mi">140</span><span class="n">G</span>
        <span class="o">--</span><span class="n">min_kmer_cov</span><span class="p">:</span>     <span class="mi">2</span>
        <span class="o">--</span><span class="n">full_cleanup</span><span class="p">:</span>
</pre></div>
</div>
</div>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<p>Grabherr, M.G., Haas, B.J., Yassour, M., Levin, J.Z., Thompson, D.A., Amit, I., Adiconis, X., Fan, L., Raychowdhury, R., Zeng, Q. and Chen, Z., 2011. <strong>Trinity: reconstructing a full-length transcriptome without a genome from RNA-Seq data</strong>. <em>Nature biotechnology</em>, 29(7), p.644.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.RNA_seq.add_trinity_tags"></span><div class="section" id="add-trinity-tags">
<h2><a class="toc-backref" href="#id27"><code class="docutils literal"><span class="pre">add_trinity_tags</span></code> <sup>*</sup></a><a class="headerlink" href="#add-trinity-tags" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for adding the tags required by Trinity to the ends of the read names.
See the <a class="reference external" href="https://github.com/trinityrnaseq/trinityrnaseq/wiki/Running-Trinity#strand_specific_assembly">Strand specific assembly</a> section of the Trinity manual.</p>
<p>The module uses awk, so you don’t need to pass a <code class="docutils literal"><span class="pre">script_path</span></code>. 
Since you must pass a <code class="docutils literal"><span class="pre">script_path</span></code>, just leave it blank.</p>
<div class="section" id="id1">
<h3>Requires<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">fastq</span></code> files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id2">
<h3>Output:<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">puts <code class="docutils literal"><span class="pre">fastq</span></code> output files (with added tags) in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id3">
<h3>Lines for parameter file<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">trintags</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>      <span class="n">add_trinity_tags</span>
    <span class="n">base</span><span class="p">:</span>        <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="n">NOT_USED</span>
</pre></div>
</div>
</div>
</div>
<span class="target" id="module-Transcriptomics.trinity_mapping"></span><div class="section" id="trinity-mapping">
<h2><a class="toc-backref" href="#id28"><code class="docutils literal"><span class="pre">trinity_mapping</span></code></a><a class="headerlink" href="#trinity-mapping" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for running <code class="docutils literal"><span class="pre">align_and_estimate_abundance.pl</span></code> on a Trinity assembly and the raw reads.</p>
<p>See the script documentation <a class="reference external" href="https://github.com/trinityrnaseq/trinityrnaseq/wiki/Trinity-Transcript-Quantification#estimating-transcript-abundance">here</a>.</p>
<div class="section" id="id4">
<h3>Requires<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">fastq</span></code> files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">A Trinity assembly in one of (depending on <code class="docutils literal"><span class="pre">scope</span></code>)</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id5">
<h3>Output:<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts BAM output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;isoforms.results&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;genes.results&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id6">
<h3>Parameters that can be set<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample|project</td>
<td>Set if project-wide fasta slot should be used</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id7">
<h3>Lines for parameter file<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">trin_map1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>               <span class="n">trinity_mapping</span>
    <span class="n">base</span><span class="p">:</span>                 <span class="n">trinity1</span>
    <span class="n">script_path</span><span class="p">:</span>          <span class="o">/</span><span class="n">fastspace</span><span class="o">/</span><span class="n">bioinfo_apps</span><span class="o">/</span><span class="n">trinityrnaseq_r20140717</span><span class="o">/</span><span class="n">util</span><span class="o">/</span><span class="n">align_and_estimate_abundance</span><span class="o">.</span><span class="n">pl</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">est_method</span><span class="p">:</span>     <span class="n">RSEM</span>
        <span class="o">--</span><span class="n">aln_method</span><span class="p">:</span>     <span class="n">bowtie</span>
        <span class="o">--</span><span class="n">trinity_mode</span><span class="p">:</span>
        <span class="o">--</span><span class="n">seqType</span><span class="p">:</span>        <span class="n">fq</span>
</pre></div>
</div>
</div>
<div class="section" id="id8">
<h3>References<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<p>Grabherr, M.G., Haas, B.J., Yassour, M., Levin, J.Z., Thompson, D.A., Amit, I., Adiconis, X., Fan, L., Raychowdhury, R., Zeng, Q. and Chen, Z., 2011. <strong>Trinity: reconstructing a full-length transcriptome without a genome from RNA-Seq data</strong>. <em>Nature biotechnology</em>, 29(7), p.644.</p>
</div>
</div>
<span class="target" id="module-Transcriptomics.trinmap_statistics"></span><div class="section" id="id9">
<h2><a class="toc-backref" href="#id29"><code class="docutils literal"><span class="pre">trinity</span></code></a><a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for running <code class="docutils literal"><span class="pre">abundance_estimates_to_matrix.pl</span></code> on genes or isoforms counts tables produced by <code class="docutils literal"><span class="pre">align_and_estimate_abundance.pl</span></code></p>
<p>See the script documentation <a class="reference external" href="https://github.com/trinityrnaseq/trinityrnaseq/wiki/Trinity-Transcript-Quantification#building-expression-matrices">here</a>.</p>
<p>This conversion makes sense at the project level - combining all sample matrices into a single, normalized, comparison table. However, for completeness, we included a sample scope option for running the script in each sample separately.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><code class="docutils literal"><span class="pre">scope</span></code> is not defined for this module. It only makes sense to run <code class="docutils literal"><span class="pre">abundance_estimates_to_matrix</span></code> when comparing many samples against a single assembly</p>
</div>
<div class="section" id="id11">
<h3>Requires<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Either <code class="docutils literal"><span class="pre">genes.results</span></code> or <code class="docutils literal"><span class="pre">isoforms.results</span></code> files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;genes.results&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;isoforms.results&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id12">
<h3>Output:<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Creates the following files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">&lt;project&gt;.counts.matrix</span></code>                        in <code class="docutils literal"><span class="pre">self.sample_data[&quot;counts.matrix&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">&lt;project&gt;.not_cross_norm.fpkm.tmp</span></code>              in <code class="docutils literal"><span class="pre">self.sample_data[&quot;not_cross_norm.fpkm.tmp&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">&lt;project&gt;.not_cross_norm.fpkm.tmp.TMM_info.txt</span></code> in <code class="docutils literal"><span class="pre">self.sample_data[&quot;not_cross_norm.fpkm.tmp.TMM_info.txt&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">&lt;project&gt;.TMM.fpkm.matrix</span></code>                      in <code class="docutils literal"><span class="pre">self.sample_data[&quot;TMM.fpkm.matrix&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id13">
<h3>Parameters that can be set<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>use_isoforms</td>
<td>&#160;</td>
<td>Use ‘isoforms.results’ matrix. If not passed, use ‘genes.results’</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id14">
<h3>Lines for parameter file<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">trin_map_stats</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>             <span class="n">trinmap_statistics</span>
    <span class="n">base</span><span class="p">:</span>               <span class="n">trin_map1</span>
    <span class="n">script_path</span><span class="p">:</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">abundance_estimates_to_matrix</span><span class="o">.</span><span class="n">pl</span>
    <span class="n">use_isoforms</span><span class="p">:</span>       
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">est_method</span><span class="p">:</span>   <span class="n">RSEM</span>
</pre></div>
</div>
</div>
<div class="section" id="id15">
<h3>References<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<p>Grabherr, M.G., Haas, B.J., Yassour, M., Levin, J.Z., Thompson, D.A., Amit, I., Adiconis, X., Fan, L., Raychowdhury, R., Zeng, Q. and Chen, Z., 2011. <strong>Trinity: reconstructing a full-length transcriptome without a genome from RNA-Seq data</strong>. <em>Nature biotechnology</em>, 29(7), p.644.</p>
</div>
</div>
<span class="target" id="module-Liron.RSEM_module.RSEM"></span><div class="section" id="module-rsem">
<h2><a class="toc-backref" href="#id30">Module <code class="docutils literal"><span class="pre">RSEM</span></code></a><a class="headerlink" href="#module-rsem" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module for running RSEM</dd>
</dl>
<div class="section" id="id16">
<h3>Requires<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>fastq file in </dt>
<dd>self.sample_data[sample][“fastq.F”]
self.sample_data[sample][“fastq.R”]
self.sample_data[sample][“fastq.S”]</dd>
<dt>or bam file in </dt>
<dd>self.sample_data[sample][“bam”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id17">
<h3>Output<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>puts output bam files (if the input is fastq) in:</dt>
<dd>self.sample_data[sample][“bam”]
self.sample_data[sample][“unsorted_bam”]</dd>
<dt>puts the location of RSEM results in:</dt>
<dd>self.sample_data[sample][“RSEM”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id18">
<h3>Parameters that can be set<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>mode</td>
<td>transcriptome/genome</td>
<td>Is the reference is a genome or a transcriptome?</td>
</tr>
<tr class="row-odd"><td>gff3</td>
<td>None</td>
<td>Use if the mode is genome and the annotation file is in gff3 format</td>
</tr>
<tr class="row-even"><td>del_unsorted_bam</td>
<td>None</td>
<td>Delete unsorted bam files to save space</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="comments">
<h3>Comments<a class="headerlink" href="#comments" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id19">
<h3>Lines for parameter file<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Step_Name</span><span class="p">:</span>                                                   <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">RSEM</span>                                             <span class="c1"># Name of the module used</span>
    <span class="n">base</span><span class="p">:</span>                                                    <span class="c1"># Name of the step [or list of names] to run after [must be after a bam file generator step or merge with fastq files]</span>
    <span class="n">script_path</span><span class="p">:</span>                                             <span class="c1"># Command for running the RSEM script </span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>                                                 <span class="c1"># Number of CPUs to reserve for this analysis</span>
    <span class="n">mode</span><span class="p">:</span>                                                    <span class="c1"># transcriptome or genome</span>
    <span class="n">annotation</span><span class="p">:</span>                                              <span class="c1"># For Genome mode: the location of GTF file [the default] , for GFF3 use the gff3 flag. For Transcriptome mode: transcript-to-gene-map file.</span>
                                                             <span class="c1"># If annotation is set to Trinity the transcript-to-gene-map file will be generated using the from_Trinity_to_gene_map script</span>
    <span class="n">from_Trinity_to_gene_map_script_path</span><span class="p">:</span>                    <span class="c1"># If the mode is transcriptome and the reference was assembled using Trinity it is possible to generate the transcript-to-gene-map file automatically using this script</span>
                                                             <span class="c1"># If annotation is set to Trinity and this line is empty or missing it will try using the module&#39;s associated script</span>
    <span class="n">gff3</span><span class="p">:</span>                                                    <span class="c1"># Use if the mode is genome and the annotation file is in gff3 format</span>
    <span class="n">mapper</span><span class="p">:</span>                                                  <span class="c1"># bowtie/bowtie2/star </span>
    <span class="n">mapper_path</span><span class="p">:</span>                                             <span class="c1"># Location of mapper script</span>
    <span class="n">rsem_prepare_reference_script_path</span><span class="p">:</span>                      <span class="c1"># Location of preparing reference script</span>
    <span class="n">plot_stat</span><span class="p">:</span>                                               <span class="c1"># Generate statistical plots</span>
    <span class="n">plot_stat_script_path</span><span class="p">:</span>                                   <span class="c1"># Location of statistical plot generating script</span>
    <span class="n">reference</span><span class="p">:</span>                                               <span class="c1"># The reference genome/transcriptome location [FASTA file]</span>
    <span class="n">rsem_generate_data_matrix_script_path</span><span class="p">:</span>                   <span class="c1"># Location of the final matrix generating script</span>
                                                             <span class="c1"># If this line is empty or missing it will try using the module&#39;s associated script</span>
    <span class="n">del_unsorted_bam</span><span class="p">:</span>                                        <span class="c1"># Delete unsorted bam files to save space.</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">append</span><span class="o">-</span><span class="n">names</span><span class="p">:</span>                                      <span class="c1"># RSEM will append gene_name/transcript_name to the result files</span>
        <span class="o">--</span><span class="n">estimate</span><span class="o">-</span><span class="n">rspd</span><span class="p">:</span>                                     <span class="c1"># Enables RSEM to learn from the data how the reads are distributed across a transcript</span>
        <span class="o">-</span><span class="n">p</span><span class="p">:</span>                                                  <span class="c1"># Number of CPUs to use in this analysis</span>
        <span class="o">--</span><span class="n">bam</span><span class="p">:</span>                                               <span class="c1"># Will use bam files and not fastq</span>
        <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">bam</span><span class="o">-</span><span class="n">output</span><span class="p">:</span>
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">genome</span><span class="o">-</span><span class="n">bam</span><span class="p">:</span>                                 <span class="c1"># Alignments in genomic coordinates (only if mode is genome)</span>
</pre></div>
</div>
</div>
<div class="section" id="id20">
<h3>References<a class="headerlink" href="#id20" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<span class="target" id="module-main_NSF_classes.Assembly.quast"></span><div class="section" id="quast">
<h2><a class="toc-backref" href="#id31"><code class="docutils literal"><span class="pre">quast</span></code> <sup>*</sup></a><a class="headerlink" href="#quast" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running quast on fasta assemblies:</p>
<p>QUAST is executed on the fasta file along the following lines:</p>
<ul class="simple">
<li>If ‘scope’ is specified, the appropriate fasta will be used. An error will occur if the fasta does not exist.</li>
<li>If ‘scope’ is not specified, if a project-wide fasta exists, it will be used. Otherwise, sample-wise fasta files will be used. If none exist, an error will occur.</li>
</ul>
<div class="section" id="id21">
<h3>Requires<a class="headerlink" href="#id21" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id22">
<h3>Output<a class="headerlink" href="#id22" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output directory in one of:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;quast&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;quast&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id23">
<h3>Parameters that can be set<a class="headerlink" href="#id23" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project or sample bowtie2 index.</td>
</tr>
<tr class="row-odd"><td>compare_mode</td>
<td>&#160;</td>
<td>If ‘scope’ is ‘sample’, specifies wether to analyse each sample separately or to create a single comparison report for all samples.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id24">
<h3>Lines for parameter file<a class="headerlink" href="#id24" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">quast1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">quast</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">spades1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">quast</span><span class="o">.</span><span class="n">py</span>
    <span class="n">compare_mode</span><span class="p">:</span> <span class="n">null</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">fast</span><span class="p">:</span> <span class="n">null</span>
</pre></div>
</div>
</div>
<div class="section" id="id25">
<h3>References<a class="headerlink" href="#id25" title="Permalink to this headline">¶</a></h3>
<p>Gurevich, A., Saveliev, V., Vyahhi, N. and Tesler, G., 2013. <strong>QUAST: quality assessment tool for genome assemblies</strong>. <em>Bioinformatics</em>, 29(8), pp.1072-1075.</p>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Variants.html" class="btn btn-neutral float-right" title="Variant-related Tasks" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="SequenceSearching.html" class="btn btn-neutral" title="Sequence-Searching Related Tasks" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Menachem Sklarz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>