

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Preparation and QC &mdash; NeatSeq_Flow Module Repository 1.1.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="NeatSeq_Flow Module Repository 1.1.0 documentation" href="index.html"/>
        <link rel="next" title="Assembly and Mapping" href="Workflow_docs/Assembly_Indexing_mapping.html"/>
        <link rel="prev" title="NeatSeq_Flow’s Module and Workflow Repository" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> NeatSeq_Flow Module Repository
          

          
            
            <img src="_static/logo_Hor.png" class="logo" />
          
          </a>

          
            
            
              <div class="version">
                1.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <p class="caption"><span class="caption-text">Modules</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Preparation and QC</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#merge"><code class="docutils literal"><span class="pre">merge</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#requires">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parameters-that-can-be-set">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#lines-for-parameter-file">Lines for parameter file</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastqc-html"><code class="docutils literal"><span class="pre">fastqc_html</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#trimmo"><code class="docutils literal"><span class="pre">trimmo</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id4">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id6">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id7">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id8">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#multiqc"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id9">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id10">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id11">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-cutadapt">Module <code class="docutils literal"><span class="pre">Cutadapt</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id13">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id14">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id15">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#comments">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id16">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id17">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-trim-galore">Module <code class="docutils literal"><span class="pre">Trim_Galore</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id18">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id19">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id20">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id21">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id22">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#merge-project"><code class="docutils literal"><span class="pre">merge_project</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id24">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id25">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id26">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id27">Lines for parameter file</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#mapping">Mapping</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#bowtie2-builder"><code class="docutils literal"><span class="pre">bowtie2_builder</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id29">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id30">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id31">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id32">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id33">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bowtie2-mapper"><code class="docutils literal"><span class="pre">bowtie2_mapper</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id34">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id35">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id36">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id37">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id38">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bowtie1-builder"><code class="docutils literal"><span class="pre">bowtie1_builder</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id39">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id40">output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id41">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id42">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id43">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bowtie1-mapper"><code class="docutils literal"><span class="pre">bowtie1_mapper</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id44">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id45">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id46">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id47">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id48">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bwa-builder"><code class="docutils literal"><span class="pre">bwa_builder</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id49">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id50">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id51">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id52">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id53">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bwa-mapper"><code class="docutils literal"><span class="pre">bwa_mapper</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id54">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id55">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id56">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id57">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id58">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#star-mapper"><code class="docutils literal"><span class="pre">STAR_mapper</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id59">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id60">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id61">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id62">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id63">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#star-builder"><code class="docutils literal"><span class="pre">STAR_builder</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id64">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id65">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id66">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id67">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id68">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#samtools"><code class="docutils literal"><span class="pre">samtools</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id69">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id70">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id71"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id72">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id73">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id74">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id75">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#genome-assembly">Genome Assembly</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#clc-assembl"><code class="docutils literal"><span class="pre">clc_assembl</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id77">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id78">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id79">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id80">Lines for parameter file</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#megahit-assembl"><code class="docutils literal"><span class="pre">megahit_assembl</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id81">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id82">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id83">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id84">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id85">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#spades-assembl"><code class="docutils literal"><span class="pre">spades_assembl</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id86">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id87">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id88">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id89">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id90">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#transcriptome-assembly">Transcriptome Assembly</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#trinity"><code class="docutils literal"><span class="pre">trinity</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id92">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id93">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id94">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id95">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id96">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#add-trinity-tags"><code class="docutils literal"><span class="pre">add_trinity_tags</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id97">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id98">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id99">Lines for parameter file</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-rsem">Module <code class="docutils literal"><span class="pre">RSEM</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id100">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id101">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id102">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id103">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id104">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id105">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#assembly-qc">Assembly QC</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#quast"><code class="docutils literal"><span class="pre">quast</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id107">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id108">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id109">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id110">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id111">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#sequence-annotation">Sequence Annotation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-prokka">Module <code class="docutils literal"><span class="pre">Prokka</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id113">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id114">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id115">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id116">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id117">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id118">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#prokka-old"><code class="docutils literal"><span class="pre">prokka_old</span></code>  <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id119">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id120">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id121">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id122">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id123">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id124">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#sequence-clustering">Sequence Clustering</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#cd-hit"><code class="docutils literal"><span class="pre">cd_hit</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id126">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id127">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id128">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id129">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id130">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#vsearch-cluster"><code class="docutils literal"><span class="pre">vsearch_cluster</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id131">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id132">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id133">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id134">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id135">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#vsearch-derepel"><code class="docutils literal"><span class="pre">vsearch_derepel</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id136">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id137">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id138">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id139">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id140">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#metagenomics">Metagenomics</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#humann2"><code class="docutils literal"><span class="pre">HUMAnN2</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id142">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id143">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id144">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id145">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id146">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#kaiju"><code class="docutils literal"><span class="pre">kaiju</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id147">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id148">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id149">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id150">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id151">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#kraken"><code class="docutils literal"><span class="pre">kraken</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id152">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id153">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id154">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id155">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id156">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#metaphlan2"><code class="docutils literal"><span class="pre">metaphlan2</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id157">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id158">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id159">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id160">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id161">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#microbiology">Microbiology</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#card-rgi"><code class="docutils literal"><span class="pre">CARD_RGI</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id163">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id164">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id165">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id166">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id167">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id168">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-cgmlst-and-mlst-typing">Module <code class="docutils literal"><span class="pre">cgMLST_and_MLST_typing</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id169">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id170">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id171">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id172">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id173">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id174">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-roary">Module <code class="docutils literal"><span class="pre">Roary</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id175">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id176">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id177">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id178">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id179">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id180">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#acknowledgments">Acknowledgments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-snippy">Module <code class="docutils literal"><span class="pre">Snippy</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id181">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id182">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id183">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id184">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id185">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id186">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#qiime-version-1-9">QIIME (version 1.9)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#qiime-prep"><code class="docutils literal"><span class="pre">qiime_prep</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id188">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id189">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id190">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id191">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id192">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-demult"><code class="docutils literal"><span class="pre">qiime_demult</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id193">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id194">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id195">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id196">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-chimera"><code class="docutils literal"><span class="pre">qiime_chimera</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id197">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id198">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id199">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id200">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id201">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-pick-otus"><code class="docutils literal"><span class="pre">qiime_pick_otus</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id202">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id203">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id204">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id205">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id206">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-pick-rep-set"><code class="docutils literal"><span class="pre">qiime_pick_rep_set</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id207">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id208">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id209">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id210">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id211">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-align-seqs"><code class="docutils literal"><span class="pre">qiime_align_seqs</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id212">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id213">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id214">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id215">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id216">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-filter-alignment"><code class="docutils literal"><span class="pre">qiime_filter_alignment</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id217">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id218">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id219">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id220">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id221">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-assign-taxonomy"><code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id222">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id223">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id224">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id225">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id226">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-make-phylogeny"><code class="docutils literal"><span class="pre">qiime_make_phylogeny</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id227">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id228">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id229">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id230">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id231">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-make-otu-table"><code class="docutils literal"><span class="pre">qiime_make_otu_table</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id232">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#optional">Optional</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id233">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id234">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id235">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id236">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-filter-samples-from-otu-table"><code class="docutils literal"><span class="pre">qiime_filter_samples_from_otu_table</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id237">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id238">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id239">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id240">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id241">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-filter-otus"><code class="docutils literal"><span class="pre">qiime_filter_otus</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id242">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id243">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id244">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id245">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id246">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-sort-otu-table"><code class="docutils literal"><span class="pre">qiime_sort_otu_table</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id247">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id248">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id249">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id250">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id251">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qiime-divers"><code class="docutils literal"><span class="pre">qiime_divers</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id252">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id253">Optional</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id254">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id255">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id256">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id257">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#various-reporting-programs">Various Reporting Programs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#ngsplot"><code class="docutils literal"><span class="pre">NGSplot</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id259">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id260">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id261">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id262">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id263">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id264"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id265">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id266">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id267">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id268">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-collect-results">Module <code class="docutils literal"><span class="pre">Collect_results</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id269">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id270">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id271">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id272">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id273">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id274">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#sequence-searching-related-tasks">Sequence-Searching Related Tasks</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#makeblastdb"><code class="docutils literal"><span class="pre">makeblastdb</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id276">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id277">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id278">Parameters that can be set:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id279">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id280">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#blast"><code class="docutils literal"><span class="pre">blast</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id281">Requires:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id282">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id283">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id284">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id285">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#blast-new-alternative-implementation"><code class="docutils literal"><span class="pre">blast_new</span></code> - Alternative implementation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id286">Requires:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id287">Output:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id288">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id291">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id292">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#parse-blast"><code class="docutils literal"><span class="pre">parse_blast</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id293">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id294">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id295">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id296">Lines for parameter file</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-gassst">Module <code class="docutils literal"><span class="pre">Gassst</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id297">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id298">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id299">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id300">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id301">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id302">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#bam-conversion-to-other-formats">BAM Conversion to Other Formats</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#genomecoveragebed"><code class="docutils literal"><span class="pre">genomeCoverageBed</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id304">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id305">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id306">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id307">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id308">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ucsc-bw-wig"><code class="docutils literal"><span class="pre">UCSC_BW_wig</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id309">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id310">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id311">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id312">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id313">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#igv-count"><code class="docutils literal"><span class="pre">IGV_count</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id314">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id315">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id316">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id317">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id318">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#igv-totdf"><code class="docutils literal"><span class="pre">IGV_toTDF</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id319">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id320">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id321">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id322">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id323">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#variant-related-tasks">Variant-related Tasks</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#freebayes"><code class="docutils literal"><span class="pre">freebayes</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id325">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id326">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id327">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id328">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id329">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id330">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#mpileup-varscan"><code class="docutils literal"><span class="pre">mpileup_varscan</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id331">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id332">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id333">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id334">Comments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id335">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id336">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#vcftools"><code class="docutils literal"><span class="pre">vcftools</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id337">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id338">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id339">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id340">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id341">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#chip-seq">ChIP-seq</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#macs2-callpeak"><code class="docutils literal"><span class="pre">macs2_callpeak</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id343">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id344">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id345">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id346">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id347">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#macs2-bdgcmp"><code class="docutils literal"><span class="pre">macs2_bdgcmp</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id348">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id349">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id350">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id351">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id352">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ceas"><code class="docutils literal"><span class="pre">CEAS</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id353">Requires</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id354">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id355">Parameters that can be set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id356">Lines for parameter file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id357">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#generic-module">Generic Module</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#generic"><code class="docutils literal"><span class="pre">Generic</span></code> <sup>*</sup></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#parameters-file-arguments">Parameters file arguments:</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Workflow_docs/Assembly_Indexing_mapping.html">Assembly and Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="Workflow_docs/BLAST_db.html">BLAST on <cite>fasta</cite> files</a></li>
<li class="toctree-l1"><a class="reference internal" href="Workflow_docs/BLAST_fasta.html">Assembly and Searching</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">NeatSeq_Flow Module Repository</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Preparation and QC</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/all_modules.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="preparation-and-qc">
<h1>Preparation and QC<a class="headerlink" href="#preparation-and-qc" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="modules-included-in-this-section">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#merge" id="id361"><code class="docutils literal"><span class="pre">merge</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#fastqc-html" id="id362"><code class="docutils literal"><span class="pre">fastqc_html</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#trimmo" id="id363"><code class="docutils literal"><span class="pre">trimmo</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#multiqc" id="id364"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#module-cutadapt" id="id365">Module <code class="docutils literal"><span class="pre">Cutadapt</span></code></a></li>
<li><a class="reference internal" href="#module-trim-galore" id="id366">Module <code class="docutils literal"><span class="pre">Trim_Galore</span></code></a></li>
<li><a class="reference internal" href="#merge-project" id="id367"><code class="docutils literal"><span class="pre">merge_project</span></code></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.preparing.merge"></span><div class="section" id="merge">
<h2><a class="toc-backref" href="#id361"><code class="docutils literal"><span class="pre">merge</span></code> <sup>*</sup></a><a class="headerlink" href="#merge" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for merging &lt;and unzipping&gt; fastqc and fasta files</p>
<div class="section" id="requires">
<h3>Requires<a class="headerlink" href="#requires" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;Forward&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;Reverse&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;Single&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">or fasta files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;Nucleotide&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;Protein&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts fastq output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;|&quot;fastq.R&quot;|&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">puts fasta output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;|&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In the <em>merge</em> parameters, set the <em>script_path</em> parameter according to the type of raw files you’ve got. 
e.g., if they are gzipped, it should be <code class="docutils literal"><span class="pre">gzip</span> <span class="pre">-cd</span></code>, etc.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>If you want to do something more complex with the combined files, you can use the <code class="docutils literal"><span class="pre">pipe</span></code> parameter to send extra commands to be piped on the files after the main command.</p>
<p class="last">e.g.: You can get files from a remote location by setting <code class="docutils literal"><span class="pre">script_path</span></code> to <code class="docutils literal"><span class="pre">curl</span></code> and <code class="docutils literal"><span class="pre">pipe</span></code> to <code class="docutils literal"><span class="pre">gzip</span> <span class="pre">-cd</span></code>. This will download the files with curl, unzip them and concatenate them into the target file.  In the sample file, specify remote URLs instead of local pathes.</p>
</div>
</div>
<div class="section" id="parameters-that-can-be-set">
<h3>Parameters that can be set<a class="headerlink" href="#parameters-that-can-be-set" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>pipe</td>
<td>&#160;</td>
<td>Additional commands to be piped on the files before writing to file.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="lines-for-parameter-file">
<h3>Lines for parameter file<a class="headerlink" href="#lines-for-parameter-file" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">merge1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">merge</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="n">gzip</span> <span class="o">-</span><span class="n">cd</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">merge1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">merge</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="n">curl</span>
    <span class="n">pipe</span><span class="p">:</span>  <span class="n">gzip</span> <span class="o">-</span><span class="n">cd</span>
</pre></div>
</div>
</div>
</div>
<span class="target" id="module-main_NSF_classes.preparing.fastqc_html"></span><div class="section" id="fastqc-html">
<h2><a class="toc-backref" href="#id362"><code class="docutils literal"><span class="pre">fastqc_html</span></code> <sup>*</sup></a><a class="headerlink" href="#fastqc-html" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running fastqc.</p>
<p>Creates scripts that run fastqc on all available fastq files.</p>
<div class="section" id="id1">
<h3>Requires<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id2">
<h3>Output<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts fastqc output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastqc_fastq.F_html&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastqc_fastq.R_html&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastqc_fastq.S_html&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">puts fastqc zip files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastqc_fastq.F_zip&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastqc_fastq.R_zip&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastqc_fastq.S_zip&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id3">
<h3>Lines for parameter file<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">fqc_merge1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">fastqc_html</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">merge1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">FastQC</span><span class="o">/</span><span class="n">fastqc</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">15</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">threads</span><span class="p">:</span> <span class="mi">15</span>
</pre></div>
</div>
</div>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<p>Andrews, S., 2010. FastQC: a quality control tool for high throughput sequence data.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.preparing.trimmo"></span><div class="section" id="trimmo">
<h2><a class="toc-backref" href="#id363"><code class="docutils literal"><span class="pre">trimmo</span></code> <sup>*</sup></a><a class="headerlink" href="#trimmo" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running trimmomatic on fastq files</p>
<div class="section" id="id4">
<h3>Requires<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">fastq files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id5">
<h3>Output<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">puts fastq output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;|&quot;fastq.R&quot;|&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id6">
<h3>Parameters that can be set<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>spec_dir</td>
<td>path</td>
<td>If <em>trimmomatic</em> must be executed within a particular directory, specify that directory here</td>
</tr>
<tr class="row-odd"><td>todo</td>
<td>LEADING:20 TRAILING:20</td>
<td>The trimmomatic arguments</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id7">
<h3>Lines for parameter file<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">trim1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">trimmo</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">merge1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="n">java</span> <span class="o">-</span><span class="n">jar</span> <span class="n">trimmomatic</span><span class="o">-</span><span class="mf">0.32</span><span class="o">.</span><span class="n">jar</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
        <span class="n">node</span><span class="p">:</span> <span class="n">node1</span>
    <span class="n">spec_dir</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">Trimmomatic_dir</span><span class="o">/</span>
    <span class="n">todo</span><span class="p">:</span> <span class="n">LEADING</span><span class="p">:</span><span class="mi">20</span> <span class="n">TRAILING</span><span class="p">:</span><span class="mi">20</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">threads</span><span class="p">:</span> <span class="mi">20</span>
</pre></div>
</div>
</div>
<div class="section" id="id8">
<h3>References<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<p>Bolger, A.M., Lohse, M. and Usadel, B., 2014. <strong>Trimmomatic: a flexible trimmer for Illumina sequence data</strong>. <em>Bioinformatics</em>, 30(15), pp.2114-2120.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.Reports.Multiqc"></span><div class="section" id="multiqc">
<h2><a class="toc-backref" href="#id364"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a><a class="headerlink" href="#multiqc" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for preparing a MultiQC report for all samples</p>
<div class="section" id="id9">
<h3>Requires<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>No real requirements. Will give a report with information if one of the base steps produces reports that MultiQC can read, <em>e.g.</em> fastqc, bowtie2, samtools etc.</li>
</ul>
</div>
<div class="section" id="id10">
<h3>Output<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts report dir in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;Multiqc_report&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id11">
<h3>Lines for parameter file<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">firstMultQC</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Multiqc</span>
    <span class="n">base</span><span class="p">:</span>
    <span class="o">-</span> <span class="n">sam_bwt2_1</span>
    <span class="o">-</span> <span class="n">fqc_trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">multiqc</span>
</pre></div>
</div>
</div>
<div class="section" id="id12">
<h3>References<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<p>Ewels, P., Magnusson, M., Lundin, S. and Käller, M., 2016. <strong>MultiQC: summarize analysis results for multiple tools and samples in a single report</strong>. <em>Bioinformatics</em>, 32(19), pp.3047-3048.</p>
</div>
</div>
<span class="target" id="module-Liron.Cutadapt_module.Cutadapt"></span><div class="section" id="module-cutadapt">
<h2><a class="toc-backref" href="#id365">Module <code class="docutils literal"><span class="pre">Cutadapt</span></code></a><a class="headerlink" href="#module-cutadapt" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Levin Liron</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module for running cutadapt on fastqc files</dd>
</dl>
<div class="section" id="id13">
<h3>Requires<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>fastq files in at least one of the following slots:</dt>
<dd>sample_data[&lt;sample&gt;][“fastq.F”]
sample_data[&lt;sample&gt;][“fastq.R”]
sample_data[&lt;sample&gt;][“fastq.S”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id14">
<h3>Output<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>puts fastq output files in the following slots:</dt>
<dd>sample_data[&lt;sample&gt;][“fastq.F”]
sample_data[&lt;sample&gt;][“fastq.R”]
sample_data[&lt;sample&gt;][“fastq.S”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id15">
<h3>Parameters that can be set<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>PARAMETER NAME</td>
<td>POSSIBLE VALUES</td>
<td>DESCRIPTION</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="comments">
<h3>Comments<a class="headerlink" href="#comments" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id16">
<h3>Lines for parameter file<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Step_Name</span><span class="p">:</span>                       <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Cutadapt</span>             <span class="c1"># Name of the module used</span>
    <span class="n">base</span><span class="p">:</span>                        <span class="c1"># Name of the step [or list of names] to run after [must be after a merge step]</span>
    <span class="n">script_path</span><span class="p">:</span>                 <span class="c1"># Command for running the Cutadapt script</span>
    <span class="n">paired</span><span class="p">:</span>                      <span class="c1"># Analyse Forward and Reverse reads together.</span>
    <span class="n">Demultiplexing</span><span class="p">:</span>              <span class="c1"># Use to Demultiplex the adaptors, needs to be in the format of name=adaptor_seq</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>                     <span class="c1"># Number of CPUs to reserve for this analysis</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">too</span><span class="o">-</span><span class="n">short</span><span class="o">-</span><span class="n">output</span><span class="p">:</span>      <span class="c1"># will replace @ with the location of the sample dir  [e.g. @too_short.fq] </span>
        <span class="o">-</span><span class="n">a</span><span class="p">:</span>                      <span class="c1"># Use to trim poly A in SE reads [e.g. &quot;A{100} -A T{100}&quot;]</span>
</pre></div>
</div>
</div>
<div class="section" id="id17">
<h3>References<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<span class="target" id="module-Liron.Trim_Galore_module.Trim_Galore"></span><div class="section" id="module-trim-galore">
<h2><a class="toc-backref" href="#id366">Module <code class="docutils literal"><span class="pre">Trim_Galore</span></code></a><a class="headerlink" href="#module-trim-galore" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module for running Trim Galore on fastq files</dd>
</dl>
<div class="section" id="id18">
<h3>Requires<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>fastq files in at least one of the following slots:</dt>
<dd>sample_data[&lt;sample&gt;][“fastq.F”]
sample_data[&lt;sample&gt;][“fastq.R”]
sample_data[&lt;sample&gt;][“fastq.S”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id19">
<h3>Output<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>puts fastq output files in the following slots:</dt>
<dd>sample_data[&lt;sample&gt;][“fastq.F”]
sample_data[&lt;sample&gt;][“fastq.R”]
sample_data[&lt;sample&gt;][“fastq.S”]</dd>
<dt>puts unpaired fastq output files in the following slots:</dt>
<dd>sample_data[&lt;sample&gt;][“fastq.F.unpaired”]
sample_data[&lt;sample&gt;][“fastq.R.unpaired”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id20">
<h3>Parameters that can be set<a class="headerlink" href="#id20" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>PARAMETER NAME</td>
<td>POSSIBLE VALUES</td>
<td>DESCRIPTION</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id21">
<h3>Comments<a class="headerlink" href="#id21" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id22">
<h3>Lines for parameter file<a class="headerlink" href="#id22" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt>::</dt>
<dd><dl class="first last docutils">
<dt>Step_Name:                       # Name of this step</dt>
<dd><p class="first">module: Trim_Galore          # Name of the module used
base:                        # Name of the step [or list of names] to run after [must be after a merge step]
script_path:                 # Command for running the Trim Galore script
qsub_params:</p>
<blockquote>
<div>-pe:                     # Number of CPUs to reserve for this analysis</div></blockquote>
<p>cutadapt_path:               # Location of cutadapt executable 
redirects:</p>
<blockquote class="last">
<div>–length:                # Parameters for running Trim Galore
-q:                      # Parameters for running Trim Galore</div></blockquote>
</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="id23">
<h3>References<a class="headerlink" href="#id23" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<span class="target" id="module-miscellaneous.merge_project"></span><div class="section" id="merge-project">
<h2><a class="toc-backref" href="#id367"><code class="docutils literal"><span class="pre">merge_project</span></code></a><a class="headerlink" href="#merge-project" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for merging sample <cite>fasta</cite> files into a single project wide <cite>fasta</cite> file.</p>
<div class="section" id="id24">
<h3>Requires<a class="headerlink" href="#id24" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A <cite>fasta</cite> file in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id25">
<h3>Output<a class="headerlink" href="#id25" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id26">
<h3>Parameters that can be set<a class="headerlink" href="#id26" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils" id="id359">
<caption><span class="caption-text">Parameters that can be set:</span><a class="headerlink" href="#id359" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>type</td>
<td>nucl|prot</td>
<td>Type of fasta to merge. If ommitted, both will be merged in parallel (if they exist).</td>
</tr>
<tr class="row-odd"><td>script_path</td>
<td>cat</td>
<td>This must be <em>cat</em>. Do not change it…</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id27">
<h3>Lines for parameter file<a class="headerlink" href="#id27" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">merge_proj</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>         <span class="n">merge_project</span>
    <span class="n">base</span><span class="p">:</span>           <span class="n">merge1</span>
    <span class="n">script_path</span><span class="p">:</span>    <span class="n">cat</span>
    <span class="nb">type</span><span class="p">:</span>           <span class="n">fasta</span><span class="o">.</span><span class="n">nucl</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="mapping">
<h1>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id28">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#bowtie2-builder" id="id368"><code class="docutils literal"><span class="pre">bowtie2_builder</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#bowtie2-mapper" id="id369"><code class="docutils literal"><span class="pre">bowtie2_mapper</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#bowtie1-builder" id="id370"><code class="docutils literal"><span class="pre">bowtie1_builder</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#bowtie1-mapper" id="id371"><code class="docutils literal"><span class="pre">bowtie1_mapper</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#bwa-builder" id="id372"><code class="docutils literal"><span class="pre">bwa_builder</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#bwa-mapper" id="id373"><code class="docutils literal"><span class="pre">bwa_mapper</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#star-mapper" id="id374"><code class="docutils literal"><span class="pre">STAR_mapper</span></code></a></li>
<li><a class="reference internal" href="#star-builder" id="id375"><code class="docutils literal"><span class="pre">STAR_builder</span></code></a></li>
<li><a class="reference internal" href="#samtools" id="id376"><code class="docutils literal"><span class="pre">samtools</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#id71" id="id377"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.mapping.bowtie2_builder"></span><div class="section" id="bowtie2-builder">
<h2><a class="toc-backref" href="#id368"><code class="docutils literal"><span class="pre">bowtie2_builder</span></code> <sup>*</sup></a><a class="headerlink" href="#bowtie2-builder" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running bowtie2 index builder:</p>
<p>Builds a bowtie2 index for a fasta file stored at the project or sample level.</p>
<p>Determine which one will be used by specifying <code class="docutils literal"><span class="pre">scope</span></code> as either <code class="docutils literal"><span class="pre">project</span></code> or <code class="docutils literal"><span class="pre">sample</span></code>.</p>
<div class="section" id="id29">
<h3>Requires<a class="headerlink" href="#id29" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id30">
<h3>Output<a class="headerlink" href="#id30" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output index files in one of the following slots:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bowtie2_index&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;bowtie2_index&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts the fasta file in the following slot:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;reference&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id31">
<h3>Parameters that can be set<a class="headerlink" href="#id31" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project fasta or a sample fasta.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id32">
<h3>Lines for parameter file<a class="headerlink" href="#id32" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwt2_build</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bowtie2_builder</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trinity1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">-</span><span class="n">build</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
</pre></div>
</div>
</div>
<div class="section" id="id33">
<h3>References<a class="headerlink" href="#id33" title="Permalink to this headline">¶</a></h3>
<p>Langmead, B. and Salzberg, S.L., 2012. <strong>Fast gapped-read alignment with Bowtie 2</strong>. <em>Nature methods</em>, 9(4), pp.357-359.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.mapping.bowtie2_mapper"></span><div class="section" id="bowtie2-mapper">
<h2><a class="toc-backref" href="#id369"><code class="docutils literal"><span class="pre">bowtie2_mapper</span></code> <sup>*</sup></a><a class="headerlink" href="#bowtie2-mapper" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running bowtie2 mapper:</p>
<p>The reads stored in each sample are aligned to one of the following bowtie2 indices:</p>
<ul class="simple">
<li>An external index passed with the <code class="docutils literal"><span class="pre">-x</span></code> parameter.</li>
<li>A bowtie2 index on a project fasta files, such as an assembly from all samples. Specify with <code class="docutils literal"><span class="pre">bowtie2_mapper:scope</span>&#160; <span class="pre">project</span></code></li>
<li>A sample bowtie2 index on a sample-specific fasta file, such as from a sample-wise assembly or from the sample file. Specify with <code class="docutils literal"><span class="pre">bowtie2_mapper:scope</span>&#160; <span class="pre">sample</span></code></li>
</ul>
<p>The latter two options <strong>must</strong> come after a <code class="docutils literal"><span class="pre">bowtie2_builder</span></code> instance.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">See the documentation for the <code class="docutils literal"><span class="pre">bowtie2_builder</span></code> module.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>fastq files are never defined project-wide</p>
<p class="last">The <code class="docutils literal"><span class="pre">scope</span></code> parameter controls the origin of the index files, <em>i.e.</em> wheather the fasta file to map to is an assembly of the sample reads (scope: sample) or an assembly of all reads in the project (scope: project). The reads to be mapped are always saple reads, as a ‘fastq’ slot is not defined at the project level.</p>
</div>
<div class="section" id="id34">
<h3>Requires<a class="headerlink" href="#id34" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id35">
<h3>Output<a class="headerlink" href="#id35" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output sam files in the following slots:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;sam&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts the name of the mapper in:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;mapper&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>puts fasta of reference genome (if one is given in param file) in:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;reference&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id36">
<h3>Parameters that can be set<a class="headerlink" href="#id36" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-x</td>
<td>path to bowtie2 index</td>
<td>If not given, will look for a project bowtie2 index and then for a sample bowtie2 index</td>
</tr>
<tr class="row-odd"><td>ref_genome</td>
<td>path to genome fasta</td>
<td>If -x is NOT given, will use the equivalent internal fasta. If -x is passed, and ref_genome is NOT passed, will leave the reference slot empty</td>
</tr>
<tr class="row-even"><td>get_map_log</td>
<td>&#160;</td>
<td>Store the log produced by bowtie2 (This is bowtie2 standard output)</td>
</tr>
<tr class="row-odd"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project or sample bowtie2 index.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id37">
<h3>Lines for parameter file<a class="headerlink" href="#id37" title="Permalink to this headline">¶</a></h3>
<p><strong>For external index:</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwt2_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bowtie2_mapper</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie2</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">get_map_log</span><span class="p">:</span>
    <span class="n">ref_genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">.</span><span class="n">fna</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">p</span><span class="p">:</span> <span class="mi">20</span>
        <span class="o">-</span><span class="n">q</span><span class="p">:</span> <span class="n">null</span>
        <span class="o">-</span><span class="n">x</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie2_index</span><span class="o">/</span><span class="n">ref_genome</span>
</pre></div>
</div>
<p><strong>Using a bowtie2 index constructed from a project fasta:</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwt2_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bowtie2_mapper</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">bwt2_bld1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie2</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">get_map_log</span><span class="p">:</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">p</span><span class="p">:</span> <span class="mi">20</span>
        <span class="o">-</span><span class="n">q</span><span class="p">:</span> <span class="n">null</span>
</pre></div>
</div>
</div>
<div class="section" id="id38">
<h3>References<a class="headerlink" href="#id38" title="Permalink to this headline">¶</a></h3>
<p>Langmead, B. and Salzberg, S.L., 2012. <strong>Fast gapped-read alignment with Bowtie 2</strong>. <em>Nature methods</em>, 9(4), pp.357-359.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.mapping.bowtie1_builder"></span><div class="section" id="bowtie1-builder">
<h2><a class="toc-backref" href="#id370"><code class="docutils literal"><span class="pre">bowtie1_builder</span></code> <sup>*</sup></a><a class="headerlink" href="#bowtie1-builder" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running bowtie1 index builder:</p>
<div class="section" id="id39">
<h3>Requires<a class="headerlink" href="#id39" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id40">
<h3>output<a class="headerlink" href="#id40" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt>Puts output index files in one of the following slot:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bowtie1_index&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;bowtie1_index&quot;]</span></code></li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="id41">
<h3>Parameters that can be set<a class="headerlink" href="#id41" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>path to bowtie1 index</td>
<td>If not given, will look for a project bowtie1 index and then for a sample bowtie1 index</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id42">
<h3>Lines for parameter file<a class="headerlink" href="#id42" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwt1_bld_ind</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bowtie1_builder</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trinity1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
</pre></div>
</div>
</div>
<div class="section" id="id43">
<h3>References<a class="headerlink" href="#id43" title="Permalink to this headline">¶</a></h3>
<p>Langmead, B., Trapnell, C., Pop, M. and Salzberg, S.L., 2009. <strong>Ultrafast and memory-efficient alignment of short DNA sequences to the human genome</strong>. <em>Genome biology</em>, 10(3), p.R25.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.mapping.bowtie1_mapper"></span><div class="section" id="bowtie1-mapper">
<h2><a class="toc-backref" href="#id371"><code class="docutils literal"><span class="pre">bowtie1_mapper</span></code> <sup>*</sup></a><a class="headerlink" href="#bowtie1-mapper" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running bowtie1 mapper:</p>
<p>The reads stored in each sample are aligned to one of the following bowtie indices:</p>
<ul class="simple">
<li>An external index passed with the <code class="docutils literal"><span class="pre">ebwt</span></code> parameter.</li>
<li>A bowtie index on a project fasta files, such as an assembly from all samples. Specify with <code class="docutils literal"><span class="pre">bowtie1_mapper:scope</span>&#160; <span class="pre">project</span></code></li>
<li>A sample bowtie1 index on a sample-specific fasta file, such as from a sample-wise assembly or from the sample file. Specify with <code class="docutils literal"><span class="pre">bowtie1_mapper:scope</span>&#160; <span class="pre">sample</span></code></li>
</ul>
<p>The latter two options <strong>must</strong> come after a <code class="docutils literal"><span class="pre">bowtie1_builder</span></code> instance.</p>
<div class="section" id="id44">
<h3>Requires<a class="headerlink" href="#id44" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id45">
<h3>Output<a class="headerlink" href="#id45" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output sam files in the following slots:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;sam&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts the name of the mapper in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;mapper&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts fasta of reference genome (if one is given in param file) in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;reference&quot;]</span></code></p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id46">
<h3>Parameters that can be set<a class="headerlink" href="#id46" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ebwt</td>
<td>path to bowtie1 index</td>
<td>If not given, will look for a project bowtie1 index and then for a sample bowtie1 index</td>
</tr>
<tr class="row-odd"><td>ref_genome</td>
<td>path to genome fasta</td>
<td>If ebwt is NOT given, will use the equivalent internal fasta. If ebwt IS given, and ref_genome is NOT passed, will leave the reference slot empty.</td>
</tr>
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project or sample bowtie1 index.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id47">
<h3>Lines for parameter file<a class="headerlink" href="#id47" title="Permalink to this headline">¶</a></h3>
<p><strong>For external index:</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwt1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bowtie1_mapper</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">ebwt</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie1_index</span><span class="o">/</span><span class="n">ref_genome</span>
    <span class="n">ref_genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">.</span><span class="n">fna</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">p</span><span class="p">:</span> <span class="mi">20</span>
</pre></div>
</div>
<p><strong>For project bowtie index:</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwt1_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bowtie1_mapper</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">bwt1_bld_ind</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
</pre></div>
</div>
</div>
<div class="section" id="id48">
<h3>References<a class="headerlink" href="#id48" title="Permalink to this headline">¶</a></h3>
<p>Langmead, B., Trapnell, C., Pop, M. and Salzberg, S.L., 2009. <strong>Ultrafast and memory-efficient alignment of short DNA sequences to the human genome</strong>. <em>Genome biology</em>, 10(3), p.R25.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.mapping.bwa_builder"></span><div class="section" id="bwa-builder">
<h2><a class="toc-backref" href="#id372"><code class="docutils literal"><span class="pre">bwa_builder</span></code> <sup>*</sup></a><a class="headerlink" href="#bwa-builder" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running bwa index builder:</p>
<p>Builds a bwa index for a fasta file stored at the project or sample level.</p>
<p>Determine which one will be used by specifying <code class="docutils literal"><span class="pre">scope</span></code> as either <code class="docutils literal"><span class="pre">project</span></code> or <code class="docutils literal"><span class="pre">sample</span></code>.</p>
<div class="section" id="id49">
<h3>Requires<a class="headerlink" href="#id49" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id50">
<h3>Output<a class="headerlink" href="#id50" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output index files in one of the following slots:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bwa_index&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;bwa_index&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts the fasta file in one of the following slot:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;reference&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id51">
<h3>Parameters that can be set<a class="headerlink" href="#id51" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project fasta or a sample fasta.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id52">
<h3>Lines for parameter file<a class="headerlink" href="#id52" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwa_bld_ind</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bwa_builder</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">spades1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bwa</span> <span class="n">index</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
</pre></div>
</div>
</div>
<div class="section" id="id53">
<h3>References<a class="headerlink" href="#id53" title="Permalink to this headline">¶</a></h3>
<p>Li, H. and Durbin, R., 2009. <strong>Fast and accurate short read alignment with Burrows–Wheeler transform</strong>. <em>Bioinformatics</em>, 25(14), pp.1754-1760.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.mapping.bwa_mapper"></span><div class="section" id="bwa-mapper">
<h2><a class="toc-backref" href="#id373"><code class="docutils literal"><span class="pre">bwa_mapper</span></code> <sup>*</sup></a><a class="headerlink" href="#bwa-mapper" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running bwa mapper:</p>
<p>The reads stored in each sample are aligned to one of the following bwa indices:</p>
<ul class="simple">
<li>An external index passed with the <code class="docutils literal"><span class="pre">ref_index</span></code> parameter.</li>
<li>A bwa index on a project fasta files, such as an assembly from all samples. Specify with <code class="docutils literal"><span class="pre">bwa_mapper:scope</span>&#160; <span class="pre">project</span></code></li>
<li>A sample bwa index on a sample-specific fasta file, such as from a sample-wise assembly or from the sample fasta file. Specify with <code class="docutils literal"><span class="pre">bwa_mapper:scope</span>&#160; <span class="pre">sample</span></code></li>
</ul>
<p>The latter two options <strong>must</strong> come after a <code class="docutils literal"><span class="pre">bwa_builder</span></code> instance.</p>
<div class="section" id="id54">
<h3>Requires<a class="headerlink" href="#id54" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><dl class="first docutils">
<dt>If <code class="docutils literal"><span class="pre">mod</span></code> is one of <code class="docutils literal"><span class="pre">samse,</span> <span class="pre">sampe</span></code>, the sai files are required as well (created by a <code class="docutils literal"><span class="pre">bwa</span> <span class="pre">aln</span></code> step:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;saiF|saiR|saiS&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id55">
<h3>Output<a class="headerlink" href="#id55" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output sam files in the following slots:</dt>
<dd><ul class="first last">
<li><dl class="first docutils">
<dt>If <code class="docutils literal"><span class="pre">mod</span></code> is one of <code class="docutils literal"><span class="pre">mem,</span> <span class="pre">samse,</span> <span class="pre">sampe,</span> <span class="pre">bwasw</span></code>:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;sam&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>If <code class="docutils literal"><span class="pre">mod</span></code> is <code class="docutils literal"><span class="pre">aln</span></code>:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;saiF|saiR|saiS&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts the name of the mapper in:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;mapper&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>puts fasta of reference genome (if one is given in param file) in:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;reference&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id56">
<h3>Parameters that can be set<a class="headerlink" href="#id56" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ref_index</td>
<td>path to bwa index</td>
<td>If not given, will look for a project bwa index and then for a sample bwa index</td>
</tr>
<tr class="row-odd"><td>ref_genome</td>
<td>path to genome fasta</td>
<td>If ref_index is NOT given, will use the equivalent internal fasta. If ref_index is passed, and ref_genome is NOT passed, will leave the reference slot empty</td>
</tr>
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project or sample bwa index.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id57">
<h3>Lines for parameter file<a class="headerlink" href="#id57" title="Permalink to this headline">¶</a></h3>
<p><strong>For external index:</strong></p>
<ol class="arabic simple">
<li>Using <code class="docutils literal"><span class="pre">mem</span></code>:</li>
</ol>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwa_mem_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bwa_mapper</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bwa</span>
    <span class="n">mod</span><span class="p">:</span> <span class="n">mem</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">ref_genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">.</span><span class="n">fna</span>
    <span class="n">ref_index</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bwa_index</span><span class="o">/</span><span class="n">ref_genome</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">t</span><span class="p">:</span> <span class="mi">20</span>
</pre></div>
</div>
<p><strong>2. Using ``aln - samse/sampe``:</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwa_aln_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bwa_mapper</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bwa_mapper</span>
    <span class="n">mod</span><span class="p">:</span> <span class="n">aln</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">ref_genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">.</span><span class="n">fna</span>
    <span class="n">ref_index</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bwa_index</span><span class="o">/</span><span class="n">ref_genome</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">t</span><span class="p">:</span> <span class="mi">20</span>
<span class="n">bwa_samse_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bwa_mapper</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">bwt2_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bwa</span>
    <span class="n">mod</span><span class="p">:</span> <span class="n">samse</span>
    <span class="n">ref_genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">.</span><span class="n">fna</span>
    <span class="n">ref_index</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bwa_index</span><span class="o">/</span><span class="n">ref_genome</span>
</pre></div>
</div>
<p><strong>For project bwa index:</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bwa_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">bwa_mapper</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">bwa_bld_ind</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bwa</span>
    <span class="n">mod</span><span class="p">:</span> <span class="n">mem</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
</pre></div>
</div>
</div>
<div class="section" id="id58">
<h3>References<a class="headerlink" href="#id58" title="Permalink to this headline">¶</a></h3>
<p>Li, H. and Durbin, R., 2009. <strong>Fast and accurate short read alignment with Burrows–Wheeler transform</strong>. <em>Bioinformatics</em>, 25(14), pp.1754-1760.</p>
</div>
</div>
<span class="target" id="module-mapping.STAR_mapper"></span><div class="section" id="star-mapper">
<h2><a class="toc-backref" href="#id374"><code class="docutils literal"><span class="pre">STAR_mapper</span></code></a><a class="headerlink" href="#star-mapper" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running STAR mapper:</p>
<div class="section" id="id59">
<h3>Requires<a class="headerlink" href="#id59" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set (must come after <code class="docutils literal"><span class="pre">STAR_builder</span></code> module which populates the required slots):</p>
<blockquote>
<div><ul>
<li><p class="first">STAR index directories in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;STAR_index&quot;]</span></code>  if <code class="docutils literal"><span class="pre">scope</span></code> = “sample”</li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;STAR_index&quot;]</span></code>  if <code class="docutils literal"><span class="pre">scope</span></code> = “project”</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Reference fasta files in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;STAR_fasta&quot;]</span></code>  if <code class="docutils literal"><span class="pre">scope</span></code> = “sample”</li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;STAR_fasta&quot;]</span></code>  if <code class="docutils literal"><span class="pre">scope</span></code> = “project”</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id60">
<h3>Output<a class="headerlink" href="#id60" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output sam files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;sam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Alternatively, if <code class="docutils literal"><span class="pre">--outSAMtype</span></code> is set to <code class="docutils literal"><span class="pre">BAM</span></code>, puts output BAM files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bam_unsorted&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">--outWigType</span></code> is set, will store outputs in:</p>
<blockquote>
<div><ul>
<li><p class="first">if <code class="docutils literal"><span class="pre">--outWigType</span></code> is <code class="docutils literal"><span class="pre">wiggle</span></code></p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig2_UniqueMultiple&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig2_Unique&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig1_UniqueMultiple&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig1_Unique&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">if <code class="docutils literal"><span class="pre">--outWigType</span></code> is <code class="docutils literal"><span class="pre">bedGraph</span></code></p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bdg2_UniqueMultiple&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bdg2_Unique&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bdg1_UniqueMultiple&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bdg1_Unique&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bdg&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
<li><dl class="first docutils">
<dt>Puts the name of the mapper in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;mapper&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts fasta of reference genome (if one is given in param file) in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;reference&quot;]</span></code></p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id61">
<h3>Parameters that can be set<a class="headerlink" href="#id61" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ref_genome</td>
<td>path to genome fasta</td>
<td>If ebwt is NOT given, will use the equivalent internal fasta. If ebwt IS given, and ref_genome is NOT passed, will leave the reference slot empty.</td>
</tr>
<tr class="row-odd"><td>scope</td>
<td>project | sample</td>
<td>Not used</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id62">
<h3>Lines for parameter file<a class="headerlink" href="#id62" title="Permalink to this headline">¶</a></h3>
<p><strong>For external index:</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">STAR_map</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>             <span class="n">STAR_mapper</span>
    <span class="n">base</span><span class="p">:</span>               <span class="n">STAR_bld_ind</span>
    <span class="n">script_path</span><span class="p">:</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">STAR</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">readMapNumber</span><span class="p">:</span>    <span class="mi">1000</span>
        <span class="o">--</span><span class="n">genomeDir</span><span class="p">:</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">/</span><span class="n">STAR_index</span><span class="o">/</span>
</pre></div>
</div>
<p><strong>For project STAR index:</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">STAR_map</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>             <span class="n">STAR_mapper</span>
    <span class="n">base</span><span class="p">:</span>               <span class="n">STAR_bld_ind</span>
    <span class="n">script_path</span><span class="p">:</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">STAR</span>
    <span class="n">scope</span><span class="p">:</span>              <span class="n">project</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">readMapNumber</span><span class="p">:</span>    <span class="mi">1000</span>
</pre></div>
</div>
</div>
<div class="section" id="id63">
<h3>References<a class="headerlink" href="#id63" title="Permalink to this headline">¶</a></h3>
<p>Dobin, A., Davis, C.A., Schlesinger, F., Drenkow, J., Zaleski, C., Jha, S., Batut, P., Chaisson, M. and Gingeras, T.R., 2013. <strong>STAR: ultrafast universal RNA-seq aligner</strong>. <em>Bioinformatics</em>, 29(1), pp.15-21.</p>
</div>
</div>
<span class="target" id="module-mapping.STAR_builder"></span><div class="section" id="star-builder">
<h2><a class="toc-backref" href="#id375"><code class="docutils literal"><span class="pre">STAR_builder</span></code></a><a class="headerlink" href="#star-builder" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running STAR genome index construction:</p>
<div class="section" id="id64">
<h3>Requires<a class="headerlink" href="#id64" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id65">
<h3>Output<a class="headerlink" href="#id65" title="Permalink to this headline">¶</a></h3>
<p>Puts output index files in one of the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;STAR_index&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;STAR_index&quot;]</span></code></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id66">
<h3>Parameters that can be set<a class="headerlink" href="#id66" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Not used</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id67">
<h3>Lines for parameter file<a class="headerlink" href="#id67" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">STAR_bld_ind</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>             <span class="n">STAR_builder</span>
    <span class="n">base</span><span class="p">:</span>               <span class="n">trinity1</span>
    <span class="n">script_path</span><span class="p">:</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">STAR</span>
    <span class="n">scope</span><span class="p">:</span>              <span class="n">project</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="n">queue</span><span class="p">:</span>          <span class="n">star</span><span class="o">.</span><span class="n">q</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">genomeSAindexNbases</span><span class="p">:</span>  <span class="mi">12</span>
        <span class="o">--</span><span class="n">genomeChrBinNbits</span><span class="p">:</span>    <span class="mi">10</span>
</pre></div>
</div>
</div>
<div class="section" id="id68">
<h3>References<a class="headerlink" href="#id68" title="Permalink to this headline">¶</a></h3>
<p>Dobin, A., Davis, C.A., Schlesinger, F., Drenkow, J., Zaleski, C., Jha, S., Batut, P., Chaisson, M. and Gingeras, T.R., 2013. <strong>STAR: ultrafast universal RNA-seq aligner</strong>. <em>Bioinformatics</em>, 29(1), pp.15-21.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.mapping.samtools"></span><div class="section" id="samtools">
<h2><a class="toc-backref" href="#id376"><code class="docutils literal"><span class="pre">samtools</span></code> <sup>*</sup></a><a class="headerlink" href="#samtools" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for executing samtools on a SAM file.</p>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">The module was tested on samtools 1.3</p>
</div>
<p>The samtools programs included in the module are the following:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">view</span></code> to convert the SAM file to a BAM file</li>
<li><code class="docutils literal"><span class="pre">sort</span></code> to sort the BAM file</li>
<li><code class="docutils literal"><span class="pre">index</span></code> creates an index for the BAM file</li>
<li><code class="docutils literal"><span class="pre">flagstat</span></code> Runs flagstat on the BAM file</li>
<li><code class="docutils literal"><span class="pre">stats</span></code> Runs stats on the BAM file</li>
<li><code class="docutils literal"><span class="pre">idxstats</span></code> Runs idxstats on the BAM file</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Order of samtools subprogram execution:</p>
<p class="last">The <code class="docutils literal"><span class="pre">samtools</span></code> programs are executed in the order above. It is up to you to have a sensible combination…</p>
</div>
<p><strong>Requires</strong>:</p>
<ul>
<li><p class="first">A SAM file in the following location:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;sam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<p><strong>Output</strong>:</p>
<ul>
<li><p class="first">Depending on the parameters, will put files in the following locations:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;index&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;unfiltered_bam&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;unsorted_bam&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;flagstat&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;stats&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;idxstats&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<table border="1" class="docutils" id="id360">
<caption><span class="caption-text">Parameters that can be set:</span><a class="headerlink" href="#id360" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>view</td>
<td><em>e.g.</em>: -buh  -q 30</td>
<td><code class="docutils literal"><span class="pre">samtools</span> <span class="pre">view</span></code> parameters.</td>
</tr>
<tr class="row-odd"><td>sort</td>
<td><em>e.g.</em>: -&#64; 20</td>
<td><code class="docutils literal"><span class="pre">samtools</span> <span class="pre">sort</span></code> parameters.</td>
</tr>
<tr class="row-even"><td>index</td>
<td>&#160;</td>
<td><code class="docutils literal"><span class="pre">samtools</span> <span class="pre">index</span></code> parameters.</td>
</tr>
<tr class="row-odd"><td>flagstat</td>
<td>&#160;</td>
<td>Leave empty. flagstat takes no parameters</td>
</tr>
<tr class="row-even"><td>stats</td>
<td><code class="docutils literal"><span class="pre">samtools</span> <span class="pre">stats</span></code> parameters</td>
<td>Adds code for <code class="docutils literal"><span class="pre">samtools</span> <span class="pre">stats</span></code></td>
</tr>
<tr class="row-odd"><td>idxstats</td>
<td>&#160;</td>
<td>Adds code for <code class="docutils literal"><span class="pre">samtools</span> <span class="pre">idxstats</span></code></td>
</tr>
<tr class="row-even"><td>filter_by_tag</td>
<td><em>e.g.</em>: NM:i:[01]</td>
<td>Filter BAM by one of the tags. Use an awk-compliant regular expression. In this example, keep only lines where the edit distance is 0 or 1. This is an experimental feature and should be used with caution…</td>
</tr>
<tr class="row-odd"><td>del_sam</td>
<td>&#160;</td>
<td>Remove SAM file</td>
</tr>
<tr class="row-even"><td>del_unsorted</td>
<td>&#160;</td>
<td>Remove unsorted bam file.</td>
</tr>
</tbody>
</table>
<div class="section" id="id69">
<h3>Lines for parameter file<a class="headerlink" href="#id69" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">sam_bwt1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">samtools</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">bwt1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">samtools</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">samtools</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">view</span><span class="p">:</span> <span class="o">-</span><span class="n">buh</span>  <span class="o">-</span><span class="n">q</span> <span class="mi">30</span> <span class="o">-@</span> <span class="mi">20</span> <span class="o">-</span><span class="n">F</span> <span class="mi">4</span>
    <span class="n">sort</span><span class="p">:</span> <span class="o">-@</span> <span class="mi">20</span>
    <span class="n">flagstat</span><span class="p">:</span> <span class="n">null</span>
    <span class="n">index</span><span class="p">:</span> <span class="n">null</span>
    <span class="n">stats</span><span class="p">:</span> <span class="o">--</span><span class="n">remove</span><span class="o">-</span><span class="n">dups</span>
    <span class="n">del_sam</span><span class="p">:</span> <span class="n">null</span>
    <span class="n">del_unsorted</span><span class="p">:</span> <span class="n">null</span>
</pre></div>
</div>
</div>
<div class="section" id="id70">
<h3>References<a class="headerlink" href="#id70" title="Permalink to this headline">¶</a></h3>
<p>Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., Marth, G., Abecasis, G. and Durbin, R., 2009. <strong>The sequence alignment/map format and SAMtools</strong>. <em>Bioinformatics</em>, 25(16), pp.2078-2079.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.Reports.Multiqc"></span><div class="section" id="id71">
<h2><a class="toc-backref" href="#id377"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a><a class="headerlink" href="#id71" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for preparing a MultiQC report for all samples</p>
<div class="section" id="id72">
<h3>Requires<a class="headerlink" href="#id72" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>No real requirements. Will give a report with information if one of the base steps produces reports that MultiQC can read, <em>e.g.</em> fastqc, bowtie2, samtools etc.</li>
</ul>
</div>
<div class="section" id="id73">
<h3>Output<a class="headerlink" href="#id73" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts report dir in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;Multiqc_report&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id74">
<h3>Lines for parameter file<a class="headerlink" href="#id74" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">firstMultQC</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Multiqc</span>
    <span class="n">base</span><span class="p">:</span>
    <span class="o">-</span> <span class="n">sam_bwt2_1</span>
    <span class="o">-</span> <span class="n">fqc_trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">multiqc</span>
</pre></div>
</div>
</div>
<div class="section" id="id75">
<h3>References<a class="headerlink" href="#id75" title="Permalink to this headline">¶</a></h3>
<p>Ewels, P., Magnusson, M., Lundin, S. and Käller, M., 2016. <strong>MultiQC: summarize analysis results for multiple tools and samples in a single report</strong>. <em>Bioinformatics</em>, 32(19), pp.3047-3048.</p>
</div>
</div>
</div>
<div class="section" id="genome-assembly">
<h1>Genome Assembly<a class="headerlink" href="#genome-assembly" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id76">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#clc-assembl" id="id378"><code class="docutils literal"><span class="pre">clc_assembl</span></code></a></li>
<li><a class="reference internal" href="#megahit-assembl" id="id379"><code class="docutils literal"><span class="pre">megahit_assembl</span></code></a></li>
<li><a class="reference internal" href="#spades-assembl" id="id380"><code class="docutils literal"><span class="pre">spades_assembl</span></code> <sup>*</sup></a></li>
</ul>
</div>
<span class="target" id="module-Assembly.clc_assembl"></span><div class="section" id="clc-assembl">
<h2><a class="toc-backref" href="#id378"><code class="docutils literal"><span class="pre">clc_assembl</span></code></a><a class="headerlink" href="#clc-assembl" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for assembling reads using CLC assembler.</p>
<div class="section" id="id77">
<h3>Requires<a class="headerlink" href="#id77" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id78">
<h3>Output:<a class="headerlink" href="#id78" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts fasta output files in the following slots:</p>
<blockquote>
<div><ul>
<li><p class="first">if <code class="docutils literal"><span class="pre">scope</span></code> set to <code class="docutils literal"><span class="pre">sample</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;clc_assembl.contigs&quot;]</span></code></li>
<li>Also, sets <code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;assembler&quot;]</span> <span class="pre">=</span> <span class="pre">&quot;clc_assembl&quot;</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">if <code class="docutils literal"><span class="pre">scope</span></code> set to <code class="docutils literal"><span class="pre">project</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;clc_assembl.contigs&quot;]</span></code></li>
<li>Also, sets <code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;assembler&quot;]</span> <span class="pre">=</span> <span class="pre">&quot;clc_assembl&quot;</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id79">
<h3>Parameters that can be set<a class="headerlink" href="#id79" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample|project</td>
<td>Set to <code class="docutils literal"><span class="pre">project</span></code> to assembl all project reads into one assembly.</td>
</tr>
<tr class="row-odd"><td>p</td>
<td>e.g. ‘fb ss 180 250’</td>
<td>Sets the <code class="docutils literal"><span class="pre">-p</span></code> parameter passed to CLC for paired-end reads. Required only if the project includes paired end reads.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id80">
<h3>Lines for parameter file<a class="headerlink" href="#id80" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">clc1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">clc_assembl</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">clc_assembler</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>    <span class="n">shared</span> <span class="mi">30</span>
        <span class="n">node</span><span class="p">:</span>   <span class="n">sge37</span>
    <span class="n">scope</span><span class="p">:</span>      <span class="n">sample</span>
    <span class="n">p</span><span class="p">:</span>          <span class="n">fb</span> <span class="n">ss</span> <span class="mi">180</span> <span class="mi">250</span> 
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">cpus</span><span class="p">:</span> <span class="mi">30</span>
</pre></div>
</div>
</div>
</div>
<span class="target" id="module-Assembly.megahit_assembl"></span><div class="section" id="megahit-assembl">
<h2><a class="toc-backref" href="#id379"><code class="docutils literal"><span class="pre">megahit_assembl</span></code></a><a class="headerlink" href="#megahit-assembl" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for assembling reads using MEGAHIT assembler.</p>
<div class="section" id="id81">
<h3>Requires<a class="headerlink" href="#id81" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id82">
<h3>Output:<a class="headerlink" href="#id82" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts fasta output files in the following slots:</p>
<blockquote>
<div><ul>
<li><p class="first">if <code class="docutils literal"><span class="pre">scope</span></code> set to <code class="docutils literal"><span class="pre">sample</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;megahit_assembl.contigs&quot;]</span></code></li>
<li>Also, sets <code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;assembler&quot;]</span> <span class="pre">=</span> <span class="pre">&quot;megahit_assembl&quot;</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">if <code class="docutils literal"><span class="pre">scope</span></code> set to <code class="docutils literal"><span class="pre">project</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;megahit_assembl.contigs&quot;]</span></code></li>
<li>Also, sets <code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;assembler&quot;]</span> <span class="pre">=</span> <span class="pre">&quot;megahit_assembl&quot;</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id83">
<h3>Parameters that can be set<a class="headerlink" href="#id83" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample|project</td>
<td>Set to <code class="docutils literal"><span class="pre">project</span></code> to assembl all project reads into one assembly.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id84">
<h3>Lines for parameter file<a class="headerlink" href="#id84" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">megahit1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">megahit_assembl</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">megahit</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">30</span>
        <span class="n">node</span><span class="p">:</span> <span class="n">sge37</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="k">continue</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">cpu</span><span class="o">-</span><span class="n">threads</span><span class="p">:</span> <span class="mi">30</span>
</pre></div>
</div>
</div>
<div class="section" id="id85">
<h3>References<a class="headerlink" href="#id85" title="Permalink to this headline">¶</a></h3>
<p>Li, D., Liu, C.M., Luo, R., Sadakane, K. and Lam, T.W., 2015. <strong>MEGAHIT: an ultra-fast single-node solution for large and complex metagenomics assembly via succinct de Bruijn graph</strong>. <em>Bioinformatics</em>, 31(10), pp.1674-1676.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.Assembly.spades_assembl"></span><div class="section" id="spades-assembl">
<h2><a class="toc-backref" href="#id380"><code class="docutils literal"><span class="pre">spades_assembl</span></code> <sup>*</sup></a><a class="headerlink" href="#spades-assembl" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for assembling reads using spades assembler.</p>
<div class="section" id="id86">
<h3>Requires<a class="headerlink" href="#id86" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">fastq files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id87">
<h3>Output:<a class="headerlink" href="#id87" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">puts fasta output files in the following slots:</p>
<blockquote>
<div><ul>
<li><p class="first">for sample-wise assembly:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;spades_assembl.contigs&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;spades_assembl.scaffolds&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">for mega assembly (not defined yet):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;spades_assembl.contigs&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;spades_assembl.scaffolds&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id88">
<h3>Parameters that can be set<a class="headerlink" href="#id88" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample|project</td>
<td>Set if project-wide fasta slot should be used</td>
</tr>
<tr class="row-odd"><td>truncate_names</td>
<td>&#160;</td>
<td>truncates contig names, <em>e.g.</em> ‘&gt;NODE_82_length_18610_cov_38.4999_ID_165’ will be changed to ‘&gt;NODE_82_length_18610’</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id89">
<h3>Lines for parameter file<a class="headerlink" href="#id89" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">spades1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">spades_assembl</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">spades</span><span class="o">.</span><span class="n">py</span>
    <span class="n">truncate_names</span><span class="p">:</span> 
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">careful</span><span class="p">:</span> 
</pre></div>
</div>
</div>
<div class="section" id="id90">
<h3>References<a class="headerlink" href="#id90" title="Permalink to this headline">¶</a></h3>
<p>Bankevich, A., Nurk, S., Antipov, D., Gurevich, A.A., Dvorkin, M., Kulikov, A.S., Lesin, V.M., Nikolenko, S.I., Pham, S., Prjibelski, A.D. and Pyshkin, A.V., 2012. <strong>SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing</strong>. <em>Journal of computational biology</em>, 19(5), pp.455-477.</p>
</div>
</div>
</div>
<div class="section" id="transcriptome-assembly">
<h1>Transcriptome Assembly<a class="headerlink" href="#transcriptome-assembly" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id91">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#trinity" id="id381"><code class="docutils literal"><span class="pre">trinity</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#add-trinity-tags" id="id382"><code class="docutils literal"><span class="pre">add_trinity_tags</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#module-rsem" id="id383">Module <code class="docutils literal"><span class="pre">RSEM</span></code></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.RNA_seq.trinity"></span><div class="section" id="trinity">
<h2><a class="toc-backref" href="#id381"><code class="docutils literal"><span class="pre">trinity</span></code> <sup>*</sup></a><a class="headerlink" href="#trinity" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for RNA_seq assembly using the <a class="reference external" href="https://github.com/trinityrnaseq/trinityrnaseq/wiki">Trinity assembler</a>.</p>
<div class="section" id="id92">
<h3>Requires<a class="headerlink" href="#id92" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">fastq</span></code> files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id93">
<h3>Output:<a class="headerlink" href="#id93" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">puts <code class="docutils literal"><span class="pre">fasta</span></code> output files in the following slots:</p>
<blockquote>
<div><ul>
<li><p class="first">for sample-wise assembly:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;Trinity.contigs&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">for project-wise assembly:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;Trinity.contigs&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id94">
<h3>Parameters that can be set<a class="headerlink" href="#id94" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample|project</td>
<td>Set if project-wide fasta slot should be used</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id95">
<h3>Lines for parameter file<a class="headerlink" href="#id95" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">trinity1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>     <span class="n">trinity</span>
    <span class="n">base</span><span class="p">:</span>       <span class="n">trin_tags1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">Trinity</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="n">node</span><span class="p">:</span>      <span class="n">sge213</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>       <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">grid_conf</span><span class="p">:</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SGE_Trinity_conf</span><span class="o">.</span><span class="n">txt</span>
        <span class="o">--</span><span class="n">CPU</span><span class="p">:</span>              <span class="mi">20</span>
        <span class="o">--</span><span class="n">seqType</span><span class="p">:</span>          <span class="n">fq</span>
        <span class="o">--</span><span class="n">JM</span><span class="p">:</span>               <span class="mi">140</span><span class="n">G</span>
        <span class="o">--</span><span class="n">min_kmer_cov</span><span class="p">:</span>     <span class="mi">2</span>
        <span class="o">--</span><span class="n">full_cleanup</span><span class="p">:</span>
</pre></div>
</div>
</div>
<div class="section" id="id96">
<h3>References<a class="headerlink" href="#id96" title="Permalink to this headline">¶</a></h3>
<p>Grabherr, M.G., Haas, B.J., Yassour, M., Levin, J.Z., Thompson, D.A., Amit, I., Adiconis, X., Fan, L., Raychowdhury, R., Zeng, Q. and Chen, Z., 2011. <strong>Trinity: reconstructing a full-length transcriptome without a genome from RNA-Seq data</strong>. <em>Nature biotechnology</em>, 29(7), p.644.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.RNA_seq.add_trinity_tags"></span><div class="section" id="add-trinity-tags">
<h2><a class="toc-backref" href="#id382"><code class="docutils literal"><span class="pre">add_trinity_tags</span></code> <sup>*</sup></a><a class="headerlink" href="#add-trinity-tags" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for adding the tags required by Trinity to the ends of the read names.
See the <a class="reference external" href="https://github.com/trinityrnaseq/trinityrnaseq/wiki/Running-Trinity#strand_specific_assembly">Strand specific assembly</a> section of the Trinity manual.</p>
<p>The module uses awk, so you don’t need to pass a <code class="docutils literal"><span class="pre">script_path</span></code>. 
Since you must pass a <code class="docutils literal"><span class="pre">script_path</span></code>, just leave it blank.</p>
<div class="section" id="id97">
<h3>Requires<a class="headerlink" href="#id97" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">fastq</span></code> files in at least one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id98">
<h3>Output:<a class="headerlink" href="#id98" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><p class="first">puts <code class="docutils literal"><span class="pre">fastq</span></code> output files (with added tags) in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id99">
<h3>Lines for parameter file<a class="headerlink" href="#id99" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">trintags</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>      <span class="n">add_trinity_tags</span>
    <span class="n">base</span><span class="p">:</span>        <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="n">NOT_USED</span>
</pre></div>
</div>
</div>
</div>
<span class="target" id="module-Liron.RSEM_module.RSEM"></span><div class="section" id="module-rsem">
<h2><a class="toc-backref" href="#id383">Module <code class="docutils literal"><span class="pre">RSEM</span></code></a><a class="headerlink" href="#module-rsem" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module for running RSEM</dd>
</dl>
<div class="section" id="id100">
<h3>Requires<a class="headerlink" href="#id100" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>fastq file in </dt>
<dd>self.sample_data[sample][“fastq.F”]
self.sample_data[sample][“fastq.R”]
self.sample_data[sample][“fastq.S”]</dd>
<dt>or bam file in </dt>
<dd>self.sample_data[sample][“bam”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id101">
<h3>Output<a class="headerlink" href="#id101" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>puts output bam files (if the input is fastq) in:</dt>
<dd>self.sample_data[sample][“bam”]
self.sample_data[sample][“unsorted_bam”]</dd>
<dt>puts the location of RSEM results in:</dt>
<dd>self.sample_data[sample][“RSEM”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id102">
<h3>Parameters that can be set<a class="headerlink" href="#id102" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>mode</td>
<td>transcriptome/genome</td>
<td>Is the reference is a genome or a transcriptome?</td>
</tr>
<tr class="row-odd"><td>gff3</td>
<td>None</td>
<td>Use if the mode is genome and the annotation file is in gff3 format</td>
</tr>
<tr class="row-even"><td>del_unsorted_bam</td>
<td>None</td>
<td>Delete unsorted bam files to save space</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id103">
<h3>Comments<a class="headerlink" href="#id103" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id104">
<h3>Lines for parameter file<a class="headerlink" href="#id104" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Step_Name</span><span class="p">:</span>                                                   <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">RSEM</span>                                             <span class="c1"># Name of the module used</span>
    <span class="n">base</span><span class="p">:</span>                                                    <span class="c1"># Name of the step [or list of names] to run after [must be after a bam file generator step or merge with fastq files]</span>
    <span class="n">script_path</span><span class="p">:</span>                                             <span class="c1"># Command for running the RSEM script </span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>                                                 <span class="c1"># Number of CPUs to reserve for this analysis</span>
    <span class="n">mode</span><span class="p">:</span>                                                    <span class="c1"># transcriptome or genome</span>
    <span class="n">annotation</span><span class="p">:</span>                                              <span class="c1"># For Genome mode: the location of GTF file [the default] , for GFF3 use the gff3 flag. For Transcriptome mode: transcript-to-gene-map file.</span>
                                                             <span class="c1"># If annotation is set to Trinity the transcript-to-gene-map file will be generated using the from_Trinity_to_gene_map script</span>
    <span class="n">from_Trinity_to_gene_map_script_path</span><span class="p">:</span>                    <span class="c1"># If the mode is transcriptome and the reference was assembled using Trinity it is possible to generate the transcript-to-gene-map file automatically using this script</span>
                                                             <span class="c1"># If annotation is set to Trinity and this line is empty or missing it will try using the module&#39;s associated script</span>
    <span class="n">gff3</span><span class="p">:</span>                                                    <span class="c1"># Use if the mode is genome and the annotation file is in gff3 format</span>
    <span class="n">mapper</span><span class="p">:</span>                                                  <span class="c1"># bowtie/bowtie2/star </span>
    <span class="n">mapper_path</span><span class="p">:</span>                                             <span class="c1"># Location of mapper script</span>
    <span class="n">rsem_prepare_reference_script_path</span><span class="p">:</span>                      <span class="c1"># Location of preparing reference script</span>
    <span class="n">plot_stat</span><span class="p">:</span>                                               <span class="c1"># Generate statistical plots</span>
    <span class="n">plot_stat_script_path</span><span class="p">:</span>                                   <span class="c1"># Location of statistical plot generating script</span>
    <span class="n">reference</span><span class="p">:</span>                                               <span class="c1"># The reference genome/transcriptome location [FASTA file]</span>
    <span class="n">rsem_generate_data_matrix_script_path</span><span class="p">:</span>                   <span class="c1"># Location of the final matrix generating script</span>
                                                             <span class="c1"># If this line is empty or missing it will try using the module&#39;s associated script</span>
    <span class="n">del_unsorted_bam</span><span class="p">:</span>                                        <span class="c1"># Delete unsorted bam files to save space.</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">append</span><span class="o">-</span><span class="n">names</span><span class="p">:</span>                                      <span class="c1"># RSEM will append gene_name/transcript_name to the result files</span>
        <span class="o">--</span><span class="n">estimate</span><span class="o">-</span><span class="n">rspd</span><span class="p">:</span>                                     <span class="c1"># Enables RSEM to learn from the data how the reads are distributed across a transcript</span>
        <span class="o">-</span><span class="n">p</span><span class="p">:</span>                                                  <span class="c1"># Number of CPUs to use in this analysis</span>
        <span class="o">--</span><span class="n">bam</span><span class="p">:</span>                                               <span class="c1"># Will use bam files and not fastq</span>
        <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">bam</span><span class="o">-</span><span class="n">output</span><span class="p">:</span>
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">genome</span><span class="o">-</span><span class="n">bam</span><span class="p">:</span>                                 <span class="c1"># Alignments in genomic coordinates (only if mode is genome)</span>
</pre></div>
</div>
</div>
<div class="section" id="id105">
<h3>References<a class="headerlink" href="#id105" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>
<div class="section" id="assembly-qc">
<h1>Assembly QC<a class="headerlink" href="#assembly-qc" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id106">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#quast" id="id384"><code class="docutils literal"><span class="pre">quast</span></code> <sup>*</sup></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.Assembly.quast"></span><div class="section" id="quast">
<h2><a class="toc-backref" href="#id384"><code class="docutils literal"><span class="pre">quast</span></code> <sup>*</sup></a><a class="headerlink" href="#quast" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running quast on fasta assemblies:</p>
<p>QUAST is executed on the fasta file along the following lines:</p>
<ul class="simple">
<li>If ‘scope’ is specified, the appropriate fasta will be used. An error will occur if the fasta does not exist.</li>
<li>If ‘scope’ is not specified, if a project-wide fasta exists, it will be used. Otherwise, sample-wise fasta files will be used. If none exist, an error will occur.</li>
</ul>
<div class="section" id="id107">
<h3>Requires<a class="headerlink" href="#id107" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id108">
<h3>Output<a class="headerlink" href="#id108" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output directory in one of:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;quast&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;quast&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id109">
<h3>Parameters that can be set<a class="headerlink" href="#id109" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project or sample bowtie2 index.</td>
</tr>
<tr class="row-odd"><td>compare_mode</td>
<td>&#160;</td>
<td>If ‘scope’ is ‘sample’, specifies wether to analyse each sample separately or to create a single comparison report for all samples.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id110">
<h3>Lines for parameter file<a class="headerlink" href="#id110" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">quast1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">quast</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">spades1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">quast</span><span class="o">.</span><span class="n">py</span>
    <span class="n">compare_mode</span><span class="p">:</span> <span class="n">null</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">fast</span><span class="p">:</span> <span class="n">null</span>
</pre></div>
</div>
</div>
<div class="section" id="id111">
<h3>References<a class="headerlink" href="#id111" title="Permalink to this headline">¶</a></h3>
<p>Gurevich, A., Saveliev, V., Vyahhi, N. and Tesler, G., 2013. <strong>QUAST: quality assessment tool for genome assemblies</strong>. <em>Bioinformatics</em>, 29(8), pp.1072-1075.</p>
</div>
</div>
</div>
<div class="section" id="sequence-annotation">
<h1>Sequence Annotation<a class="headerlink" href="#sequence-annotation" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id112">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#module-prokka" id="id385">Module <code class="docutils literal"><span class="pre">Prokka</span></code></a></li>
<li><a class="reference internal" href="#prokka-old" id="id386"><code class="docutils literal"><span class="pre">prokka_old</span></code>  <sup>*</sup></a></li>
</ul>
</div>
<span class="target" id="module-Liron.Prokka_module.Prokka"></span><div class="section" id="module-prokka">
<h2><a class="toc-backref" href="#id385">Module <code class="docutils literal"><span class="pre">Prokka</span></code></a><a class="headerlink" href="#module-prokka" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>Runs Prokka on all samples</dd>
</dl>
<div class="section" id="id113">
<h3>Requires<a class="headerlink" href="#id113" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>For each Sample, a fasta.nucl file type [e.g. an assembly result] in:</dt>
<dd>sample_data[sample][“fasta.nucl”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id114">
<h3>Output<a class="headerlink" href="#id114" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>For each Sample, puts the location of the Sample’s GFF file in:</dt>
<dd>sample_data[sample][“GFF”]</dd>
<dt>For each Sample, puts the location of the Sample’s identified genes file in:    </dt>
<dd>sample_data[sample][“fasta.nucl”]</dd>
<dt>For each Sample, puts the location of the Sample’s identified genes [translated] file in:    </dt>
<dd>sample_data[sample][“fasta.prot”]</dd>
<dt>if generate_GFF_dir option exist, puts the directory location of all Samples GFFs in:</dt>
<dd>sample_data[“GFF_dir”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id115">
<h3>Parameters that can be set<a class="headerlink" href="#id115" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>generate_GFF_dir</td>
<td>&#160;</td>
<td>Create GFF directory</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id116">
<h3>Comments<a class="headerlink" href="#id116" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id117">
<h3>Lines for parameter file<a class="headerlink" href="#id117" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Step_Name</span><span class="p">:</span>                                  <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Prokka</span>                          <span class="c1"># Name of the module to use</span>
    <span class="n">base</span><span class="p">:</span>                                   <span class="c1"># Name of the step [or list of names] to run after [must be after a fasta file generator step like an assembly program or start the analysis with fasta files]</span>
    <span class="n">script_path</span><span class="p">:</span>                            <span class="c1"># Command for running Prokka </span>
    <span class="n">env</span><span class="p">:</span>                                    <span class="c1"># env parameters that needs to be in the PATH for running this module</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>                                <span class="c1"># Number of CPUs to reserve for this analysis</span>
    <span class="n">generate_GFF_dir</span><span class="p">:</span>                       <span class="c1"># Create GFF directory</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">cpus</span><span class="p">:</span>                             <span class="c1"># parameters for running Prokka</span>
        <span class="o">--</span><span class="n">force</span><span class="p">:</span>                            <span class="c1"># parameters for running Prokka</span>
        <span class="o">--</span><span class="n">genus</span><span class="p">:</span>                            <span class="c1"># parameters for running Prokka</span>
        <span class="o">--</span><span class="n">kingdom</span><span class="p">:</span>                          <span class="c1"># parameters for running Prokka</span>
        <span class="o">--</span><span class="n">proteins</span><span class="p">:</span>                         <span class="c1"># Use the location of a protein DB [FASTA] for extra annotation or use &quot;VFDB&quot; to use the module VFDB built-in virulence/resistance DB  </span>
</pre></div>
</div>
</div>
<div class="section" id="id118">
<h3>References<a class="headerlink" href="#id118" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<span class="target" id="module-Annotation.prokka_old"></span><div class="section" id="prokka-old">
<h2><a class="toc-backref" href="#id386"><code class="docutils literal"><span class="pre">prokka_old</span></code>  <sup>*</sup></a><a class="headerlink" href="#prokka-old" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running prokka:</p>
<p>Prokka is executed on the contigs stored in sample_data.</p>
<div class="section" id="id119">
<h3>Requires<a class="headerlink" href="#id119" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A nucleotide fasta file in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id120">
<h3>Output<a class="headerlink" href="#id120" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set to <code class="docutils literal"><span class="pre">sample</span></code>:</p>
<blockquote>
<div><ul>
<li><dl class="first docutils">
<dt>Puts output predicted protein sequences (faa file) in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.prot&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts output predicted protein genomic sequences (fna file) in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts the annotation file (gff) in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;gff&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Stores the prokks dir in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;prokka.dir&quot;]</span></code></p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set to <code class="docutils literal"><span class="pre">project</span></code>:</p>
<blockquote>
<div><ul>
<li><dl class="first docutils">
<dt>Puts output predicted protein sequences (faa file) in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.prot&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts output predicted protein genomic sequences (fna file) in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts the annotation file (gff) in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">sample_data[&quot;gff&quot;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Stores the prokks dir in:</dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">sample_data[&quot;prokka.dir&quot;]</span></code></p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id121">
<h3>Parameters that can be set<a class="headerlink" href="#id121" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>generate_GFF_dir</td>
<td>empty</td>
<td>Create a dir with links to the gff files for use downstream by others. Only relevant when <code class="docutils literal"><span class="pre">scope=='sample'</span></code></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id122">
<h3>Comments<a class="headerlink" href="#id122" title="Permalink to this headline">¶</a></h3>
<p>If you set values to <code class="docutils literal"><span class="pre">--locustag</span></code>, <code class="docutils literal"><span class="pre">--genus</span></code>, <code class="docutils literal"><span class="pre">--species</span></code> and <code class="docutils literal"><span class="pre">--strain</span></code>, these will hold for all the samples, and will be passed as-is to the scripts.</p>
<p>If you pass the parameters without setting their values, the values will be set to the sample names (or to the project name, when <code class="docutils literal"><span class="pre">scope</span> <span class="pre">==</span> <span class="pre">'project'</span></code>).</p>
</div>
<div class="section" id="id123">
<h3>Lines for parameter file<a class="headerlink" href="#id123" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">prokka1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">prokka_old</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">spades1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">prokka</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">generate_GFF_dir</span><span class="p">:</span> 
    <span class="n">scope</span><span class="p">:</span> <span class="n">sample</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">cpus</span><span class="p">:</span> <span class="mi">20</span>
        <span class="o">--</span><span class="n">fast</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">force</span><span class="p">:</span>
        <span class="o">--</span><span class="n">genus</span><span class="p">:</span> <span class="n">Staphylococcus</span>
        <span class="o">--</span><span class="n">metagenome</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">strain</span><span class="p">:</span> 
</pre></div>
</div>
</div>
<div class="section" id="id124">
<h3>References<a class="headerlink" href="#id124" title="Permalink to this headline">¶</a></h3>
<p>Seemann, T., 2014. <strong>Prokka: rapid prokaryotic genome annotation</strong>. <em>Bioinformatics</em>, 30(14), pp.2068-2069.</p>
</div>
</div>
</div>
<div class="section" id="sequence-clustering">
<h1>Sequence Clustering<a class="headerlink" href="#sequence-clustering" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id125">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#cd-hit" id="id387"><code class="docutils literal"><span class="pre">cd_hit</span></code></a></li>
<li><a class="reference internal" href="#vsearch-cluster" id="id388"><code class="docutils literal"><span class="pre">vsearch_cluster</span></code></a></li>
<li><a class="reference internal" href="#vsearch-derepel" id="id389"><code class="docutils literal"><span class="pre">vsearch_derepel</span></code></a></li>
</ul>
</div>
<span class="target" id="module-clustering.cd_hit"></span><div class="section" id="cd-hit">
<h2><a class="toc-backref" href="#id387"><code class="docutils literal"><span class="pre">cd_hit</span></code></a><a class="headerlink" href="#cd-hit" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for clustering with cd-hit/ch-hit-est:</p>
<p>This module runs both cd-hit and cd-hit-est. The type of sequence (<cite>nucl</cite> or <cite>prot</cite>) will be determined by the program supplied in <cite>script_path</cite>.</p>
<p>You must make sure that the required file exists: If clustering prot sequences with <code class="docutils literal"><span class="pre">cd-hit-est</span></code>, make sure there is a <code class="docutils literal"><span class="pre">fasta.prot</span></code> file, etc.</p>
<p><strong>CD-HIT: a fast program for clustering and comparing large sets of protein or nucleotide sequences</strong>, Weizhong Li &amp; Adam Godzik. <em>Bioinformatics</em>, (2006) 22:1658-1659</p>
<p><strong>CD-HIT: accelerated for clustering the next generation sequencing data</strong>, Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu &amp; Weizhong Li. <em>Bioinformatics</em>, (2012) 28:3150-3152</p>
<div class="section" id="id126">
<h3>Requires<a class="headerlink" href="#id126" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in the following slot (scope = sample):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;|&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">fasta files in the following slot (scope = project):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;|&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id127">
<h3>Output<a class="headerlink" href="#id127" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the output fasta file in the fasta slot:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;|&quot;fasta.prot&quot;]</span></code></p>
</div></blockquote>
</li>
<li><p class="first">Or</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;|&quot;fasta.prot&quot;]</span></code></p>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id128">
<h3>Parameters that can be set<a class="headerlink" href="#id128" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project or sample fasta.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id129">
<h3>Lines for parameter file<a class="headerlink" href="#id129" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">clust_proj</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">cd_hit</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">derepel_proj</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;path/to/cd-hit-est&#39;</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">40</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">T</span><span class="p">:</span> <span class="mi">40</span>
</pre></div>
</div>
</div>
<div class="section" id="id130">
<h3>References<a class="headerlink" href="#id130" title="Permalink to this headline">¶</a></h3>
<p>Fu, L., Niu, B., Zhu, Z., Wu, S. and Li, W., 2012. <strong>CD-HIT: accelerated for clustering the next-generation sequencing data.</strong> <em>Bioinformatics</em>, 28(23), pp.3150-3152.</p>
</div>
</div>
<span class="target" id="module-clustering.vsearch_cluster"></span><div class="section" id="vsearch-cluster">
<h2><a class="toc-backref" href="#id388"><code class="docutils literal"><span class="pre">vsearch_cluster</span></code></a><a class="headerlink" href="#vsearch-cluster" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running vsearch clustering:</p>
<p>The reads stored in fasta files are clustered with one of the 3 methods available: cluster_fast, cluster_size or cluster_smallmem.</p>
<p>..Note: At the moment this works on the <cite>nucl</cite> fasta only. See the web: <a class="reference external" href="https://github.com/torognes/vsearch/issues/42">https://github.com/torognes/vsearch/issues/42</a></p>
<p>Output types are defined with the <cite>outputs</cite> parameter which can be a comma separated list of the following:</p>
<blockquote>
<div>biomout,mothur_shared_out,otutabout,profile,uc</div></blockquote>
<p>Fasta output files are defined with the <cite>fasta_outputs</cite> parameter which can be a comma separated list of the following:</p>
<blockquote>
<div>centroids,consout,msaout</div></blockquote>
<p>By default, the <cite>centroids</cite> file is stored in the <cite>fasta</cite> slot. Change this by setting <cite>store_fasta</cite> to one of the types listed above, i.e. centroids,consout or msaout</p>
<div class="section" id="id131">
<h3>Requires<a class="headerlink" href="#id131" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in the following slot (scope = sample):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">fasta files in the following slot (scope = project):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id132">
<h3>Output<a class="headerlink" href="#id132" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts required output in similarly named slots, e.g.:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;vsearch.centroids&quot;]</span></code> or
<code class="docutils literal"><span class="pre">self.sample_data[&quot;vsearch.centroids&quot;]</span></code></p>
</div></blockquote>
</li>
<li><p class="first">Puts the required fasta in the fasta slot:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code> or <code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></p>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id133">
<h3>Parameters that can be set<a class="headerlink" href="#id133" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>outputs</td>
<td>biomout,mothur_shared_out,otutabout,profile,uc</td>
<td>List of outputs other than fasta type outputs (see <cite>fasta_outputs</cite></td>
</tr>
<tr class="row-odd"><td>fasta_outputs</td>
<td>centroids,consout,msaout</td>
<td>A list of fasta types to produce.</td>
</tr>
<tr class="row-even"><td>store_fasta</td>
<td>centroids|consout|msaout</td>
<td>The fasta type to store in fasta slot</td>
</tr>
<tr class="row-odd"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project or sample nucl fasta.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id134">
<h3>Lines for parameter file<a class="headerlink" href="#id134" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">clust_proj</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">vsearch_cluster</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">derepel_proj</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.vsearch_path}</span><span class="s1">/vsearch&#39;</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">40</span>
    <span class="n">fasta_outputs</span><span class="p">:</span> <span class="n">centroids</span><span class="p">,</span><span class="n">consout</span>
    <span class="n">outputs</span><span class="p">:</span> <span class="n">uc</span>
    <span class="n">store_fasta</span><span class="p">:</span> <span class="n">centroids</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="nb">type</span><span class="p">:</span> <span class="n">cluster_fast</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="nb">id</span><span class="p">:</span> <span class="mf">0.85</span>  <span class="c1"># From ipyrad defaults</span>
        <span class="o">--</span><span class="n">qmask</span><span class="p">:</span> <span class="n">dust</span>
        <span class="o">--</span><span class="n">strand</span><span class="p">:</span> <span class="n">both</span>
        <span class="o">--</span><span class="n">threads</span><span class="p">:</span> <span class="mi">40</span>
        <span class="o">--</span><span class="n">sizein</span><span class="p">:</span>
        <span class="o">--</span><span class="n">sizeout</span><span class="p">:</span>
</pre></div>
</div>
</div>
<div class="section" id="id135">
<h3>References<a class="headerlink" href="#id135" title="Permalink to this headline">¶</a></h3>
<p>Rognes, T., Flouri, T., Nichols, B., Quince, C. and Mahé, F., 2016. <strong>VSEARCH: a versatile open source tool for metagenomics</strong>. <em>PeerJ</em>, 4, p.e2584.</p>
</div>
</div>
<span class="target" id="module-clustering.vsearch_derepel"></span><div class="section" id="vsearch-derepel">
<h2><a class="toc-backref" href="#id389"><code class="docutils literal"><span class="pre">vsearch_derepel</span></code></a><a class="headerlink" href="#vsearch-derepel" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running vsearch read dereplication:</p>
<p>Performs dereplication on fastq and fasta files.</p>
<p>..Note:: Dereplication with vsearch is not defined on paired end reads.</p>
<blockquote>
<div>At the moment, this module is defined only for <code class="docutils literal"><span class="pre">fasta.nucl</span></code> or for <code class="docutils literal"><span class="pre">fastq.S</span></code>.</div></blockquote>
<div class="section" id="id136">
<h3>Requires<a class="headerlink" href="#id136" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">or fasta files the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id137">
<h3>Output<a class="headerlink" href="#id137" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output <strong>fasta</strong> file in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;vsearch_derepl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id138">
<h3>Parameters that can be set<a class="headerlink" href="#id138" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample | project</td>
<td>Which file to use for dereplication: sample-wise or project-wise files</td>
</tr>
<tr class="row-odd"><td>uc</td>
<td><ul class="first last simple">
<li></li>
</ul>
</td>
<td>Save UCLUST-like dereplication output? (see –uc in manual)</td>
</tr>
<tr class="row-even"><td>type</td>
<td>derep_fulllength | derep_prefix</td>
<td>Type of derelpication strategy. See manual</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id139">
<h3>Lines for parameter file<a class="headerlink" href="#id139" title="Permalink to this headline">¶</a></h3>
<p>For external index:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">derepel_proj</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">vsearch_derepel</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">merge_proj</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.vsearch_path}</span><span class="s1">/vsearch&#39;</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="nb">type</span><span class="p">:</span> <span class="n">derep_fulllength</span>
    <span class="n">uc</span><span class="p">:</span> 
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">sizein</span><span class="p">:</span>
        <span class="o">--</span><span class="n">sizeout</span><span class="p">:</span>
</pre></div>
</div>
</div>
<div class="section" id="id140">
<h3>References<a class="headerlink" href="#id140" title="Permalink to this headline">¶</a></h3>
<p>Rognes, T., Flouri, T., Nichols, B., Quince, C. and Mahé, F., 2016. <strong>VSEARCH: a versatile open source tool for metagenomics</strong>. <em>PeerJ</em>, 4, p.e2584.</p>
</div>
</div>
</div>
<div class="section" id="metagenomics">
<h1>Metagenomics<a class="headerlink" href="#metagenomics" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id141">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#humann2" id="id390"><code class="docutils literal"><span class="pre">HUMAnN2</span></code></a></li>
<li><a class="reference internal" href="#kaiju" id="id391"><code class="docutils literal"><span class="pre">kaiju</span></code></a></li>
<li><a class="reference internal" href="#kraken" id="id392"><code class="docutils literal"><span class="pre">kraken</span></code></a></li>
<li><a class="reference internal" href="#metaphlan2" id="id393"><code class="docutils literal"><span class="pre">metaphlan2</span></code></a></li>
</ul>
</div>
<span class="target" id="module-metagenomics.HUMAnN2"></span><div class="section" id="humann2">
<h2><a class="toc-backref" href="#id390"><code class="docutils literal"><span class="pre">HUMAnN2</span></code></a><a class="headerlink" href="#humann2" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running <code class="docutils literal"><span class="pre">HUMAnN2</span></code>:</p>
<div class="section" id="id142">
<h3>Requires<a class="headerlink" href="#id142" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files, either forward or single:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id143">
<h3>Output<a class="headerlink" href="#id143" title="Permalink to this headline">¶</a></h3>
<p>self.sample_data[sample][“HUMAnN2.genefamilies.norm”]
self.sample_data[sample][“HUMAnN2.pathabundance.norm”]</p>
<ul>
<li><p class="first">Puts the <code class="docutils literal"><span class="pre">HUMAnN2</span></code> output files in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[sample][&quot;HUMAnN2.genefamilies&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[sample][&quot;HUMAnN2.pathabundance&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[sample][&quot;HUMAnN2.pathcoverage&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If “renorm_table” is set in params:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[sample][&quot;HUMAnN2.genefamilies.norm&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[sample][&quot;HUMAnN2.pathabundance.norm&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If “join_tables” is set in params:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;HUMAnN2.genefamilies&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;HUMAnN2.pathabundance&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;HUMAnN2.pathcoverage&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If “join_tables” and “renorm_table” are set in params:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;HUMAnN2.genefamilies.norm&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;HUMAnN2.pathabundance.norm&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;HUMAnN2.pathcoverage&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id144">
<h3>Parameters that can be set<a class="headerlink" href="#id144" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>renorm_table</td>
<td>Empty or parameters to pass to <code class="docutils literal"><span class="pre">humann2_renorm_table</span></code></td>
<td>Runs <code class="docutils literal"><span class="pre">humann2_renorm_table</span></code> on HUMAnN2 outputs with the specified parameters</td>
</tr>
<tr class="row-odd"><td>join_tables</td>
<td>&#160;</td>
<td>Runs <code class="docutils literal"><span class="pre">humann2_join_tables</span></code> to gather all sample tables.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id145">
<h3>Lines for parameter file<a class="headerlink" href="#id145" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">HUMAnN2_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">HUMAnN2</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">paths</span><span class="o">.</span><span class="n">humann2</span><span class="p">}</span>
    <span class="n">join_tables</span><span class="p">:</span> 
    <span class="n">renorm_table</span><span class="p">:</span> <span class="o">--</span><span class="n">units</span> <span class="n">cpm</span> <span class="o">-</span><span class="n">p</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">bowtie2</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie2</span>
        <span class="o">--</span><span class="n">gap</span><span class="o">-</span><span class="n">fill</span><span class="p">:</span> <span class="n">on</span>
        <span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">format</span><span class="p">:</span> <span class="n">fastq</span>
        <span class="o">--</span><span class="n">metaphlan</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">paths</span><span class="o">.</span><span class="n">metaphlan2</span><span class="p">}</span>
        <span class="o">--</span><span class="n">minpath</span><span class="p">:</span> <span class="n">on</span>
        <span class="o">--</span><span class="n">nucleotide</span><span class="o">-</span><span class="n">database</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">chocophlan</span><span class="p">}</span>
        <span class="o">--</span><span class="n">protein</span><span class="o">-</span><span class="n">database</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">uniref</span><span class="p">}</span>
        <span class="o">--</span><span class="n">threads</span><span class="p">:</span> <span class="mi">30</span>
</pre></div>
</div>
</div>
<div class="section" id="id146">
<h3>References<a class="headerlink" href="#id146" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="http://huttenhower.sph.harvard.edu/humann2">HUMAnN2 home page</a></p>
</div>
</div>
<span class="target" id="module-metagenomics.kaiju"></span><div class="section" id="kaiju">
<h2><a class="toc-backref" href="#id391"><code class="docutils literal"><span class="pre">kaiju</span></code></a><a class="headerlink" href="#kaiju" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running <code class="docutils literal"><span class="pre">kaiju</span></code>:</p>
<div class="section" id="id147">
<h3>Requires<a class="headerlink" href="#id147" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files, either paired end or single:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id148">
<h3>Output<a class="headerlink" href="#id148" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the <code class="docutils literal"><span class="pre">kaiju</span></code> output files in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;raw_classification&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If  ‘kaiju2krona’ is set:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;classification&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">ktImportText_path</span></code> parameter was passed, puts the krona reports in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;krona&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id149">
<h3>Parameters that can be set<a class="headerlink" href="#id149" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ktImportText_path</td>
<td>&#160;</td>
<td>Path to ktImportText.</td>
</tr>
<tr class="row-odd"><td>kaiju2krona</td>
<td>&#160;</td>
<td>Path to kaiju2krona. If not specified, will derive it from the <code class="docutils literal"><span class="pre">script_path</span></code></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id150">
<h3>Lines for parameter file<a class="headerlink" href="#id150" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">kaiju1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">kaiju</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">paths</span><span class="o">.</span><span class="n">kaiju</span><span class="p">}</span>
    <span class="n">kaiju2krona</span><span class="p">:</span> 
    <span class="n">ktImportText_path</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">paths</span><span class="o">.</span><span class="n">ktImportText</span><span class="p">}</span>
    <span class="n">names_dmp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">kaijudb</span><span class="o">/</span><span class="n">names</span><span class="o">.</span><span class="n">dmp</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">f</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">kaijudb</span><span class="o">/</span><span class="n">kaiju_db</span><span class="o">.</span><span class="n">fmi</span>
        <span class="o">-</span><span class="n">t</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">kaijudb</span><span class="o">/</span><span class="n">nodes</span><span class="o">.</span><span class="n">dmp</span>
        <span class="o">-</span><span class="n">z</span><span class="p">:</span> <span class="mi">40</span>            
</pre></div>
</div>
</div>
<div class="section" id="id151">
<h3>References<a class="headerlink" href="#id151" title="Permalink to this headline">¶</a></h3>
<p>Menzel, P., Ng, K.L. and Krogh, A., 2016. <strong>Fast and sensitive taxonomic classification for metagenomics with Kaiju</strong>. Nature communications, 7.</p>
</div>
</div>
<span class="target" id="module-metagenomics.kraken"></span><div class="section" id="kraken">
<h2><a class="toc-backref" href="#id392"><code class="docutils literal"><span class="pre">kraken</span></code></a><a class="headerlink" href="#kraken" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running <code class="docutils literal"><span class="pre">kraken</span></code>:</p>
<p>Note that <code class="docutils literal"><span class="pre">kraken</span></code> executable must be in a folder together with <code class="docutils literal"><span class="pre">kraken-translate</span></code> and <code class="docutils literal"><span class="pre">kraken-report</span></code>. This is the default for <code class="docutils literal"><span class="pre">kraken</span></code> installation.</p>
<p>Pass the full path to the <code class="docutils literal"><span class="pre">kraken</span></code> executable in <code class="docutils literal"><span class="pre">script_path</span></code>.</p>
<p>Merging of sample kraken reports in done in two ways: with an R script shipped with the module, and with krona. See the section on Parameters that can be set.</p>
<div class="section" id="id152">
<h3>Requires<a class="headerlink" href="#id152" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files, either paired end or single:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id153">
<h3>Output<a class="headerlink" href="#id153" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the <code class="docutils literal"><span class="pre">kraken</span></code> output files in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;raw_classification&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;classification&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;classification_report&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">ktImportTaxonomy_path</span></code> parameter was passed, puts the krona reports in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;krona&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">merge_kraken_reports</span></code> was passed, puts the merged reports in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id154">
<h3>Parameters that can be set<a class="headerlink" href="#id154" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>merge_kraken_reports</td>
<td>&#160;</td>
<td>If set, will merge the reports with <code class="docutils literal"><span class="pre">merge_kraken_reports.R</span></code>, which is shipped with the module. If <code class="docutils literal"><span class="pre">Rscript</span></code> is not in the path, you will have to use <code class="docutils literal"><span class="pre">setenv</span></code> to add it to the path.</td>
</tr>
<tr class="row-odd"><td>ktImportTaxonomy_path</td>
<td>&#160;</td>
<td>Path to ktImportTaxonomy. You can additional <code class="docutils literal"><span class="pre">ktImportTaxonomy</span></code> parameters at the end of the path. If not passed, the <code class="docutils literal"><span class="pre">krona</span></code> report will not be built.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id155">
<h3>Lines for parameter file<a class="headerlink" href="#id155" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">kraken1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">kraken</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">paths</span><span class="o">.</span><span class="n">kraken</span><span class="p">}</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">merge_kraken_reports</span><span class="p">:</span>
    <span class="n">ktImportTaxonomy_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ktImportTaxonomy</span>  <span class="o">-</span><span class="n">u</span>  <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">krona</span><span class="o">.</span><span class="n">sourceforge</span><span class="o">.</span><span class="n">net</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">db</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">kraken_std_db</span>
        <span class="o">--</span><span class="n">preload</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">quick</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">threads</span><span class="p">:</span> <span class="mi">20</span>
</pre></div>
</div>
</div>
<div class="section" id="id156">
<h3>References<a class="headerlink" href="#id156" title="Permalink to this headline">¶</a></h3>
<p>Wood, D.E. and Salzberg, S.L., 2014. <strong>Kraken: ultrafast metagenomic sequence classification using exact alignments</strong>. <em>Genome biology</em>, 15(3), p.R46.</p>
</div>
</div>
<span class="target" id="module-metagenomics.metaphlan2"></span><div class="section" id="metaphlan2">
<h2><a class="toc-backref" href="#id393"><code class="docutils literal"><span class="pre">metaphlan2</span></code></a><a class="headerlink" href="#metaphlan2" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running <code class="docutils literal"><span class="pre">metaphlan2</span></code>:</p>
<div class="section" id="id157">
<h3>Requires<a class="headerlink" href="#id157" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files, either paired end or single:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id158">
<h3>Output<a class="headerlink" href="#id158" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the <code class="docutils literal"><span class="pre">metaphlan2</span></code> output files in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;raw_classification&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If</p>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">ktImportText_path</span></code> parameter was passed, puts the krona reports in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;krona&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">merge_metaphlan_tables</span></code> was passed, puts the merged reports in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;merged_metaphlan2&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If ‘–biom’ is set in <code class="docutils literal"><span class="pre">redirects</span></code>, the biom table is put in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If ‘–bowtie2out’ is set in <code class="docutils literal"><span class="pre">redirects</span></code>, the SAM file is put in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;sam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If  ‘metaphlan2krona_path’ is set:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;classification&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id159">
<h3>Parameters that can be set<a class="headerlink" href="#id159" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ktImportText_path</td>
<td>&#160;</td>
<td>Path to ktImportText.</td>
</tr>
<tr class="row-odd"><td>merge_metaphlan_tables</td>
<td>&#160;</td>
<td>Path to merge_metaphlan_tables.py. If not specified, will derive it from the location of <code class="docutils literal"><span class="pre">metaphlan2</span></code></td>
</tr>
<tr class="row-even"><td>metaphlan2krona_path</td>
<td>&#160;</td>
<td>Path to metaphlan2krona.py</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id160">
<h3>Lines for parameter file<a class="headerlink" href="#id160" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">metph1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">metaphlan2</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">paths</span><span class="o">.</span><span class="n">metaphlan2</span><span class="p">}</span>
    <span class="n">ktImportText_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ktImportText</span>
    <span class="n">merge_metaphlan_tables</span><span class="p">:</span> 
    <span class="n">metaphlan2krona_path</span><span class="p">:</span>   <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">metaphlan2krona</span><span class="o">.</span><span class="n">py</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">biom</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">bowtie2_exe</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bowtie2</span>
        <span class="o">--</span><span class="n">bowtie2db</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">database</span>
        <span class="o">--</span><span class="n">bowtie2out</span><span class="p">:</span>
        <span class="o">--</span><span class="n">input_type</span><span class="p">:</span> <span class="n">fastq</span>
        <span class="o">--</span><span class="n">mdelim</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span>
        <span class="o">--</span><span class="n">mpa_pkl</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">mpa_v20_m200</span><span class="o">.</span><span class="n">pkl</span>
</pre></div>
</div>
</div>
<div class="section" id="id161">
<h3>References<a class="headerlink" href="#id161" title="Permalink to this headline">¶</a></h3>
<p>Truong, D.T., Franzosa, E.A., Tickle, T.L., Scholz, M., Weingart, G., Pasolli, E., Tett, A., Huttenhower, C. and Segata, N., 2015. <strong>MetaPhlAn2 for enhanced metagenomic taxonomic profiling</strong>. <em>Nature methods</em>, 12(10), pp.902-903.</p>
</div>
</div>
</div>
<div class="section" id="microbiology">
<h1>Microbiology<a class="headerlink" href="#microbiology" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id162">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#card-rgi" id="id394"><code class="docutils literal"><span class="pre">CARD_RGI</span></code></a></li>
<li><a class="reference internal" href="#module-cgmlst-and-mlst-typing" id="id395">Module <code class="docutils literal"><span class="pre">cgMLST_and_MLST_typing</span></code></a></li>
<li><a class="reference internal" href="#module-roary" id="id396">Module <code class="docutils literal"><span class="pre">Roary</span></code></a></li>
<li><a class="reference internal" href="#module-snippy" id="id397">Module <code class="docutils literal"><span class="pre">Snippy</span></code></a></li>
</ul>
</div>
<span class="target" id="module-microbiology.CARD_RGI"></span><div class="section" id="card-rgi">
<h2><a class="toc-backref" href="#id394"><code class="docutils literal"><span class="pre">CARD_RGI</span></code></a><a class="headerlink" href="#card-rgi" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running CARD RGI:</p>
<p>RGI is executed on the contigs stored in a Nucleotide fasta file.</p>
<div class="section" id="id163">
<h3>Requires<a class="headerlink" href="#id163" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A nucleotide fasta file in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id164">
<h3>Output<a class="headerlink" href="#id164" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set to <code class="docutils literal"><span class="pre">sample</span></code>:</p>
<blockquote>
<div><ul>
<li><p class="first">Puts output files in:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;CARD_RGI.json&quot;]</span></code>
<code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;CARD_RGI.tsv&quot;]</span></code></p>
</div></blockquote>
</li>
<li><p class="first">Puts index of output files in:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">self.sample_data[&quot;CARD_RGI.files_index&quot;]</span></code></p>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">merge_script_path</span></code> is specified in parameters, puts the merged file in</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">self.sample_data[&quot;CARD_RGI.merged_reports&quot;]</span></code></p>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set to <code class="docutils literal"><span class="pre">project</span></code>:</p>
<blockquote>
<div><ul>
<li><p class="first">Puts output files in:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">sample_data[&quot;CARD_RGI.json&quot;]</span></code>
<code class="docutils literal"><span class="pre">sample_data[&quot;CARD_RGI.tsv&quot;]</span></code></p>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id165">
<h3>Parameters that can be set<a class="headerlink" href="#id165" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>JSON2tsv_script</td>
<td>path</td>
<td><dl class="first last docutils">
<dt>The path to the CARD script for converting the JSON output to tsv </dt>
<dd>(find ‘convertJsonToTSV.py’ in your RGI installation)</dd>
</dl>
</td>
</tr>
<tr class="row-odd"><td>merge_script_path</td>
<td>path</td>
<td><dl class="first last docutils">
<dt>Path to a script that takes an index of RGI output files (‘–ind’) and a place to put </dt>
<dd>the output (–output). This script will be executed in the wrapping up stage.
(Note, the script can take more parameters. These should be passed with the path in 
the parameter files, e.g. ‘python /path/to/script –param1 val1 –param2 val2’)
If the parameters is not passed, no action will be taken on the output files.</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id166">
<h3>Comments<a class="headerlink" href="#id166" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id167">
<h3>Lines for parameter file<a class="headerlink" href="#id167" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rgi_inst</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">CARD_RGI</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">spades1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="n">python</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">rgi</span><span class="o">.</span><span class="n">py</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">15</span>
    <span class="n">JSON2tsv_script</span><span class="p">:</span> <span class="n">python</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">convertJsonToTSV</span><span class="o">.</span><span class="n">py</span>
    <span class="n">merge_script_path</span><span class="p">:</span> <span class="n">Rscript</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">merge_reports</span><span class="o">.</span><span class="n">R</span> <span class="o">--</span><span class="n">variable</span> <span class="n">bit_score</span>
    <span class="n">orf_to_use</span><span class="p">:</span> <span class="o">-</span><span class="n">x</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">sample</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">n</span><span class="p">:</span> <span class="mi">20</span>
        <span class="o">-</span><span class="n">x</span><span class="p">:</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="id168">
<h3>References<a class="headerlink" href="#id168" title="Permalink to this headline">¶</a></h3>
<p>McArthur, A.G., Waglechner, N., Nizam, F., Yan, A., Azad, M.A., Baylay, A.J., Bhullar, K., Canova, M.J., De Pascale, G., Ejim, L. and Kalan, L., 2013. <strong>The comprehensive antibiotic resistance database</strong>. <em>Antimicrobial agents and chemotherapy</em>, 57(7), pp.3348-3357.</p>
</div>
</div>
<span class="target" id="module-Liron.cgMLST_and_MLST_typing_module.cgMLST_and_MLST_typing"></span><div class="section" id="module-cgmlst-and-mlst-typing">
<h2><a class="toc-backref" href="#id395">Module <code class="docutils literal"><span class="pre">cgMLST_and_MLST_typing</span></code></a><a class="headerlink" href="#module-cgmlst-and-mlst-typing" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module for a MLST and cgMLST Typing</dd>
</dl>
<div class="section" id="id169">
<h3>Requires<a class="headerlink" href="#id169" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>Blast results after parsing in: </dt>
<dd>self.sample_data[&lt;sample&gt;][“blast.parsed”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id170">
<h3>Output<a class="headerlink" href="#id170" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>Typing results in:</dt>
<dd>self.sample_data[&lt;sample&gt;][“Typing”]</dd>
<dt>Merge of typing results in: </dt>
<dd>self.sample_data[“Typing”]</dd>
<dt>Files for phyloviz in:</dt>
<dd>self.sample_data[“phyloviz_MetaData”]
self.sample_data[“phyloviz_Alleles”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id171">
<h3>Parameters that can be set<a class="headerlink" href="#id171" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>cut_samples_not_in_metadata</td>
<td>&#160;</td>
<td>In the final merge file consider only samples found in the Meta-Data file</td>
</tr>
<tr class="row-odd"><td>sample_cutoff</td>
<td>[0-1]</td>
<td>In the final merge file consider only samples that have at least this fraction of identified alleles</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id172">
<h3>Comments<a class="headerlink" href="#id172" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id173">
<h3>Lines for parameter file<a class="headerlink" href="#id173" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Step_Name</span><span class="p">:</span>                                   <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">cgMLST_and_MLST_typing</span>           <span class="c1"># Name of the module to use</span>
    <span class="n">base</span><span class="p">:</span>                                    <span class="c1"># Name of the step [or list of names] to run after [must be after steps that generates blast.parsed File_Types] </span>
    <span class="n">script_path</span><span class="p">:</span>                             <span class="c1"># Leave blank</span>
    <span class="n">metadata</span><span class="p">:</span>                                <span class="c1"># Path to Meta-Data file</span>
    <span class="n">metadata_samples_ID_field</span><span class="p">:</span>               <span class="c1"># Column name in the Meta-Data file of the samples ID</span>
    <span class="n">cut_samples_not_in_metadata</span><span class="p">:</span>             <span class="c1"># In the final merge file consider only samples found in the Meta-Data file</span>
    <span class="n">sample_cutoff</span><span class="p">:</span>                           <span class="c1"># In the final merge file consider only samples that have at least this fraction of identified alleles</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">scheme</span><span class="p">:</span>                            <span class="c1"># Path to the Typing scheme file [Tab delimited]</span>
        <span class="o">--</span><span class="n">num_of_genes</span><span class="p">:</span>                      <span class="c1"># Number of genes/locus in the scheme</span>
        <span class="o">--</span><span class="n">cols_to_return</span><span class="p">:</span>                    <span class="c1"># Column names in the scheme file to include in the final Meta-Data file</span>
</pre></div>
</div>
</div>
<div class="section" id="id174">
<h3>References<a class="headerlink" href="#id174" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<span class="target" id="module-Liron.Roary_module.Roary"></span><div class="section" id="module-roary">
<h2><a class="toc-backref" href="#id396">Module <code class="docutils literal"><span class="pre">Roary</span></code></a><a class="headerlink" href="#module-roary" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module for running Roary on GFF files</dd>
</dl>
<div class="section" id="id175">
<h3>Requires<a class="headerlink" href="#id175" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>for each Sample, GFF file location in:</dt>
<dd>sample_data[&lt;sample&gt;][“GFF”]</dd>
<dt>if there is a GFF directory in:</dt>
<dd>sample_data[“GFF_dir”]
No new GFF directory will be created and ONLY the GFF files in this directory will be analysed</dd>
</dl>
<p>If the search_GFF flag is on GFF files will be searched in the last base name directory</p>
</div></blockquote>
</div>
<div class="section" id="id176">
<h3>Output<a class="headerlink" href="#id176" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>puts output GFF directory location in the following slots:</dt>
<dd><blockquote class="first">
<div>sample_data[“GFF”]</div></blockquote>
<dl class="last docutils">
<dt>puts output pan_genome results directory location in the following slots:</dt>
<dd>sample_data[“pan_genome_results_dir”]</dd>
<dt>puts output pan_genome presence_absence_matrix file location in the following slots:</dt>
<dd>sample_data[“presence_absence_matrix”]</dd>
<dt>puts output pan_genome clustered_proteins file location in the following slots:</dt>
<dd>sample_data[“clustered_proteins”]</dd>
<dt>puts output GWAS directory location in the following slot:</dt>
<dd>sample_data[“GWAS_results_dir”]</dd>
<dt>puts output Biclustering directory location in the following slot:</dt>
<dd>sample_data[“Bicluster_results_dir”]</dd>
<dt>puts output Biclustering cluster file location in the following slot:</dt>
<dd>sample_data[“Bicluster_clusters”]</dd>
<dt>puts output Gecko directory location in the following slot:</dt>
<dd>sample_data[“Gecko_results_dir”]</dd>
</dl>
</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id177">
<h3>Parameters that can be set<a class="headerlink" href="#id177" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>PARAMETER NAME</td>
<td>POSSIBLE VALUES</td>
<td>DESCRIPTION</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id178">
<h3>Comments<a class="headerlink" href="#id178" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id179">
<h3>Lines for parameter file<a class="headerlink" href="#id179" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Step_Name</span><span class="p">:</span>                                   <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Roary</span>                            <span class="c1"># Name of the module used</span>
    <span class="n">base</span><span class="p">:</span>                                    <span class="c1"># Name of the step [or list of names] to run after [must be after a GFF file generator step like Prokka]</span>
    <span class="n">script_path</span><span class="p">:</span>                             <span class="c1"># Command for running the Roary script </span>
    <span class="n">env</span><span class="p">:</span>                                     <span class="c1"># env parameters that needs to be in the PATH for running this module</span>
    <span class="n">qsub_params</span><span class="p">:</span>                             
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>                                 <span class="c1"># Number of CPUs to reserve for this analysis</span>
    <span class="n">virulence_resistance_tag</span><span class="p">:</span>                <span class="c1"># Use the name of the db used in prokka or use &quot;VFDB&quot; if you used the VFDB built-in Prokka module DB </span>
    <span class="n">search_GFF</span><span class="p">:</span>                              <span class="c1"># Search for GFF files?</span>
    <span class="n">Bi</span><span class="o">-</span><span class="n">cluster</span><span class="p">:</span>                              <span class="c1"># Do Bi-cluster analysis using the Roary results, if empty or this line dose not exist will not do Bi-cluster analysis </span>
        <span class="o">--</span><span class="n">Annotation</span><span class="p">:</span>                        <span class="c1"># location of virulence annotation file to use to annotate the clusters or use &quot;VFDB&quot; if you used the VFDB built-in Prokka module DB</span>
        <span class="o">--</span><span class="n">ID_field</span><span class="p">:</span>                          <span class="c1"># The column name in the MetaData file of the samples IDs</span>
        <span class="o">--</span><span class="n">cols_to_use</span><span class="p">:</span>                       <span class="c1"># list of the MetaData columns to use to annotate the clusters  example: &#39;&quot;ST&quot;,&quot;CC&quot;,&quot;source&quot;,&quot;host&quot;,&quot;geographic.location&quot;,&quot;Date&quot;&#39;</span>
        <span class="o">--</span><span class="n">metadata</span><span class="p">:</span>                          <span class="c1"># location of MetaData file to use to annotate the clusters</span>
    <span class="n">plot</span><span class="p">:</span>                                    <span class="c1"># plot gene presence/absence matrix</span>
        <span class="nb">format</span><span class="p">:</span>                              <span class="c1"># The gene presence/absence matrix plot output format. example: pdf</span>
        <span class="n">Clustering_method</span>                    <span class="c1"># The gene presence/absence matrix plot clustering method. example: ward</span>
    <span class="n">scoary</span><span class="p">:</span>
        <span class="n">script_path</span><span class="p">:</span>                         <span class="c1"># Command for running the scoary script, if empty or this line dose not exist will not run scoary </span>
        <span class="n">BH_cutoff</span><span class="p">:</span>                           <span class="c1"># Scoary BH correction for multiple testing cut-off</span>
        <span class="n">Bonferroni_cutoff</span><span class="p">:</span>                   <span class="c1"># Scoary Bonferroni correction for multiple testing cut-off</span>
        <span class="n">metadata_file</span><span class="p">:</span>                       <span class="c1"># location of MetaData file to use to create the scoary traits file</span>
        <span class="n">metadata_samples_ID_field</span><span class="p">:</span>           <span class="c1"># The column name in the MetaData file of the sample&#39;s IDs</span>
        <span class="n">traits_file</span><span class="p">:</span>                         <span class="c1"># Path to a traits file</span>
        <span class="n">scoary_traits_to_pars</span><span class="p">:</span>               <span class="c1"># If a traits file is not provided use a list of conditions to create the scoary traits file from MetaData file. example:&quot;source/==&#39;blood&#39;&quot;  &quot;source/==&#39;wound&#39;&quot;</span>
                                             <span class="c1"># Pairs of field and operator + value to convert to boolean traits: field_name1/op_value1 .. field_nameN/op_valueN Example: &quot;field_1/&gt;=val_1&lt;val_2&quot;    &quot;feild_2/==&#39;str_val&#39;&quot;</span>
                                             <span class="c1"># A Filter can be used by FILTER_field_name1/FILTER_op_value1&amp;field_name1/op_value1</span>
                                             <span class="c1"># Note that Gecko can&#39;t run if the Bi-clustering was not run</span>
    <span class="n">Gecko</span><span class="p">:</span>
        <span class="n">script_path</span><span class="p">:</span>                         <span class="c1"># Command for running the Gecko script, if empty or this line dose not exist will not run Gecko</span>
        <span class="o">-</span><span class="n">d</span><span class="p">:</span>                                  <span class="c1"># Parameters for running Gecko</span>
        <span class="o">-</span><span class="n">s</span><span class="p">:</span>                                  <span class="c1"># Parameters for running Gecko</span>
        <span class="o">-</span><span class="n">q</span><span class="p">:</span>                                  <span class="c1"># Parameters for running Gecko</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">k</span><span class="p">:</span>                                  <span class="c1"># Parameters for running Roary</span>
        <span class="o">-</span><span class="n">p</span><span class="p">:</span>                                  <span class="c1"># Parameters for running Roary</span>
        <span class="o">-</span><span class="n">qc</span><span class="p">:</span>                                 <span class="c1"># Parameters for running Roary</span>
        <span class="o">-</span><span class="n">s</span><span class="p">:</span>                                  <span class="c1"># Parameters for running Roary</span>
        <span class="o">-</span><span class="n">v</span><span class="p">:</span>                                  <span class="c1"># Parameters for running Roary</span>
        <span class="o">-</span><span class="n">y</span><span class="p">:</span>                                  <span class="c1"># Parameters for running Roary</span>
</pre></div>
</div>
</div>
<div class="section" id="id180">
<h3>References<a class="headerlink" href="#id180" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="acknowledgments">
<h3>Acknowledgments<a class="headerlink" href="#acknowledgments" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<span class="target" id="module-Liron.Snippy_module.Snippy"></span><div class="section" id="module-snippy">
<h2><a class="toc-backref" href="#id397">Module <code class="docutils literal"><span class="pre">Snippy</span></code></a><a class="headerlink" href="#module-snippy" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module for running Snippy on fastq files</dd>
</dl>
<div class="section" id="id181">
<h3>Requires<a class="headerlink" href="#id181" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>fastq files in at least one of the following slots:</dt>
<dd>sample_data[&lt;sample&gt;][“fastq.F”]
sample_data[&lt;sample&gt;][“fastq.R”]
sample_data[&lt;sample&gt;][“fastq.S”]</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id182">
<h3>Output<a class="headerlink" href="#id182" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><dl class="docutils">
<dt>puts Results directory location in:</dt>
<dd>self.sample_data[&lt;sample&gt;][“Snippy”]</dd>
<dt>puts for each sample the vcf file location in:</dt>
<dd>self.sample_data[&lt;sample&gt;][“vcf”]</dd>
<dt>if snippy-core is set to run:</dt>
<dd><dl class="first last docutils">
<dt>puts the core Multi-FASTA alignment location in:</dt>
<dd>self.sample_data[“fasta.nucl”]</dd>
<dt>puts core vcf file location of all analyzed samples in the following slot:</dt>
<dd>self.sample_data[“vcf”]</dd>
</dl>
</dd>
<dt>if Gubbins is set to run: </dt>
<dd><dl class="first last docutils">
<dt>puts result Tree file location of all analyzed samples in:</dt>
<dd>self.sample_data[“newick”]</dd>
<dt>update the core Multi-FASTA alignment in:</dt>
<dd>self.sample_data[“fasta.nucl”]</dd>
<dt>and update the core vcf file in the slot:</dt>
<dd>self.sample_data[“vcf”]</dd>
</dl>
</dd>
<dt>if pars is set to run, puts phyloviz ready to use files in:</dt>
<dd><dl class="first last docutils">
<dt>Alleles:</dt>
<dd>self.sample_data[“phyloviz_Alleles”]</dd>
<dt>MetaData:</dt>
<dd>self.sample_data[“phyloviz_MetaData”]</dd>
</dl>
</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="id183">
<h3>Parameters that can be set<a class="headerlink" href="#id183" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>PARAMETER NAME</td>
<td>POSSIBLE VALUES</td>
<td>DESCRIPTION</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id184">
<h3>Comments<a class="headerlink" href="#id184" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id185">
<h3>Lines for parameter file<a class="headerlink" href="#id185" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Step_Name</span><span class="p">:</span>                                  <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Snippy</span>                          <span class="c1"># Name of the module used</span>
    <span class="n">base</span><span class="p">:</span>                                   <span class="c1"># Name of the step [or list of names] to run after [must be after a merge step]</span>
    <span class="n">script_path</span><span class="p">:</span>                            <span class="c1"># Command for running the Snippy script</span>
    <span class="n">env</span><span class="p">:</span>                                    <span class="c1"># env parameters that needs to be in the PATH for running this module</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>                                <span class="c1"># Number of CPUs to reserve for this analysis</span>
    <span class="n">gubbins</span><span class="p">:</span>
        <span class="n">script_path</span><span class="p">:</span>                        <span class="c1"># Command for running the gubbins script, if empty or this line dose not exist will not run gubbins</span>
        <span class="o">--</span><span class="n">STR</span><span class="p">:</span>                              <span class="c1"># More redirects arguments for running gubbins</span>
    <span class="n">pars</span><span class="p">:</span>                                   <span class="c1"># Generate phyloviz ready to use files</span>
        <span class="o">-</span><span class="n">M</span><span class="p">:</span>                                 <span class="c1"># Location of a MetaData file </span>
        <span class="o">--</span><span class="n">Cut</span><span class="p">:</span>                              <span class="c1"># Use only Samples found in the metadata file</span>
        <span class="o">--</span><span class="n">S_MetaData</span><span class="p">:</span>                       <span class="c1"># The name of the samples ID column</span>
        <span class="o">-</span><span class="n">C</span><span class="p">:</span>                                 <span class="c1"># Use only Samples that has at least this fraction of identified alleles</span>
    <span class="n">snippy</span><span class="o">-</span><span class="n">core</span><span class="p">:</span>
        <span class="n">script_path</span><span class="p">:</span>                        <span class="c1"># Command for running the snippy-core script, if empty or this line dose not exist will not run snippy-core</span>
        <span class="o">--</span><span class="n">noref</span><span class="p">:</span>                            <span class="c1"># Exclude reference </span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">cpus</span><span class="p">:</span>                             <span class="c1"># Parameters for running Snippy</span>
        <span class="o">--</span><span class="n">force</span><span class="p">:</span>                            <span class="c1"># Force overwrite of existing output folder (default OFF)</span>
        <span class="o">--</span><span class="n">mapqual</span><span class="p">:</span>                          <span class="c1"># Minimum mapping quality to allow</span>
        <span class="o">--</span><span class="n">mincov</span><span class="p">:</span>                           <span class="c1"># Minimum coverage of variant site</span>
        <span class="o">--</span><span class="n">minfrac</span><span class="p">:</span>                          <span class="c1"># Minumum proportion for variant evidence</span>
        <span class="o">--</span><span class="n">reference</span><span class="p">:</span>                        <span class="c1"># Reference Genome location</span>
        <span class="o">--</span><span class="n">cleanup</span>                           <span class="c1"># Remove all non-SNP files: BAMs, indices etc (default OFF)            </span>
</pre></div>
</div>
</div>
<div class="section" id="id186">
<h3>References<a class="headerlink" href="#id186" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>
<div class="section" id="qiime-version-1-9">
<h1>QIIME (version 1.9)<a class="headerlink" href="#qiime-version-1-9" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id187">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#qiime-prep" id="id398"><code class="docutils literal"><span class="pre">qiime_prep</span></code></a></li>
<li><a class="reference internal" href="#qiime-demult" id="id399"><code class="docutils literal"><span class="pre">qiime_demult</span></code></a></li>
<li><a class="reference internal" href="#qiime-chimera" id="id400"><code class="docutils literal"><span class="pre">qiime_chimera</span></code></a></li>
<li><a class="reference internal" href="#qiime-pick-otus" id="id401"><code class="docutils literal"><span class="pre">qiime_pick_otus</span></code></a></li>
<li><a class="reference internal" href="#qiime-pick-rep-set" id="id402"><code class="docutils literal"><span class="pre">qiime_pick_rep_set</span></code></a></li>
<li><a class="reference internal" href="#qiime-align-seqs" id="id403"><code class="docutils literal"><span class="pre">qiime_align_seqs</span></code></a></li>
<li><a class="reference internal" href="#qiime-filter-alignment" id="id404"><code class="docutils literal"><span class="pre">qiime_filter_alignment</span></code></a></li>
<li><a class="reference internal" href="#qiime-assign-taxonomy" id="id405"><code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code></a></li>
<li><a class="reference internal" href="#qiime-make-phylogeny" id="id406"><code class="docutils literal"><span class="pre">qiime_make_phylogeny</span></code></a></li>
<li><a class="reference internal" href="#qiime-make-otu-table" id="id407"><code class="docutils literal"><span class="pre">qiime_make_otu_table</span></code></a></li>
<li><a class="reference internal" href="#qiime-filter-samples-from-otu-table" id="id408"><code class="docutils literal"><span class="pre">qiime_filter_samples_from_otu_table</span></code></a></li>
<li><a class="reference internal" href="#qiime-filter-otus" id="id409"><code class="docutils literal"><span class="pre">qiime_filter_otus</span></code></a></li>
<li><a class="reference internal" href="#qiime-sort-otu-table" id="id410"><code class="docutils literal"><span class="pre">qiime_sort_otu_table</span></code></a></li>
<li><a class="reference internal" href="#qiime-divers" id="id411"><code class="docutils literal"><span class="pre">qiime_divers</span></code></a></li>
</ul>
</div>
<span class="target" id="module-QIIME.qiime_prep"></span><div class="section" id="qiime-prep">
<h2><a class="toc-backref" href="#id398"><code class="docutils literal"><span class="pre">qiime_prep</span></code></a><a class="headerlink" href="#qiime-prep" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for preparing fastq reads for analysis with QIIME (1.9):</p>
<p>The reads stored in each sample are optinally joined and then set it a directory in such a way the downstream, QIIME’s demult can concatenate the sequences while saving the sample of origin.</p>
<p>The directory will contain symbolic links to the files to be used by demult in the following step.</p>
<div class="section" id="id188">
<h3>Requires<a class="headerlink" href="#id188" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in one of the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fastq.S&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id189">
<h3>Output<a class="headerlink" href="#id189" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts directory of links to files to use with QIIME:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;qiime.prep_links_dir&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If join is performed:</p>
<blockquote>
<div><ul>
<li><p class="first">puts the new joined reads in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fastq.J&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">puts the unjoined forward reads in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fastq.F&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">puts the unjoined reverse reads in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fastq.R&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id190">
<h3>Parameters that can be set<a class="headerlink" href="#id190" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>join</td>
<td>none, join (or join_cat - not implemented)</td>
<td>Wheather to join paired reads.</td>
</tr>
<tr class="row-odd"><td>unjoined</td>
<td>forward, reverse, both or none</td>
<td>What to do with unjoined sequences? Use only forward, only reverse, both or none. If join is none, use this parameter to indicate which reads to take for analysis.</td>
</tr>
<tr class="row-even"><td>join_algo</td>
<td>forward, reverse, both or none</td>
<td>What to do with unjoined sequences?</td>
</tr>
<tr class="row-odd"><td>parameters</td>
<td>&#160;</td>
<td>Path to QIIME parameter file to be used downstream</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id191">
<h3>Lines for parameter file<a class="headerlink" href="#id191" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_prep_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_prep</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">merge1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">join_paired_ends</span><span class="o">.</span><span class="n">py</span>
    <span class="n">join</span><span class="p">:</span> <span class="n">join</span>
    <span class="n">unjoined</span><span class="p">:</span> <span class="n">forward</span>
    <span class="n">parameters</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">qiime_params</span><span class="o">.</span><span class="n">txt</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">pe_join_method</span><span class="p">:</span> <span class="n">fastq</span><span class="o">-</span><span class="n">join</span>
</pre></div>
</div>
</div>
<div class="section" id="id192">
<h3>References<a class="headerlink" href="#id192" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_demult"></span><div class="section" id="qiime-demult">
<h2><a class="toc-backref" href="#id399"><code class="docutils literal"><span class="pre">qiime_demult</span></code></a><a class="headerlink" href="#qiime-demult" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s multiple_split_libraries_fastq.py:</p>
<p>The reads from step qiime_prep are combined into one seqs.fna file.</p>
<p>..Note:: The module has not been tested on other types of data, such as undemultiplexed reads. It should work but there will probably be unexpected problems.</p>
<div class="section" id="id193">
<h3>Requires<a class="headerlink" href="#id193" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A directory of read files with smaple names coded in the file names, such as the directory produced by qiime_prep:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;qiime.prep_links_dir&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id194">
<h3>Output<a class="headerlink" href="#id194" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting <code class="docutils literal"><span class="pre">seqs.fna</span></code> file in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;qiime.demult_seqs&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;qiime.fasta&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id195">
<h3>Lines for parameter file<a class="headerlink" href="#id195" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_demult_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_demult</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_prep_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;/path/to/multiple_split_libraries_fastq.py&#39;</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">demultiplexing_method</span><span class="p">:</span> <span class="n">sampleid_by_file</span>
        <span class="o">--</span><span class="n">include_input_dir_path</span><span class="p">:</span> <span class="n">null</span>
        <span class="o">--</span><span class="n">parameter_fp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">qiime_params</span>
        <span class="o">--</span><span class="n">remove_filepath_in_name</span><span class="p">:</span> <span class="n">null</span>
</pre></div>
</div>
</div>
<div class="section" id="id196">
<h3>References<a class="headerlink" href="#id196" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_chimera"></span><div class="section" id="qiime-chimera">
<h2><a class="toc-backref" href="#id400"><code class="docutils literal"><span class="pre">qiime_chimera</span></code></a><a class="headerlink" href="#qiime-chimera" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s identify_chimeric_seqs.py:</p>
<p>The module can operate on the raw <code class="docutils literal"><span class="pre">seqs.fna</span></code> or on an aligned version. The latter is used for ChimeraSlayer and the former for usearch61</p>
<div class="section" id="id197">
<h3>Requires<a class="headerlink" href="#id197" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;qiime.fasta&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Alternatively, an aligned fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.aligned&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id198">
<h3>Output<a class="headerlink" href="#id198" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting list of chimeras in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;chimeras&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the filtered fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.chimera_removed&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>..Note:: When using <code class="docutils literal"><span class="pre">parallel_identify_chimeric_seqs.py</span></code>, the module tries to build the scripts appropriately. <strong>It is wise to check the parallel scripts before running them…</strong></p>
</div>
<div class="section" id="id199">
<h3>Parameters that can be set<a class="headerlink" href="#id199" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>method</td>
<td>usearch61 or ChimeraSlayer</td>
<td>Method to use for the analysis (passed to the –chimera_detection_method of <code class="docutils literal"><span class="pre">identify_chimeric_seqs.py</span></code></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id200">
<h3>Lines for parameter file<a class="headerlink" href="#id200" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_chimera_usrch</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_chimera</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_demult_1</span>
    <span class="c1"># script_path: &#39;{Vars.qiime_path}/parallel_identify_chimeric_seqs.py&#39;</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/identify_chimeric_seqs.py&#39;</span>
    <span class="n">method</span><span class="p">:</span>         <span class="n">usearch61</span> <span class="c1"># Or ChimeraSlayer. Will guess depending on existing files.</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="c1"># --jobs_to_start:              20</span>
        <span class="o">--</span><span class="n">aligned_reference_seqs_fp</span><span class="p">:</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">reference_files</span><span class="o">.</span><span class="n">otus_aligned</span>
        <span class="o">--</span><span class="n">reference_seqs_fp</span><span class="p">:</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">reference_files</span><span class="o">.</span><span class="n">otus</span>
</pre></div>
</div>
</div>
<div class="section" id="id201">
<h3>References<a class="headerlink" href="#id201" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_pick_otus"></span><div class="section" id="qiime-pick-otus">
<h2><a class="toc-backref" href="#id401"><code class="docutils literal"><span class="pre">qiime_pick_otus</span></code></a><a class="headerlink" href="#qiime-pick-otus" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">pick_otus.py</span></code></p>
<div class="section" id="id202">
<h3>Requires<a class="headerlink" href="#id202" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id203">
<h3>Output<a class="headerlink" href="#id203" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting OTU table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;otu_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id204">
<h3>Parameters that can be set<a class="headerlink" href="#id204" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id205">
<h3>Lines for parameter file<a class="headerlink" href="#id205" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_pick_otu_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_pick_otus</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_chimera_usrch</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/pick_otus.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id206">
<h3>References<a class="headerlink" href="#id206" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_pick_rep_set"></span><div class="section" id="qiime-pick-rep-set">
<h2><a class="toc-backref" href="#id402"><code class="docutils literal"><span class="pre">qiime_pick_rep_set</span></code></a><a class="headerlink" href="#qiime-pick-rep-set" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">pick_rep_set.py</span></code></p>
<div class="section" id="id207">
<h3>Requires<a class="headerlink" href="#id207" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">An OTU table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;otu_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id208">
<h3>Output<a class="headerlink" href="#id208" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Saves the original fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;qiime.full_fasta&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id209">
<h3>Parameters that can be set<a class="headerlink" href="#id209" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id210">
<h3>Lines for parameter file<a class="headerlink" href="#id210" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_rep_set_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_pick_rep_set</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_pick_otu_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/pick_rep_set.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id211">
<h3>References<a class="headerlink" href="#id211" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_align_seqs"></span><div class="section" id="qiime-align-seqs">
<h2><a class="toc-backref" href="#id403"><code class="docutils literal"><span class="pre">qiime_align_seqs</span></code></a><a class="headerlink" href="#qiime-align-seqs" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running <code class="docutils literal"><span class="pre">QIIME's</span> <span class="pre">align_seqs.py</span></code>:</p>
<p>Can be used for the parallel versions thereof: <code class="docutils literal"><span class="pre">parallel_align_seqs_pynast.py</span></code></p>
<div class="section" id="id212">
<h3>Requires<a class="headerlink" href="#id212" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id213">
<h3>Output<a class="headerlink" href="#id213" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting aligned fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.aligned&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Stores the old, unaligned version in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.unaligned&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>..Note:: When using <code class="docutils literal"><span class="pre">parallel_align_seqs_pynast.py</span></code>, the module tries to build the scripts appropriately. <strong>It is wise to check the parallel scripts before running them…</strong></p>
</div>
<div class="section" id="id214">
<h3>Parameters that can be set<a class="headerlink" href="#id214" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id215">
<h3>Lines for parameter file<a class="headerlink" href="#id215" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_align_para</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_align_seqs</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_rep_set_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/parallel_align_seqs_pynast.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">jobs_to_start</span><span class="p">:</span> <span class="mi">5</span>
        <span class="o">--</span><span class="n">retain_temp_files</span><span class="p">:</span> 
</pre></div>
</div>
</div>
<div class="section" id="id216">
<h3>References<a class="headerlink" href="#id216" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_filter_alignment"></span><div class="section" id="qiime-filter-alignment">
<h2><a class="toc-backref" href="#id404"><code class="docutils literal"><span class="pre">qiime_filter_alignment</span></code></a><a class="headerlink" href="#qiime-filter-alignment" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">filter_alignment.py</span></code></p>
<div class="section" id="id217">
<h3>Requires<a class="headerlink" href="#id217" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id218">
<h3>Output<a class="headerlink" href="#id218" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting aligned fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Saves the original unaligned fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;fasta.aligned_unfiltered&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id219">
<h3>Parameters that can be set<a class="headerlink" href="#id219" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id220">
<h3>Lines for parameter file<a class="headerlink" href="#id220" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_filt_align_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_filter_alignment</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_align_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/filter_alignment.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id221">
<h3>References<a class="headerlink" href="#id221" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_assign_taxonomy"></span><div class="section" id="qiime-assign-taxonomy">
<h2><a class="toc-backref" href="#id405"><code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code></a><a class="headerlink" href="#qiime-assign-taxonomy" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">assign_taxonomy.py</span></code></p>
<p>Can also be used to run the parallel versions of the program:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">parallel_assign_taxonomy_blast.py</span></code></li>
<li><code class="docutils literal"><span class="pre">parallel_assign_taxonomy_rdp.py</span></code></li>
<li><code class="docutils literal"><span class="pre">parallel_assign_taxonomy_uclust.py</span></code></li>
</ul>
</div></blockquote>
<div class="section" id="id222">
<h3>Requires<a class="headerlink" href="#id222" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id223">
<h3>Output<a class="headerlink" href="#id223" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting list of chimeras in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;taxonomy&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>..Note:: When using the parallel version, the module tries to build the scripts appropriately. <strong>It is wise to check the parallel scripts before running them…</strong></p>
</div>
<div class="section" id="id224">
<h3>Parameters that can be set<a class="headerlink" href="#id224" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id225">
<h3>Lines for parameter file<a class="headerlink" href="#id225" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_tax_asn_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_assign_taxonomy</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_rep_set_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/parallel_assign_taxonomy_rdp.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">confidence</span><span class="p">:</span> <span class="mf">0.5</span>
        <span class="o">--</span><span class="n">id_to_taxonomy_fp</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">reference_files</span><span class="o">.</span><span class="n">id_to_taxonomy</span><span class="p">}</span>
        <span class="o">--</span><span class="n">jobs_to_start</span><span class="p">:</span> <span class="mi">20</span>
        <span class="o">--</span><span class="n">rdp_max_memory</span><span class="p">:</span> <span class="mi">50000</span>
        <span class="o">--</span><span class="n">reference_seqs_fp</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">reference_files</span><span class="o">.</span><span class="n">otus</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id226">
<h3>References<a class="headerlink" href="#id226" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_make_phylogeny"></span><div class="section" id="qiime-make-phylogeny">
<h2><a class="toc-backref" href="#id406"><code class="docutils literal"><span class="pre">qiime_make_phylogeny</span></code></a><a class="headerlink" href="#qiime-make-phylogeny" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">make_phylogeny.py</span></code></p>
<div class="section" id="id227">
<h3>Requires<a class="headerlink" href="#id227" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A fasta file in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id228">
<h3>Output<a class="headerlink" href="#id228" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting OTU table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;phylotree&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id229">
<h3>Parameters that can be set<a class="headerlink" href="#id229" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id230">
<h3>Lines for parameter file<a class="headerlink" href="#id230" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_phylo_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_make_phylogeny</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_filt_align_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/make_phylogeny.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="id231">
<h3>References<a class="headerlink" href="#id231" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_make_otu_table"></span><div class="section" id="qiime-make-otu-table">
<h2><a class="toc-backref" href="#id407"><code class="docutils literal"><span class="pre">qiime_make_otu_table</span></code></a><a class="headerlink" href="#qiime-make-otu-table" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">make_otu_table.py</span></code>:</p>
<p>The module creates a BIOM table based on the OTU table and a taxonomy assignment if avaliable (will be available if the <code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code> is in the branch).</p>
<p>If chimera checking has been performed, the suspected chimeric sequences will be removed from the BIOM table.</p>
<p>The module also adds code for creating a summary of the BIOM table and a tab-delimited version thereof.</p>
<div class="section" id="id232">
<h3>Requires<a class="headerlink" href="#id232" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">An OTU table:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;otu_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="optional">
<h3>Optional<a class="headerlink" href="#optional" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A taxonomy assignment of the sequences:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;taxonomy&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id233">
<h3>Output<a class="headerlink" href="#id233" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the BIOM table in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table summary in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table in tab-delimited format in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If a fasta.chimera_removed file exists, will put the unfiltered BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;unfiltered_biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id234">
<h3>Parameters that can be set<a class="headerlink" href="#id234" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>skip_summary</td>
<td>&#160;</td>
<td>If passed, will not create the BIOM table summary.</td>
</tr>
<tr class="row-odd"><td>skip_tsv</td>
<td>&#160;</td>
<td>If passed, will not create the tsv version of the BIOM table.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id235">
<h3>Lines for parameter file<a class="headerlink" href="#id235" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_mk_otu_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_make_otu_table</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_phylo_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/make_otu_table.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="c1"># skip_summary:</span>
    <span class="c1"># skip_tsv:</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">mapping_fp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">qiime1_mapping</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
</div>
<div class="section" id="id236">
<h3>References<a class="headerlink" href="#id236" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_filter_samples_from_otu_table"></span><div class="section" id="qiime-filter-samples-from-otu-table">
<h2><a class="toc-backref" href="#id408"><code class="docutils literal"><span class="pre">qiime_filter_samples_from_otu_table</span></code></a><a class="headerlink" href="#qiime-filter-samples-from-otu-table" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">filter_samples_from_otu_table.py</span></code></p>
<div class="section" id="id237">
<h3>Requires<a class="headerlink" href="#id237" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id238">
<h3>Output<a class="headerlink" href="#id238" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table summary in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table in tab-delimited format in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the unfiltered BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;prefilter_biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id239">
<h3>Parameters that can be set<a class="headerlink" href="#id239" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>skip_summary</td>
<td>&#160;</td>
<td>If passed, will not create the BIOM table summary.</td>
</tr>
<tr class="row-odd"><td>skip_tsv</td>
<td>&#160;</td>
<td>If passed, will not create the tsv version of the BIOM table.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id240">
<h3>Lines for parameter file<a class="headerlink" href="#id240" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">filt_samp_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_filter_samples_from_otu_table</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_mk_otu_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/filter_samples_from_otu_table.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">mapping_fp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">mapping</span><span class="o">.</span><span class="n">txt</span>
        <span class="o">--</span><span class="n">min_count</span><span class="p">:</span> <span class="mi">100000</span>
</pre></div>
</div>
</div>
<div class="section" id="id241">
<h3>References<a class="headerlink" href="#id241" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_filter_otus"></span><div class="section" id="qiime-filter-otus">
<h2><a class="toc-backref" href="#id409"><code class="docutils literal"><span class="pre">qiime_filter_otus</span></code></a><a class="headerlink" href="#qiime-filter-otus" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">filter_otus_from_otu_table.py</span></code></p>
<div class="section" id="id242">
<h3>Requires<a class="headerlink" href="#id242" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id243">
<h3>Output<a class="headerlink" href="#id243" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table summary in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table in tab-delimited format in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the unfiltered BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;prefilter_biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id244">
<h3>Parameters that can be set<a class="headerlink" href="#id244" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>skip_summary</td>
<td>&#160;</td>
<td>If passed, will not create the BIOM table summary.</td>
</tr>
<tr class="row-odd"><td>skip_tsv</td>
<td>&#160;</td>
<td>If passed, will not create the tsv version of the BIOM table.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id245">
<h3>Lines for parameter file<a class="headerlink" href="#id245" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_filt_otus_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_filter_otus</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">filt_samp_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/filter_otus_from_otu_table.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">min_count_fraction</span><span class="p">:</span> <span class="mf">0.00005</span>
        <span class="o">--</span><span class="n">min_samples</span><span class="p">:</span> <span class="mi">10</span>
</pre></div>
</div>
</div>
<div class="section" id="id246">
<h3>References<a class="headerlink" href="#id246" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_sort_otu_table"></span><div class="section" id="qiime-sort-otu-table">
<h2><a class="toc-backref" href="#id410"><code class="docutils literal"><span class="pre">qiime_sort_otu_table</span></code></a><a class="headerlink" href="#qiime-sort-otu-table" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">sort_otu_table.py</span></code></p>
<div class="section" id="id247">
<h3>Requires<a class="headerlink" href="#id247" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id248">
<h3>Output<a class="headerlink" href="#id248" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the resulting BIOM table in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table summary in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_summary&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Puts the BIOM table in tab-delimited format in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;biom_table_tsv&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id249">
<h3>Parameters that can be set<a class="headerlink" href="#id249" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>skip_summary</td>
<td>&#160;</td>
<td>If passed, will not create the BIOM table summary.</td>
</tr>
<tr class="row-odd"><td>skip_tsv</td>
<td>&#160;</td>
<td>If passed, will not create the tsv version of the BIOM table.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id250">
<h3>Lines for parameter file<a class="headerlink" href="#id250" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_sort_otus_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_sort_otu_table</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">filt_samp_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.qiime_path}</span><span class="s1">/sort_otu_table.py&#39;</span>
    <span class="n">setenv</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">qiime_env</span><span class="p">}</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">sort_field</span><span class="p">:</span>   <span class="n">XXX</span>
</pre></div>
</div>
</div>
<div class="section" id="id251">
<h3>References<a class="headerlink" href="#id251" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
<span class="target" id="module-QIIME.qiime_divers"></span><div class="section" id="qiime-divers">
<h2><a class="toc-backref" href="#id411"><code class="docutils literal"><span class="pre">qiime_divers</span></code></a><a class="headerlink" href="#qiime-divers" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<p>A module for running QIIME’s <code class="docutils literal"><span class="pre">core_diversity_analyses.py</span></code>:</p>
<p>The module creates a BIOM table based on the OTU table and a taxonomy assignment if avaliable (will be available if the <code class="docutils literal"><span class="pre">qiime_assign_taxonomy</span></code> is in the branch).</p>
<p>If chimera checking has been performed, the suspected chimeric sequences will be removed from the BIOM table.</p>
<p>The module also adds code for creating a summary of the BIOM table and a tab-delimited version thereof.</p>
<div class="section" id="id252">
<h3>Requires<a class="headerlink" href="#id252" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BIOM table:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;biom_table&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id253">
<h3>Optional<a class="headerlink" href="#id253" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A phylogenetic tree:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;phylotree&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id254">
<h3>Output<a class="headerlink" href="#id254" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the core diversity directory name in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;diversity&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id255">
<h3>Parameters that can be set<a class="headerlink" href="#id255" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>–mapping_fp</td>
<td>&#160;</td>
<td>A path to the qiime mapping file (if not set, will use the mapping file passed in <code class="docutils literal"><span class="pre">qiime_prep</span></code>.</td>
</tr>
<tr class="row-odd"><td>–parameter_fp</td>
<td>&#160;</td>
<td>A path to a qiime parameter file.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id256">
<h3>Lines for parameter file<a class="headerlink" href="#id256" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">q_divers_1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">qiime_divers</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">q_filt_otus_1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">QIIME</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">core_diversity_analyses</span><span class="o">.</span><span class="n">py</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span> <span class="n">shared</span> <span class="mi">20</span>
    <span class="n">sampling_depth</span><span class="p">:</span> <span class="mi">109897</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">categories</span><span class="p">:</span> <span class="n">Disease</span><span class="p">,</span><span class="n">sex</span>
        <span class="o">--</span><span class="n">parameter_fp</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">parameter_file</span>
</pre></div>
</div>
</div>
<div class="section" id="id257">
<h3>References<a class="headerlink" href="#id257" title="Permalink to this headline">¶</a></h3>
<p>Caporaso, J.G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F.D., Costello, E.K., Fierer, N., Peña, A.G., Goodrich, J.K., Gordon, J.I. and Huttley, G.A., 2010. “QIIME allows analysis of high-throughput community sequencing data”. <em>Nature methods</em>, 7(5), pp.335-336.</p>
</div>
</div>
</div>
<div class="section" id="various-reporting-programs">
<h1>Various Reporting Programs<a class="headerlink" href="#various-reporting-programs" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id258">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#ngsplot" id="id412"><code class="docutils literal"><span class="pre">NGSplot</span></code></a></li>
<li><a class="reference internal" href="#id264" id="id413"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#module-collect-results" id="id414">Module <code class="docutils literal"><span class="pre">Collect_results</span></code></a></li>
</ul>
</div>
<span class="target" id="module-Reports.NGSplot"></span><div class="section" id="ngsplot">
<h2><a class="toc-backref" href="#id412"><code class="docutils literal"><span class="pre">NGSplot</span></code></a><a class="headerlink" href="#ngsplot" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running NGSplot:</p>
<p>Runs NGSplot on existing <strong>sorted</strong> BAM files.</p>
<p>Please make sure the BAM is sorted, such as following the <code class="docutils literal"><span class="pre">samtools</span></code> module</p>
<p>If this is a ChIP-seq experiment and you have controls defined, it will also run NGSplot for the sample:control comparison.</p>
<p>At the moment, the module works only at the sample scope. (BAM files in the project scope are rare!)</p>
<div class="section" id="id259">
<h3>Requires<a class="headerlink" href="#id259" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">BAM files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id260">
<h3>Output<a class="headerlink" href="#id260" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output NGS reports in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;NGSplot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">For ChIP-seq data, puts comparison reports in</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;NGSplot_vs_control&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id261">
<h3>Parameters that can be set<a class="headerlink" href="#id261" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>setenv</td>
<td>NGSPLOT=/path/to/ngsplot</td>
<td>Running NGSplot requires setting this EV.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id262">
<h3>Lines for parameter file<a class="headerlink" href="#id262" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">NGSplot_genebody</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span>             <span class="n">NGSplot</span>
    <span class="n">base</span><span class="p">:</span>               <span class="n">sam_base</span>
    <span class="n">script_path</span><span class="p">:</span>        <span class="n">Rscript</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ngsplot</span><span class="o">-</span><span class="mf">2.61</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">ngs</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">r</span>
    <span class="n">setenv</span><span class="p">:</span>             <span class="n">NGSPLOT</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ngsplot</span><span class="o">-</span><span class="mf">2.61</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">G</span><span class="p">:</span>             <span class="n">mm10</span>
        <span class="o">-</span><span class="n">R</span><span class="p">:</span>             <span class="n">genebody</span>
        <span class="o">-</span><span class="n">P</span><span class="p">:</span>             <span class="mi">20</span>
        <span class="o">-</span><span class="n">GO</span><span class="p">:</span>            <span class="n">hc</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>            <span class="n">shared</span> <span class="mi">20</span>
</pre></div>
</div>
</div>
<div class="section" id="id263">
<h3>References<a class="headerlink" href="#id263" title="Permalink to this headline">¶</a></h3>
<p>Shen, L., Shao, N., Liu, X. and Nestler, E., 2014. <strong>ngs.plot: Quick mining and visualization of next-generation sequencing data by integrating genomic databases</strong>. <em>BMC genomics</em>, 15(1), p.284.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.Reports.Multiqc"></span><div class="section" id="id264">
<h2><a class="toc-backref" href="#id413"><code class="docutils literal"><span class="pre">Multiqc</span></code> <sup>*</sup></a><a class="headerlink" href="#id264" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for preparing a MultiQC report for all samples</p>
<div class="section" id="id265">
<h3>Requires<a class="headerlink" href="#id265" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>No real requirements. Will give a report with information if one of the base steps produces reports that MultiQC can read, <em>e.g.</em> fastqc, bowtie2, samtools etc.</li>
</ul>
</div>
<div class="section" id="id266">
<h3>Output<a class="headerlink" href="#id266" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts report dir in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;Multiqc_report&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id267">
<h3>Lines for parameter file<a class="headerlink" href="#id267" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">firstMultQC</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Multiqc</span>
    <span class="n">base</span><span class="p">:</span>
    <span class="o">-</span> <span class="n">sam_bwt2_1</span>
    <span class="o">-</span> <span class="n">fqc_trim1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">multiqc</span>
</pre></div>
</div>
</div>
<div class="section" id="id268">
<h3>References<a class="headerlink" href="#id268" title="Permalink to this headline">¶</a></h3>
<p>Ewels, P., Magnusson, M., Lundin, S. and Käller, M., 2016. <strong>MultiQC: summarize analysis results for multiple tools and samples in a single report</strong>. <em>Bioinformatics</em>, 32(19), pp.3047-3048.</p>
</div>
</div>
<span class="target" id="module-Liron.Collect_results_module.Collect_results"></span><div class="section" id="module-collect-results">
<h2><a class="toc-backref" href="#id414">Module <code class="docutils literal"><span class="pre">Collect_results</span></code></a><a class="headerlink" href="#module-collect-results" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module to Collect and merge/append results from all base steps directories:
This module will search for each base step for all the results files with a common name pattern [Regular expression].
The search will be done within the base step result directories.
The sample name could be inferred for each result file base on the parent directory name and added to the merged file [as new column named “Samples”].
All the results files will be append [by default] or merged by a common column name.
The merge files can then be convert individually to pivot table file</dd>
</dl>
<div class="section" id="id269">
<h3>Requires<a class="headerlink" href="#id269" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul class="simple">
<li>Tab delimited files with common name pattern found within the base step data directories:</li>
<li><code class="docutils literal"><span class="pre">For</span> <span class="pre">example</span> <span class="pre">files</span> <span class="pre">ending</span> <span class="pre">with</span> <span class="pre">.out</span></code></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id270">
<h3>Output<a class="headerlink" href="#id270" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul class="simple">
<li>Generate merged tab delimited files:</li>
<li><code class="docutils literal"><span class="pre">Will</span> <span class="pre">generate</span> <span class="pre">file</span> <span class="pre">for</span> <span class="pre">each</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">base</span> <span class="pre">steps</span> <span class="pre">with</span> <span class="pre">the</span> <span class="pre">file</span> <span class="pre">ending</span> <span class="pre">with</span> <span class="pre">.merg</span></code></li>
<li>Can also generate Excel file with sheet for each base step</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id271">
<h3>Parameters that can be set<a class="headerlink" href="#id271" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>PARAMETER NAME</td>
<td>POSSIBLE VALUES</td>
<td>DESCRIPTION</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id272">
<h3>Comments<a class="headerlink" href="#id272" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id273">
<h3>Lines for parameter file<a class="headerlink" href="#id273" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt>::</dt>
<dd><dl class="first last docutils">
<dt>Step_Name:                            # Name of this step</dt>
<dd><p class="first">module: Collect_results           # Name of the used module
base:                             # Name of the step [or list of names] to run after and collect results from [must be after a merge step]
script_path:                      # Command for running the a merging script</p>
<blockquote>
<div># If this line is empty or missing it will try using the module’s associated script</div></blockquote>
<dl class="last docutils">
<dt>redirects:</dt>
<dd><p class="first">-R:                           # Regular expression to find result files
–Merge_by:                   # Merge files by common column
–header:                     # Don’t use a header row, use integers instead [0,1,2,3…], easy to use with –pivot option
–Excel:                      # Collect all results to excel file split by sheets
–add_samples_names:          # Infer and add samples names from file parent directory to “Samples” column
–pivot:                      # Convert to pivot table by [index columns values]</p>
<blockquote>
<div># If with the options: -add_samples_names and –header  it is possible to use: ‘’‘Samples’’ ‘5’ ‘0’‘</div></blockquote>
<p class="last">–MetaData:                   # Use external MetaData file as the base for merging
–split_by:                   # Split the data in the columns [index &lt;columns&gt; values] before pivot
–sep:                        # Columns separator for input file
-T:                           # Write Transpose output</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="id274">
<h3>References<a class="headerlink" href="#id274" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>
<div class="section" id="sequence-searching-related-tasks">
<h1>Sequence-Searching Related Tasks<a class="headerlink" href="#sequence-searching-related-tasks" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id275">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#makeblastdb" id="id415"><code class="docutils literal"><span class="pre">makeblastdb</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#blast" id="id416"><code class="docutils literal"><span class="pre">blast</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#blast-new-alternative-implementation" id="id417"><code class="docutils literal"><span class="pre">blast_new</span></code> - Alternative implementation</a></li>
<li><a class="reference internal" href="#parse-blast" id="id418"><code class="docutils literal"><span class="pre">parse_blast</span></code></a></li>
<li><a class="reference internal" href="#module-gassst" id="id419">Module <code class="docutils literal"><span class="pre">Gassst</span></code></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.searching.makeblastdb"></span><div class="section" id="makeblastdb">
<h2><a class="toc-backref" href="#id415"><code class="docutils literal"><span class="pre">makeblastdb</span></code> <sup>*</sup></a><a class="headerlink" href="#makeblastdb" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>Create a blastdb from a fasta file</p>
<div class="section" id="id276">
<h3>Requires<a class="headerlink" href="#id276" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fastq files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta&quot;][&quot;nucl&quot;|&quot;prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><dl class="first docutils">
<dt>Or (if ‘projectBLAST’ is set)</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta&quot;][&quot;nucl&quot;|&quot;prot&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id277">
<h3>Output<a class="headerlink" href="#id277" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">A BLAST database in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blastdb&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blastdb.nucl&quot;|&quot;blastdb.prot&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blastdb.nucl.log&quot;|&quot;blastdb.prot.log&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Or (if ‘projectBLAST’ is set):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blastdb&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blastdb.nucl&quot;|&quot;blastdb.prot&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blastdb.nucl.log&quot;|&quot;blastdb.prot.log&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id278">
<h3>Parameters that can be set:<a class="headerlink" href="#id278" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample|project</td>
<td>Set if project-wide or sample fasta slot should be used</td>
</tr>
<tr class="row-odd"><td>-dbtype</td>
<td>nucl/prot</td>
<td>This is a compulsory redirected parameter.Helps the module decide which fasta file to use.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id279">
<h3>Lines for parameter file<a class="headerlink" href="#id279" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">mkblst1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">makeblastdb</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">trinity1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">makeblastdb</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">dbtype</span><span class="p">:</span> <span class="n">nucl</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
</pre></div>
</div>
</div>
<div class="section" id="id280">
<h3>References<a class="headerlink" href="#id280" title="Permalink to this headline">¶</a></h3>
<p>Altschul, S.F., Madden, T.L., Schäffer, A.A., Zhang, J., Zhang, Z., Miller, W. and Lipman, D.J., 1997. <strong>Gapped BLAST and PSI-BLAST: a new generation of protein database search programs</strong>. <em>Nucleic acids research</em>, 25(17), pp.3389-3402.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.searching.blast"></span><div class="section" id="blast">
<h2><a class="toc-backref" href="#id416"><code class="docutils literal"><span class="pre">blast</span></code> <sup>*</sup></a><a class="headerlink" href="#blast" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for executing blast on a nucleotide or protein fasta file.
The search can be either on a sample fasta or on a project-wide fasta.
It can use the fasta as a database or as a query.
If used as a database, you must call the makeblastdb module prior to this step.</p>
<div class="section" id="id281">
<h3>Requires:<a class="headerlink" href="#id281" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in one of the following slots for sample-wise blast:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">or fasta files in one of the following slots for project-wise blast:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">or a <code class="docutils literal"><span class="pre">makeblastdb</span></code> index in one of the following slots:</p>
<blockquote>
<div><ul>
<li><p class="first">When ‘scope’ is set to ‘project’</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blastdb.nucl&quot;|&quot;blastdb.prot&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blastdb.log.nucl&quot;|&quot;blastdb.log.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">When ‘scope’ is set to ‘sample’</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blastdb.nucl&quot;|&quot;blastdb.prot&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blastdb.log.nucl&quot;|&quot;blastdb.log.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id282">
<h3>Output:<a class="headerlink" href="#id282" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts BLAST output files in the following slots for sample-wise blast:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blast.nucl&quot;|&quot;blast.prot&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blast&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">puts fasta output files in the following slots for project-wise blast:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blast.nucl&quot;|&quot;blast.prot&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blast&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id283">
<h3>Parameters that can be set<a class="headerlink" href="#id283" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="40%" />
<col width="40%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>sample|project</td>
<td>Set if project-wide fasta slot should be used</td>
</tr>
<tr class="row-odd"><td>dbscope</td>
<td>sample|project</td>
<td>If scope=sample and you want to use internal fasta and database, this will tell</td>
</tr>
<tr class="row-even"><td>fasta2use</td>
<td>nucl|prot</td>
<td>Helps the module decide which fasta file to use.</td>
</tr>
<tr class="row-odd"><td>db2use</td>
<td>nucl|prot</td>
<td>Helps the module decide which database to use.</td>
</tr>
<tr class="row-even"><td>-query | -db</td>
<td>Path to fasta or BLAST index</td>
<td>The sequences to use as query, using the internal database as database, or a BLAST database index, using the internal fasta as query.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id284">
<h3>Lines for parameter file<a class="headerlink" href="#id284" title="Permalink to this headline">¶</a></h3>
<p>External query, project-wise fasta database (must be proceeded by <code class="docutils literal"><span class="pre">makeblastdb</span></code> module):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">blastOnAssembl</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">blast</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">mkblst1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">blastn</span>
    <span class="n">fasta2use</span><span class="p">:</span> <span class="n">nucl</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">evalue</span><span class="p">:</span> <span class="s1">&#39;0.0001&#39;</span>
        <span class="o">-</span><span class="n">num_descriptions</span><span class="p">:</span> <span class="s1">&#39;20&#39;</span>
        <span class="o">-</span><span class="n">num_threads</span><span class="p">:</span> <span class="s1">&#39;40&#39;</span>
        <span class="o">-</span><span class="n">outfmt</span><span class="p">:</span> <span class="s1">&#39;&quot;6 qseqid sallseqid qlen slen qstart qend sstart send length</span>
            <span class="n">evalue</span> <span class="n">bitscore</span> <span class="n">score</span> <span class="n">pident</span> <span class="n">qframe</span><span class="s2">&quot;&#39;</span>
        <span class="o">-</span><span class="n">query</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">query</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>Sample specific fasta, external database:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">sprot</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">blast</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">sample_assembl</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">blastx</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">db</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">database</span><span class="o">/</span><span class="n">index</span>
        <span class="o">-</span><span class="n">evalue</span><span class="p">:</span> <span class="s1">&#39;0.0001&#39;</span>
        <span class="o">-</span><span class="n">max_target_seqs</span><span class="p">:</span> <span class="s1">&#39;5&#39;</span>
        <span class="o">-</span><span class="n">num_of_proc</span><span class="p">:</span> <span class="s1">&#39;20&#39;</span>
        <span class="o">-</span><span class="n">num_threads</span><span class="p">:</span> <span class="s1">&#39;20&#39;</span>
        <span class="o">-</span><span class="n">outfmt</span><span class="p">:</span> <span class="s1">&#39;&quot;6 qseqid sallseqid qlen slen qstart qend sstart send length</span>
            <span class="n">evalue</span> <span class="n">bitscore</span> <span class="n">score</span> <span class="n">pident</span> <span class="n">qframe</span><span class="s2">&quot;&#39;</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">sample</span>
</pre></div>
</div>
</div>
<div class="section" id="id285">
<h3>References<a class="headerlink" href="#id285" title="Permalink to this headline">¶</a></h3>
<p>Altschul, S.F., Madden, T.L., Schäffer, A.A., Zhang, J., Zhang, Z., Miller, W. and Lipman, D.J., 1997. <strong>Gapped BLAST and PSI-BLAST: a new generation of protein database search programs</strong>. <em>Nucleic acids research</em>, 25(17), pp.3389-3402.</p>
</div>
</div>
<span class="target" id="module-searching.blast_new"></span><div class="section" id="blast-new-alternative-implementation">
<h2><a class="toc-backref" href="#id417"><code class="docutils literal"><span class="pre">blast_new</span></code> - Alternative implementation</a><a class="headerlink" href="#blast-new-alternative-implementation" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A class that defines a module for executing BLAST of any type on a nucleotide or protein fasta file.
The search can be either on a sample fasta or on a project-wide fasta.
It can use the fasta as a database or as a query.
If used as a database, you must call the makeblastdb module prior to this step.</p>
<p>both <code class="docutils literal"><span class="pre">-query</span></code> and <code class="docutils literal"><span class="pre">-db</span></code> must be passed in the redirected parameters. They should be set to one of the following values:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">sample</span></code> - The <code class="docutils literal"><span class="pre">-query</span></code> or <code class="docutils literal"><span class="pre">-db</span></code> should be taken from the sample scope</li>
<li><code class="docutils literal"><span class="pre">project</span></code> - The <code class="docutils literal"><span class="pre">-query</span></code> or <code class="docutils literal"><span class="pre">-db</span></code> should be taken from the project scope</li>
<li><strong>A path</strong> - A path to a <code class="docutils literal"><span class="pre">fasta</span></code> file or <code class="docutils literal"><span class="pre">makeblastdb</span></code> database to use in the parameter as-is.</li>
</ul>
<p>The type of fasta and database to use are set with the <code class="docutils literal"><span class="pre">querytype</span></code> and <code class="docutils literal"><span class="pre">dbtype</span></code> parameters, respectively. These must be set if the respective parameter (<code class="docutils literal"><span class="pre">-query</span></code> or <code class="docutils literal"><span class="pre">-db</span></code>) is set to <code class="docutils literal"><span class="pre">sample</span></code> or <code class="docutils literal"><span class="pre">project</span></code>.</p>
<div class="section" id="id286">
<h3>Requires:<a class="headerlink" href="#id286" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">fasta files in one of the following slots for sample-wise blast:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">or fasta files in one of the following slots for project-wise blast:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">or a <code class="docutils literal"><span class="pre">makeblastdb</span></code> index in one of the following slots:</p>
<blockquote>
<div><ul>
<li><p class="first">When <code class="docutils literal"><span class="pre">-db</span></code> is set to ‘project’</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blastdb.nucl&quot;|&quot;blastdb.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">When <code class="docutils literal"><span class="pre">-db</span></code> is set to ‘sample’</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blastdb.nucl&quot;|&quot;blastdb.prot&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id287">
<h3>Output:<a class="headerlink" href="#id287" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">puts BLAST output files in the following slots for sample-wise blast:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blast.nucl&quot;|&quot;blast.prot&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blast&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">puts fasta output files in the following slots for project-wise blast:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blast.nucl&quot;|&quot;blast.prot&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blast&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id288">
<h3>Parameters that can be set<a class="headerlink" href="#id288" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="40%" />
<col width="40%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>dbtype</td>
<td>sample|project</td>
<td>If scope=sample and you want to use internal fasta and database, this will tell</td>
</tr>
<tr class="row-odd"><td>querytype</td>
<td>nucl|prot</td>
<td>Helps the module decide which fasta file to use.</td>
</tr>
<tr class="row-even"><td>-query | -db</td>
<td><a href="#id289"><span class="problematic" id="id290">``</span></a>sample``|``project``|&lt;Path to fasta or BLAST index&gt;</td>
<td>Redirected compulsory parameters. Set to <code class="docutils literal"><span class="pre">sample</span></code> for sample-scope files, to <code class="docutils literal"><span class="pre">project</span></code> for project-scope files, or to a path for an external file.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">You can’t set both <code class="docutils literal"><span class="pre">-db</span></code> and <code class="docutils literal"><span class="pre">-query</span></code> to external files. One of them at least has to be <code class="docutils literal"><span class="pre">sample</span></code> or <code class="docutils literal"><span class="pre">project</span></code>.</p>
</div>
</div>
<div class="section" id="id291">
<h3>Lines for parameter file<a class="headerlink" href="#id291" title="Permalink to this headline">¶</a></h3>
<p>External query, project-wise <em>nucl</em>-type database (must be proceeded by <code class="docutils literal"><span class="pre">makeblastdb</span></code> module):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">tbl_blst_int</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">blast_new</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">mkblst1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">blastn</span>
    <span class="n">dbtype</span><span class="p">:</span> <span class="n">nucl</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">query</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">query</span><span class="o">.</span><span class="n">fasta</span>
        <span class="o">-</span><span class="n">db</span><span class="p">:</span>  <span class="n">project</span>
        <span class="o">-</span><span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
        <span class="o">-</span><span class="n">max_target_seqs</span><span class="p">:</span> <span class="mi">5</span>
        <span class="o">-</span><span class="n">num_of_proc</span><span class="p">:</span> <span class="mi">20</span>
        <span class="o">-</span><span class="n">num_threads</span><span class="p">:</span> <span class="mi">20</span>
</pre></div>
</div>
<p>Sample specific <em>prot</em>-type fasta, external database:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">tbl_blst_ext</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">blast_new</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">prokka1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">blastp</span>
    <span class="n">querytype</span><span class="p">:</span> <span class="n">prot</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">db</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">blasdb</span><span class="o">.</span><span class="n">ind</span>
        <span class="o">-</span><span class="n">query</span><span class="p">:</span> <span class="n">sample</span>
        <span class="o">-</span><span class="n">evalue</span><span class="p">:</span> <span class="mf">0.0001</span>
</pre></div>
</div>
</div>
<div class="section" id="id292">
<h3>References<a class="headerlink" href="#id292" title="Permalink to this headline">¶</a></h3>
<p>Altschul, S.F., Madden, T.L., Schäffer, A.A., Zhang, J., Zhang, Z., Miller, W. and Lipman, D.J., 1997. <strong>Gapped BLAST and PSI-BLAST: a new generation of protein database search programs</strong>. <em>Nucleic acids research</em>, 25(17), pp.3389-3402.</p>
</div>
</div>
<span class="target" id="module-searching.parse_blast"></span><div class="section" id="parse-blast">
<h2><a class="toc-backref" href="#id418"><code class="docutils literal"><span class="pre">parse_blast</span></code></a><a class="headerlink" href="#parse-blast" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running <code class="docutils literal"><span class="pre">parse_blast_general.R</span></code>:</p>
<p>The <code class="docutils literal"><span class="pre">parse_blast_general</span></code> script will be available on github in the near future.</p>
<p>The program performs the following tasks:</p>
<ol class="arabic simple">
<li>It adds annotation to raw tabular BLAST output files,</li>
<li>filters the BLAST results by several possible fields,</li>
<li>selects the best hit for a group when passed a grouping field and</li>
<li>extracts the sequences equivalent to the alignments.</li>
</ol>
<div class="section" id="id293">
<h3>Requires<a class="headerlink" href="#id293" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Tabular BLAST result files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blast.nucl|blast.fasta&quot;]</span></code> if <code class="docutils literal"><span class="pre">scope</span> <span class="pre">=</span> <span class="pre">sample</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blast.nucl|blast.fasta&quot;]</span></code> if <code class="docutils literal"><span class="pre">scope</span> <span class="pre">=</span> <span class="pre">project</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id294">
<h3>Output<a class="headerlink" href="#id294" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts the parsed report in:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;blast.parsed&quot;]</span></code>  if <code class="docutils literal"><span class="pre">scope</span> <span class="pre">=</span> <span class="pre">sample</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&quot;blast.parsed&quot;]</span></code> if <code class="docutils literal"><span class="pre">scope</span> <span class="pre">=</span> <span class="pre">project</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id295">
<h3>Parameters that can be set<a class="headerlink" href="#id295" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>fasta2use</td>
<td><code class="docutils literal"><span class="pre">nucl|prot</span></code></td>
<td>If both nucl and prot BLAST reports exist, you have to specify which one to use with this parameter.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id296">
<h3>Lines for parameter file<a class="headerlink" href="#id296" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">parse_blast_table</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">parse_blast</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">blst_table</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">paths</span><span class="o">.</span><span class="n">parse_blast</span><span class="p">}</span>
    <span class="n">blast_merge_path</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{Vars.paths.compare_blast_parsed_reports}</span><span class="s1"> --variable evalue --full_txt_output&#39;</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">sample</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">columns2keep</span><span class="p">:</span> <span class="s1">&#39;&quot;group name accession qseqid sallseqid evalue bitscore</span>
            <span class="n">score</span> <span class="n">pident</span> <span class="n">coverage</span> <span class="n">align_len</span><span class="s2">&quot;&#39;</span>
        <span class="o">--</span><span class="n">dbtable</span><span class="p">:</span> <span class="p">{</span><span class="n">Vars</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">gene_list</span><span class="o">.</span><span class="n">table</span><span class="p">}</span>
        <span class="o">--</span><span class="n">group_dif_name</span><span class="p">:</span> <span class="c1"># See parse_blast_general.R documentation for how this is to be specified</span>
        <span class="o">--</span><span class="n">max_evalue</span><span class="p">:</span> <span class="mi">1</span><span class="n">e</span><span class="o">-</span><span class="mi">7</span>
        <span class="o">--</span><span class="n">merge_blast</span><span class="p">:</span> <span class="n">qseqid</span>
        <span class="o">--</span><span class="n">merge_metadata</span><span class="p">:</span> <span class="c1"># See parse_blast_general.R documentation for how this is to be specified</span>
        <span class="o">--</span><span class="n">min_align_len</span><span class="p">:</span> <span class="mi">30</span>
        <span class="o">--</span><span class="n">min_coverage</span><span class="p">:</span> <span class="mi">60</span>
        <span class="o">--</span><span class="n">names</span><span class="p">:</span> <span class="s1">&#39;&quot;qseqid sallseqid qlen slen qstart qend sstart send length evalue</span>
            <span class="n">bitscore</span> <span class="n">score</span> <span class="n">pident</span> <span class="n">qframe</span><span class="s2">&quot;&#39;</span>
        <span class="o">--</span><span class="n">num_hits</span><span class="p">:</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
<span class="target" id="module-Liron.Gassst_module.Gassst"></span><div class="section" id="module-gassst">
<h2><a class="toc-backref" href="#id419">Module <code class="docutils literal"><span class="pre">Gassst</span></code></a><a class="headerlink" href="#module-gassst" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Liron Levin</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This module was developed as part of a study led by Dr. Jacob Moran Gilad</p>
</div>
<dl class="docutils">
<dt>SHORT DESCRIPTION</dt>
<dd>A module for executing Gassst on a nucleotide fasta file.
The search can be either on a sample fasta or on a project-wide fasta.
It can use the fasta as a database or as a query.</dd>
</dl>
<div class="section" id="id297">
<h3>Requires<a class="headerlink" href="#id297" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><dl class="first docutils">
<dt>fasta files in the following slot for sample-wise Gassst:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>or fasta files in the following slots for project-wise Gassst:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id298">
<h3>Output<a class="headerlink" href="#id298" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul>
<li><dl class="first docutils">
<dt>puts Gassst output files in the following slots for sample-wise Gassst:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blast&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;blast.nucl&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>puts fasta output files in the following slots for project-wise Gassst:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blast&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&quot;blast.nucl&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="id299">
<h3>Parameters that can be set<a class="headerlink" href="#id299" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>project/sample</td>
<td>Set if project-wide fasta.nucl file type should be used [project] the default is sample-wide fasta.nucl file type</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id300">
<h3>Comments<a class="headerlink" href="#id300" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div>Only -d [database] or -i [query] not both
The Gassst module will generate blast like output with fields:
“qseqid sallseqid qlen slen qstart qend sstart send length evalue sseq”</div></blockquote>
</div>
<div class="section" id="id301">
<h3>Lines for parameter file<a class="headerlink" href="#id301" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Step_Name</span><span class="p">:</span>                         <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Gassst</span>                 <span class="c1"># Name of the module to use</span>
    <span class="n">base</span><span class="p">:</span>                          <span class="c1"># Name of the step [or list of names] to run after [mast be after a fasta generating step]</span>
    <span class="n">script_path</span><span class="p">:</span>                   <span class="c1"># Command for running the Gassst script</span>
                                   <span class="c1"># The Gassst module will generate blast like output with fields:</span>
                                   <span class="c1"># &quot;qseqid sallseqid qlen slen qstart qend sstart send length evalue sseq&quot;</span>
    <span class="n">scope</span><span class="p">:</span>                         <span class="c1"># Set if project-wide fasta.nucl file type should be used [project] the default is sample-wide fasta.nucl file type</span>
    <span class="n">qsub_params</span><span class="p">:</span>
        <span class="o">-</span><span class="n">pe</span><span class="p">:</span>                       <span class="c1"># Number of CPUs to reserve for this analysis</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">h</span><span class="p">:</span>                        <span class="c1"># Max hits per query, for downstream best hit will be chosen!</span>
        <span class="o">-</span><span class="n">i</span><span class="p">:</span>                        <span class="c1"># Only -d [database] or -i [query] not both</span>
        <span class="o">-</span><span class="n">l</span><span class="p">:</span>                        <span class="c1"># Complexity_filter off</span>
        <span class="o">-</span><span class="n">d</span><span class="p">:</span>                        <span class="c1"># Only -d [database] or -i [query] not both</span>
        <span class="o">-</span><span class="n">n</span><span class="p">:</span>                        <span class="c1"># Number of CPUs running Gassst</span>
        <span class="o">-</span><span class="n">p</span><span class="p">:</span>                        <span class="c1"># Minimum percentage of identity. Must be in the interval [0 100]</span>
</pre></div>
</div>
</div>
<div class="section" id="id302">
<h3>References<a class="headerlink" href="#id302" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>
<div class="section" id="bam-conversion-to-other-formats">
<h1>BAM Conversion to Other Formats<a class="headerlink" href="#bam-conversion-to-other-formats" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id303">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#genomecoveragebed" id="id420"><code class="docutils literal"><span class="pre">genomeCoverageBed</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#ucsc-bw-wig" id="id421"><code class="docutils literal"><span class="pre">UCSC_BW_wig</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#igv-count" id="id422"><code class="docutils literal"><span class="pre">IGV_count</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#igv-totdf" id="id423"><code class="docutils literal"><span class="pre">IGV_toTDF</span></code> <sup>*</sup></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.mapping.genomeCoverageBed"></span><div class="section" id="genomecoveragebed">
<h2><a class="toc-backref" href="#id420"><code class="docutils literal"><span class="pre">genomeCoverageBed</span></code> <sup>*</sup></a><a class="headerlink" href="#genomecoveragebed" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running bedtools genomecov:</p>
<p>The module builds a bedgraph (bdg) file based on an existing BAM file.</p>
<div class="section" id="id304">
<h3>Requires<a class="headerlink" href="#id304" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">BAM file in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id305">
<h3>Output<a class="headerlink" href="#id305" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output BedGraph files in the following slots:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bdg&quot;]</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id306">
<h3>Parameters that can be set<a class="headerlink" href="#id306" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-g</td>
<td>path to chrom.sizes</td>
<td>You must redirect the -g parameter. Create the chrom.sizes file for the reference genome with <code class="docutils literal"><span class="pre">samtools</span> <span class="pre">faidx</span></code> followed by <code class="docutils literal"><span class="pre">cut</span> <span class="pre">-f1,2</span></code>.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id307">
<h3>Lines for parameter file<a class="headerlink" href="#id307" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">genCovBed_bwt1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">genomeCoverageBed</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">sam_bwt1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bedtools</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">genomeCoverageBed</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">bg</span><span class="p">:</span> 
        <span class="o">-</span><span class="n">g</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">/</span><span class="n">ref_genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
</pre></div>
</div>
</div>
<div class="section" id="id308">
<h3>References<a class="headerlink" href="#id308" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<span class="target" id="module-main_NSF_classes.UCSCtools.UCSC_BW_wig"></span><div class="section" id="ucsc-bw-wig">
<h2><a class="toc-backref" href="#id421"><code class="docutils literal"><span class="pre">UCSC_BW_wig</span></code> <sup>*</sup></a><a class="headerlink" href="#ucsc-bw-wig" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for creating wig and bigwig files using UCSC tools:</p>
<p>The module creates bigwig and wig files from the current active BedGraph file.</p>
<div class="section" id="id309">
<h3>Requires<a class="headerlink" href="#id309" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">BedGraph file in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bdg&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id310">
<h3>Output<a class="headerlink" href="#id310" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output sam files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bw&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id311">
<h3>Parameters that can be set<a class="headerlink" href="#id311" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>bedGraphToBigWig_params</td>
<td><em>e.g.</em> -blockSize=10 -itemsPerSlot=20</td>
<td>Parameters to pass to <code class="docutils literal"><span class="pre">bedGraphToBigWig</span></code></td>
</tr>
<tr class="row-odd"><td>bigWigToWig_params</td>
<td><em>e.g.</em> -chrom X1 -start X2 -end X3</td>
<td>Parameters to pass to <code class="docutils literal"><span class="pre">bigWigToWig</span></code></td>
</tr>
<tr class="row-even"><td>script_path</td>
<td>&#160;</td>
<td>Path to dir where UCSC tools are located.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Set <code class="docutils literal"><span class="pre">script_path</span></code> to the path ro the UCSC tools, not to a specific tool!!!
Both <code class="docutils literal"><span class="pre">bedGraphToBigWig</span></code> and <code class="docutils literal"><span class="pre">bigWigToWig</span></code> will be executed. To set specific params, use <code class="docutils literal"><span class="pre">bedGraphToBigWig_params</span></code> and <code class="docutils literal"><span class="pre">bigWigToWig_params</span></code>, respectively.</p>
</div>
</div>
<div class="section" id="id312">
<h3>Lines for parameter file<a class="headerlink" href="#id312" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">UCSCmap_bams</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">UCSC_BW_wig</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">genCovBed</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">kentUtils</span><span class="o">/</span><span class="nb">bin</span>
    <span class="n">bedGraphToBigWig_params</span><span class="p">:</span> <span class="o">-</span><span class="n">blockSize</span> <span class="mi">10</span> <span class="o">-</span><span class="n">itemsPerSlot</span> <span class="mi">20</span>
    <span class="n">bigWigToWig_params</span><span class="p">:</span> <span class="o">-</span><span class="n">chrom</span> <span class="n">X1</span> <span class="o">-</span><span class="n">start</span> <span class="n">X2</span> <span class="o">-</span><span class="n">end</span> <span class="n">X3</span>
    <span class="n">genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
</pre></div>
</div>
</div>
<div class="section" id="id313">
<h3>References<a class="headerlink" href="#id313" title="Permalink to this headline">¶</a></h3>
<p>Kent, W.J., Sugnet, C.W., Furey, T.S., Roskin, K.M., Pringle, T.H., Zahler, A.M. and Haussler, D., 2002. <strong>The human genome browser at UCSC</strong>. <em>Genome research</em>, 12(6), pp.996-1006.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.IGVtools.IGV_count"></span><div class="section" id="igv-count">
<h2><a class="toc-backref" href="#id422"><code class="docutils literal"><span class="pre">IGV_count</span></code> <sup>*</sup></a><a class="headerlink" href="#igv-count" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running IGVtools count:</p>
<div class="section" id="id314">
<h3>Requires<a class="headerlink" href="#id314" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Either SAM or BAM files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;sam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id315">
<h3>Output<a class="headerlink" href="#id315" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output tdf or wig files in one the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;tdf&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id316">
<h3>Parameters that can be set<a class="headerlink" href="#id316" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>format</td>
<td>wig|tdf</td>
<td>Determines whether to create a ‘wig’ or ‘tdf’ file.</td>
</tr>
<tr class="row-odd"><td>genome</td>
<td>&#160;</td>
<td>Path to chrom.sizes file for reference genome</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id317">
<h3>Lines for parameter file<a class="headerlink" href="#id317" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">IGVcount1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">IGV_count</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">samtools1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="n">java</span> <span class="o">-</span><span class="n">Xmx1500m</span> <span class="o">-</span><span class="n">jar</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">igvtools</span><span class="o">.</span><span class="n">jar</span> <span class="n">count</span>
    <span class="nb">format</span><span class="p">:</span> <span class="n">tdf</span>   <span class="c1"># Options: &#39;tdf&#39; or &#39;wig&#39;</span>
    <span class="n">genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
</pre></div>
</div>
</div>
<div class="section" id="id318">
<h3>References<a class="headerlink" href="#id318" title="Permalink to this headline">¶</a></h3>
<p>Thorvaldsdóttir, H., Robinson, J.T. and Mesirov, J.P., 2013. <strong>Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration</strong>. <em>Briefings in bioinformatics</em>, 14(2), pp.178-192.</p>
</div>
</div>
<span class="target" id="module-main_NSF_classes.IGVtools.IGV_toTDF"></span><div class="section" id="igv-totdf">
<h2><a class="toc-backref" href="#id423"><code class="docutils literal"><span class="pre">IGV_toTDF</span></code> <sup>*</sup></a><a class="headerlink" href="#igv-totdf" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running IGVtools toIGV:</p>
<div class="section" id="id319">
<h3>Requires<a class="headerlink" href="#id319" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">WIG file in the following slot:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;wig&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id320">
<h3>Output<a class="headerlink" href="#id320" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output tdf file in one the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;tdf&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id321">
<h3>Parameters that can be set<a class="headerlink" href="#id321" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>genome</td>
<td>&#160;</td>
<td>Path to chrom.sizes file for reference genome</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id322">
<h3>Lines for parameter file<a class="headerlink" href="#id322" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">IGV2TDF</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">IGV_toTDF</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">samtools1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">java</span> <span class="o">-</span><span class="n">Xmx1500m</span> <span class="o">-</span><span class="n">jar</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">igvtools</span><span class="o">.</span><span class="n">jar</span> <span class="n">toTDF</span> 
    <span class="n">genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
</pre></div>
</div>
</div>
<div class="section" id="id323">
<h3>References<a class="headerlink" href="#id323" title="Permalink to this headline">¶</a></h3>
<p>Thorvaldsdóttir, H., Robinson, J.T. and Mesirov, J.P., 2013. <strong>Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration</strong>. <em>Briefings in bioinformatics</em>, 14(2), pp.178-192.</p>
</div>
</div>
</div>
<div class="section" id="variant-related-tasks">
<h1>Variant-related Tasks<a class="headerlink" href="#variant-related-tasks" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id324">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#freebayes" id="id424"><code class="docutils literal"><span class="pre">freebayes</span></code></a></li>
<li><a class="reference internal" href="#mpileup-varscan" id="id425"><code class="docutils literal"><span class="pre">mpileup_varscan</span></code></a></li>
<li><a class="reference internal" href="#vcftools" id="id426"><code class="docutils literal"><span class="pre">vcftools</span></code> <sup>*</sup></a></li>
</ul>
</div>
<span class="target" id="module-variants.freebayes"></span><div class="section" id="freebayes">
<h2><a class="toc-backref" href="#id424"><code class="docutils literal"><span class="pre">freebayes</span></code></a><a class="headerlink" href="#freebayes" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for identifying variants by running freebayes:</p>
<div class="section" id="id325">
<h3>Requires<a class="headerlink" href="#id325" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">BAM files in the the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Genome reference fasta files in the the following slot (the slot should be populated by the module that created the <code class="docutils literal"><span class="pre">bam</span></code> file):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;reference&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>..Note:: Do not specify the reference (-f), since it is filled in automatically by neatseq-flow</p>
</div>
<div class="section" id="id326">
<h3>Output<a class="headerlink" href="#id326" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set to <code class="docutils literal"><span class="pre">sample</span></code>:</p>
<blockquote>
<div><ul>
<li><p class="first">Puts output files in:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;vcf&quot;]</span></code>  (if <code class="docutils literal"><span class="pre">output_type</span></code> is set to <code class="docutils literal"><span class="pre">vcf</span></code>)
<code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;gvcf&quot;]</span></code> (if <code class="docutils literal"><span class="pre">output_type</span></code> is set to <code class="docutils literal"><span class="pre">gvcf</span></code>)</p>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set to <code class="docutils literal"><span class="pre">project</span></code>:</p>
<blockquote>
<div><ul>
<li><p class="first">Puts output files in:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">sample_data[&quot;vcf&quot;]</span></code>  (if <code class="docutils literal"><span class="pre">output_type</span></code> is set to <code class="docutils literal"><span class="pre">vcf</span></code>)
<code class="docutils literal"><span class="pre">sample_data[&quot;gvcf&quot;]</span></code> (if <code class="docutils literal"><span class="pre">output_type</span></code> is set to <code class="docutils literal"><span class="pre">gvcf</span></code>)</p>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id327">
<h3>Parameters that can be set<a class="headerlink" href="#id327" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>output_type</td>
<td>vcf|gvcf</td>
<td>The type of output produced by freebayes. (Can be specified alternatively with appropriate redirects)</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id328">
<h3>Comments<a class="headerlink" href="#id328" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id329">
<h3>Lines for parameter file<a class="headerlink" href="#id329" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">freebayes1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">freebayes</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">samtools1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">freebayes</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">sample</span>
    <span class="n">output_type</span><span class="p">:</span> <span class="n">vcf</span>
    <span class="n">redirects</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">strict</span><span class="o">-</span><span class="n">vcf</span><span class="p">:</span>
</pre></div>
</div>
</div>
<div class="section" id="id330">
<h3>References<a class="headerlink" href="#id330" title="Permalink to this headline">¶</a></h3>
<p>Marth GT, Korf I, Yandell MD, Yeh RT, Gu Z, Zakeri H, Stitziel NO, Hillier L, Kwok PY, Gish WR: <strong>A general approach to single-nucleotide polymorphism discovery</strong>. <em>Nat Genet.</em> 1999, 23: 452-456. 10.1038/70570.</p>
</div>
</div>
<span class="target" id="module-variants.mpileup_varscan"></span><div class="section" id="mpileup-varscan">
<h2><a class="toc-backref" href="#id425"><code class="docutils literal"><span class="pre">mpileup_varscan</span></code></a><a class="headerlink" href="#mpileup-varscan" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for identifying variance by running mpileup and piping it’s (large) output into varscan:</p>
<div class="section" id="id331">
<h3>Requires<a class="headerlink" href="#id331" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">BAM files in the the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Genome reference fasta files in the the following slot (the slot should be populated by the module that created the <code class="docutils literal"><span class="pre">bam</span></code> file):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;reference&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id332">
<h3>Output<a class="headerlink" href="#id332" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set to <code class="docutils literal"><span class="pre">sample</span></code>:</p>
<blockquote>
<div><ul>
<li><p class="first">Puts output files in:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;vcf&quot;]</span></code>
<code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;variants&quot;]</span></code> (if <code class="docutils literal"><span class="pre">--output-vcf</span></code> is not redirected in <code class="docutils literal"><span class="pre">redirects</span></code>)</p>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">scope</span></code> is set to <code class="docutils literal"><span class="pre">project</span></code>:</p>
<blockquote>
<div><ul>
<li><p class="first">Puts output files in:</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">sample_data[&quot;vcf&quot;]</span></code>
<code class="docutils literal"><span class="pre">sample_data[&quot;variants&quot;]</span></code> (if <code class="docutils literal"><span class="pre">--output-vcf</span></code> is not redirected in <code class="docutils literal"><span class="pre">redirects</span></code>)</p>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id333">
<h3>Parameters that can be set<a class="headerlink" href="#id333" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>mpileup_path</td>
<td>path</td>
<td>The full path to the mpileup program. You can append additional mpileup arguments after the path (see example lines)</td>
</tr>
<tr class="row-odd"><td>script_path</td>
<td>path</td>
<td>The full path to the relevant varscan program (see example lines).</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id334">
<h3>Comments<a class="headerlink" href="#id334" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id335">
<h3>Lines for parameter file<a class="headerlink" href="#id335" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">mpileup_varscan1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">mpileup_varscan</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">samtools1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">java</span> <span class="o">-</span><span class="n">jar</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">VarScan</span><span class="o">.</span><span class="n">v2</span><span class="o">.</span><span class="mf">3.9</span><span class="o">.</span><span class="n">jar</span> <span class="n">mpileup2snp</span>
    <span class="n">mpileup_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">samtools</span> <span class="n">mpileup</span> <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">depth</span> <span class="mi">6000</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">sample</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">coverage</span><span class="p">:</span> <span class="mi">4</span>
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">vcf</span><span class="p">:</span>
        <span class="o">--</span><span class="n">variants</span><span class="p">:</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="id336">
<h3>References<a class="headerlink" href="#id336" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., Marth, G., Abecasis, G. and Durbin, R., 2009. <strong>The sequence alignment/map format and SAMtools</strong>. <em>Bioinformatics</em>, 25(16), pp.2078-2079.</li>
<li>Koboldt, D.C., Chen, K., Wylie, T., Larson, D.E., McLellan, M.D., Mardis, E.R., Weinstock, G.M., Wilson, R.K. and Ding, L., 2009. <strong>VarScan: variant detection in massively parallel sequencing of individual and pooled samples</strong>. <em>Bioinformatics</em>, 25(17), pp.2283-2285.</li>
</ul>
</div>
</div>
<span class="target" id="module-main_NSF_classes.variants.vcftools"></span><div class="section" id="vcftools">
<h2><a class="toc-backref" href="#id426"><code class="docutils literal"><span class="pre">vcftools</span></code> <sup>*</sup></a><a class="headerlink" href="#vcftools" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running vcftools:</p>
<p>Can take a VCF, gunzipped VCF or BCF file as input.</p>
<p>Produces an output file, as specified by the output options arguments.</p>
<div class="section" id="id337">
<h3>Requires<a class="headerlink" href="#id337" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Input files in one of the following slots (for project scope):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&quot;VCF&quot;</span> <span class="pre">|</span> <span class="pre">&quot;gzVCF&quot;</span> <span class="pre">|</span> <span class="pre">&quot;BCF&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Input files in one of the following slots (for sample scope):</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;variants&quot;][&quot;VCF&quot;</span> <span class="pre">|</span> <span class="pre">&quot;gzVCF&quot;</span> <span class="pre">|</span> <span class="pre">&quot;BCF&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id338">
<h3>Output<a class="headerlink" href="#id338" title="Permalink to this headline">¶</a></h3>
<ul>
<li><dl class="first docutils">
<dt>Puts output files in the following slots (for project scope): </dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">self.sample_data[&lt;output</span> <span class="pre">type&gt;]</span></code></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Puts output files in the following slots (for sample scope): </dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;variants&quot;][&lt;output</span> <span class="pre">type&gt;]</span></code></p>
</dd>
</dl>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Output type is set by redirecting the required type, <em>i.e.</em> any number of the following list of types.</p>
<p>For extracting several INFO fields, set <code class="docutils literal"><span class="pre">--get-INFO</span></code> to a list of INFO elements to extract (instead of passing <code class="docutils literal"><span class="pre">--get-INFO</span></code> several times). See examples below.</p>
<p>If several output types are passed, each type will be created in parallel with a different vcftools script.</p>
<p>See the vcftools manual for details (<a class="reference external" href="https://vcftools.github.io/man_latest.html">https://vcftools.github.io/man_latest.html</a>).</p>
<p class="last"><code class="docutils literal"><span class="pre">&quot;--freq&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--freq2&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--counts&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--counts2&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--depth&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--site-depth&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--site-mean-depth&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--geno-depth&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--hap-r2&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--geno-r2&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--geno-chisq&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--hap-r2-positions&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--geno-r2-positions&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--interchrom-hap-r2&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--interchrom-geno-r2&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--TsTv&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--TsTv-summary&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--TsTv-by-count&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--TsTv-by-qual&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--FILTER-summary&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--site-pi&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--window-pi&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--weir-fst-pop&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--het&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--hardy&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--TajimaD&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--indv-freq-burden&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--LROH&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--relatedness&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--relatedness2&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--site-quality&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--missing-indv&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--missing-site&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--SNPdensity&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--kept-sites&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--removed-sites&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--singletons&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--hist-indel-len&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--hapcount&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--mendel&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--extract-FORMAT-info&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--get-INFO&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--recode&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--recode-bcf&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--12&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--IMPUTE&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--ldhat&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--ldhat-geno&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--BEAGLE-GL&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--BEAGLE-PL&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--plink&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;--plink-tped&quot;</span></code>.</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">At the moment, you can’t pass more than one <code class="docutils literal"><span class="pre">extract-FORMAT-info</span></code> option at once. For more than one <code class="docutils literal"><span class="pre">extract-FORMAT-info</span></code>, create more than one instance of <code class="docutils literal"><span class="pre">vcftools</span></code>.</p>
</div>
</div>
<div class="section" id="id339">
<h3>Parameters that can be set<a class="headerlink" href="#id339" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>scope</td>
<td>project | sample</td>
<td>Indicates whether to use a project or sample bowtie1 index.</td>
</tr>
<tr class="row-odd"><td>input</td>
<td>vcf | bcf | gzvcf</td>
<td>Type of input to use. Default: vcf</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id340">
<h3>Lines for parameter file<a class="headerlink" href="#id340" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">vcftools1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">vcftools</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">freebayes1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">vcftools</span>
    <span class="n">scope</span><span class="p">:</span> <span class="n">project</span>
    <span class="nb">input</span><span class="p">:</span> <span class="n">vcf</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">recode</span><span class="p">:</span>
        <span class="o">--</span><span class="n">extract</span><span class="o">-</span><span class="n">FORMAT</span><span class="o">-</span><span class="n">info</span><span class="p">:</span> <span class="n">GT</span>
        <span class="o">--</span><span class="n">get</span><span class="o">-</span><span class="n">INFO</span><span class="p">:</span>
            <span class="o">-</span> <span class="n">NS</span>
            <span class="o">-</span> <span class="n">DB</span>
</pre></div>
</div>
</div>
<div class="section" id="id341">
<h3>References<a class="headerlink" href="#id341" title="Permalink to this headline">¶</a></h3>
<p>Danecek, P., Auton, A., Abecasis, G., Albers, C.A., Banks, E., DePristo, M.A., Handsaker, R.E., Lunter, G., Marth, G.T., Sherry, S.T. and McVean, G., 2011. <strong>The variant call format and VCFtools</strong>. <em>Bioinformatics</em>, 27(15), pp.2156-2158.</p>
</div>
</div>
</div>
<div class="section" id="chip-seq">
<h1>ChIP-seq<a class="headerlink" href="#chip-seq" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id342">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#macs2-callpeak" id="id427"><code class="docutils literal"><span class="pre">macs2_callpeak</span></code> <sup>*</sup></a></li>
<li><a class="reference internal" href="#macs2-bdgcmp" id="id428"><code class="docutils literal"><span class="pre">macs2_bdgcmp</span></code></a></li>
<li><a class="reference internal" href="#ceas" id="id429"><code class="docutils literal"><span class="pre">CEAS</span></code></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.ChIP_seq.macs2_callpeak"></span><div class="section" id="macs2-callpeak">
<h2><a class="toc-backref" href="#id427"><code class="docutils literal"><span class="pre">macs2_callpeak</span></code> <sup>*</sup></a><a class="headerlink" href="#macs2-callpeak" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running macs2 callpeak:</p>
<div class="section" id="id343">
<h3>Requires<a class="headerlink" href="#id343" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">bam files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[&lt;sample&gt;][&quot;bam&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If using control (input) samples, make sure you include a sample-control table in your sample file.</p>
</li>
</ul>
</div>
<div class="section" id="id344">
<h3>Output<a class="headerlink" href="#id344" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output macs2 output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;prefix&quot;])</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;peak_bed&quot;])</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;peak_xls&quot;])</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;summit_bed&quot;])</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">--bdg</span></code> (or <code class="docutils literal"><span class="pre">-B</span></code>) was specified, puts output bdg files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;control_lambda&quot;]</span></code> - Control BedGraph</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;treat_pileup&quot;]</span></code> - Treatment BedGraph</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bdg&quot;]</span></code> - Treatment BedGraph</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;control&gt;][&quot;bdg&quot;]</span></code> - Control BedGraph</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">bedToBigBed_path</span></code> was specified, puts output bigbed files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bb&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
<li><p class="first">If <code class="docutils literal"><span class="pre">getfasta</span></code> was specified, puts output fasta files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;peak_fasta&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;fasta.nucl&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id345">
<h3>Parameters that can be set<a class="headerlink" href="#id345" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>bedToBigBed_path</td>
<td>path to bedToBigBed</td>
<td>Runs bedToBigBed to convert the peak bed files into bigbed for uploading to UCSC.</td>
</tr>
<tr class="row-odd"><td>chrom.sizes</td>
<td>path to chrom.sizes for reference genome</td>
<td>If running bedToBigBed, you must supply the genome chrom.sizes file.</td>
</tr>
<tr class="row-even"><td>getfasta</td>
<td>&#160;</td>
<td>If set, a fasta file containing the peak sequences will be produced.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id346">
<h3>Lines for parameter file<a class="headerlink" href="#id346" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">macs1_CP</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">macs2_callpeak</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">samtools1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">macs2</span> <span class="n">callpeak</span>
    <span class="n">bedToBigBed_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">kentUtils</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">bedToBigBed</span>
    <span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span>
    <span class="n">getfasta</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">bedtools</span> <span class="n">getfasta</span> <span class="o">-</span><span class="n">name</span> <span class="o">-</span><span class="n">s</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">SPMR</span><span class="p">:</span> 
        <span class="o">--</span><span class="n">bdg</span><span class="p">:</span> 
        <span class="o">-</span><span class="n">g</span><span class="p">:</span>     <span class="n">mm</span>
        <span class="o">--</span><span class="n">bw</span><span class="p">:</span>   <span class="mi">400</span>
</pre></div>
</div>
</div>
<div class="section" id="id347">
<h3>References<a class="headerlink" href="#id347" title="Permalink to this headline">¶</a></h3>
<p>Feng, J., Liu, T., Qin, B., Zhang, Y. and Liu, X.S., 2012. <strong>Identifying ChIP-seq enrichment using MACS</strong>. <em>Nature protocols</em>, 7(9), pp.1728-1740.</p>
</div>
</div>
<span class="target" id="module-ChIP_seq.macs2_bdgcmp"></span><div class="section" id="macs2-bdgcmp">
<h2><a class="toc-backref" href="#id428"><code class="docutils literal"><span class="pre">macs2_bdgcmp</span></code></a><a class="headerlink" href="#macs2-bdgcmp" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running macs2 bdgcmp:</p>
<div class="section" id="id348">
<h3>Requires<a class="headerlink" href="#id348" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;control_lambda&quot;]</span></code> - Control BedGraph</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;treat_pileup&quot;]</span></code> - Treatment BedGraph</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id349">
<h3>Output<a class="headerlink" href="#id349" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts output macs2 output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bdg&quot;])</span></code> - The comparison bedgraph!</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;bigwig&quot;])</span></code> - if <code class="docutils literal"><span class="pre">slop_path</span></code> and <code class="docutils literal"><span class="pre">ucscTools_path</span></code> were passed</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig&quot;])</span></code> - if <code class="docutils literal"><span class="pre">slop_path</span></code> and <code class="docutils literal"><span class="pre">ucscTools_path</span></code> were passed</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;tdf&quot;])</span></code> - in TDF format (if <code class="docutils literal"><span class="pre">slop_path</span></code>, <code class="docutils literal"><span class="pre">ucscTools_path</span></code> and <code class="docutils literal"><span class="pre">toTDF_path</span></code> were passed)</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id350">
<h3>Parameters that can be set<a class="headerlink" href="#id350" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="29%" />
<col width="29%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Values</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>slop_path</td>
<td>path to bedtools slop</td>
<td>Is part of the process for converting bdg files into bigwig and wig</td>
</tr>
<tr class="row-odd"><td>ucscTools_path</td>
<td>path to ucscTools</td>
<td>UCSCtools bedClip, bedGraphToBigWig and bigWigToWig are part of the process for converting bdg files into bigwig and wig</td>
</tr>
<tr class="row-even"><td>toTDF_path</td>
<td>path to toTDF</td>
<td>Converts the wig file into TDF file.</td>
</tr>
<tr class="row-odd"><td>genome</td>
<td>path to chrom.sizes for reference genome</td>
<td>If running bedToBigBed, you must supply the genome chrom.sizes file.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id351">
<h3>Lines for parameter file<a class="headerlink" href="#id351" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bdgcmp</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">macs2_bdgcmp</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">macs1</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">macs2</span> <span class="n">bdgcmp</span>
    <span class="n">genome</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">chrom</span><span class="o">.</span><span class="n">sizes</span><span class="o">.</span><span class="n">txt</span>
    <span class="n">slop_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">bedtools</span> <span class="n">slop</span>
    <span class="n">ucscTools_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ucscTools</span><span class="o">/</span><span class="nb">bin</span>
    <span class="n">toTDF_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">java</span> <span class="o">-</span><span class="n">Xmx1500m</span> <span class="o">-</span><span class="n">jar</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">igvtools</span><span class="o">.</span><span class="n">jar</span> <span class="n">toTDF</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">--</span><span class="n">method</span><span class="p">:</span> <span class="n">FE</span>
</pre></div>
</div>
</div>
<div class="section" id="id352">
<h3>References<a class="headerlink" href="#id352" title="Permalink to this headline">¶</a></h3>
<p>Feng, J., Liu, T., Qin, B., Zhang, Y. and Liu, X.S., 2012. <strong>Identifying ChIP-seq enrichment using MACS</strong>. <em>Nature protocols</em>, 7(9), pp.1728-1740.</p>
</div>
</div>
<span class="target" id="module-ChIP_seq.CEAS"></span><div class="section" id="ceas">
<h2><a class="toc-backref" href="#id429"><code class="docutils literal"><span class="pre">CEAS</span></code></a><a class="headerlink" href="#ceas" title="Permalink to this headline">¶</a></h2>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Authors:</th><td class="field-body">Menachem Sklarz</td>
</tr>
<tr class="field-even field"><th class="field-name">Affiliation:</th><td class="field-body">Bioinformatics core facility</td>
</tr>
<tr class="field-odd field"><th class="field-name">Organization:</th><td class="field-body">National Institute of Biotechnology in the Negev, Ben Gurion University.</td>
</tr>
</tbody>
</table>
<p>A module for running CEAS:</p>
<div class="section" id="id353">
<h3>Requires<a class="headerlink" href="#id353" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;peak_bed&quot;]</span></code> - Sample <code class="docutils literal"><span class="pre">peak_bed</span></code> file</li>
<li><code class="docutils literal"><span class="pre">self.sample_data[&lt;sample&gt;][&quot;wig&quot;]</span></code> - An appropriate <code class="docutils literal"><span class="pre">wig</span></code> file</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id354">
<h3>Output<a class="headerlink" href="#id354" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p class="first">Puts CEAS output files in the following slots:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">sample_data[sample][&quot;CEAS.xls&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[sample][&quot;CEAS.R&quot;]</span></code></li>
<li><code class="docutils literal"><span class="pre">sample_data[sample][&quot;CEAS.plots&quot;]</span></code></li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="id355">
<h3>Parameters that can be set<a class="headerlink" href="#id355" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="id356">
<h3>Lines for parameter file<a class="headerlink" href="#id356" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">CEAS1</span><span class="p">:</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">CEAS</span>
    <span class="n">base</span><span class="p">:</span> <span class="n">UCSC_BW_to_wig</span>
    <span class="n">script_path</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">ceas</span>
    <span class="n">redirects</span><span class="p">:</span>
        <span class="o">-</span><span class="n">g</span><span class="p">:</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">hg19</span><span class="o">.</span><span class="n">refGene</span>
</pre></div>
</div>
</div>
<div class="section" id="id357">
<h3>References<a class="headerlink" href="#id357" title="Permalink to this headline">¶</a></h3>
<p>Shin, H., Liu, T., Manrai, A.K. and Liu, X.S., 2009. <strong>CEAS: cis-regulatory element annotation system</strong>. <em>Bioinformatics</em>, 25(19), pp.2605-2606.</p>
</div>
</div>
</div>
<div class="section" id="generic-module">
<h1>Generic Module<a class="headerlink" href="#generic-module" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="id358">
<p class="topic-title first">Modules included in this section</p>
<ul class="simple">
<li><a class="reference internal" href="#generic" id="id430"><code class="docutils literal"><span class="pre">Generic</span></code> <sup>*</sup></a></li>
</ul>
</div>
<span class="target" id="module-main_NSF_classes.Generic.Generic"></span><div class="section" id="generic">
<h2><a class="toc-backref" href="#id430"><code class="docutils literal"><span class="pre">Generic</span></code> <sup>*</sup></a><a class="headerlink" href="#generic" title="Permalink to this headline">¶</a></h2>
<p>A class that defines a module for a generic module.
It enables the user to design new modules that can handle most cases.</p>
<div class="section" id="parameters-file-arguments">
<h3>Parameters file arguments:<a class="headerlink" href="#parameters-file-arguments" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Generic_step</span><span class="p">:</span>                   <span class="c1"># Name of this step</span>
    <span class="n">module</span><span class="p">:</span> <span class="n">Generic</span>             <span class="c1"># Name of module</span>
    <span class="n">base</span><span class="p">:</span>                       <span class="c1"># Name of the step [or list of names] to run after [mast be after steps that generates the inputs File_Types] </span>
    <span class="n">script_path</span><span class="p">:</span>                <span class="c1"># Main command for this module</span>
    <span class="n">scope</span><span class="p">:</span>                      <span class="c1"># The scope of this module could be sample/project, the default is by sample</span>
    <span class="n">shell</span><span class="p">:</span>                      <span class="c1"># Type of shell [csh OR bash]. csh is the default</span>
    <span class="n">inputs</span><span class="p">:</span>                     <span class="c1"># The inputs for this module</span>
        <span class="n">STR</span><span class="p">:</span>                    <span class="c1"># Input argument, could be also &#39;empty#&#39;</span>
            <span class="n">scope</span><span class="p">:</span>              <span class="c1"># The scope of this input argument could be sample/project</span>
                                <span class="c1"># If the module scope is project and the argument scope is sample:</span>
                                <span class="c1"># all the samples inputs File_Types of this argument will be listed as: argument File_Type(of sample #) </span>
            <span class="n">File_Type</span><span class="p">:</span>          <span class="c1"># The input File_Type could be any File_Type available from previous (in this branch) steps</span>
            <span class="n">sep</span><span class="p">:</span>                <span class="c1"># If the module scope is project and the argument scope is sample:  </span>
                                <span class="c1"># all the samples inputs File_Types of this argument will be listed delimited by sep</span>
            <span class="k">del</span><span class="p">:</span>                <span class="c1"># Delete this File_Type after the step ends</span>
                                <span class="c1"># Will generate empty file with the same name and a suffix of _DELETED</span>
    <span class="n">outputs</span><span class="p">:</span>                    <span class="c1"># The outputs for this module</span>
        <span class="n">STR</span><span class="p">:</span>                    <span class="c1"># Output argument, the scope of the output arguments is determinant by the module scope</span>
                                <span class="c1"># could be also &#39;empty#&#39; OR &quot;No_run&quot; for only entering the file information to output File_Type </span>
            <span class="n">File_Type</span><span class="p">:</span>          <span class="c1"># The output File_Type could be any File_Type name for the current branch downstream work </span>
                                <span class="c1"># If the File_Type exists its content will be override for the current branch downstream work </span>
            <span class="n">prefix</span><span class="p">:</span>             <span class="c1"># A prefix for this output argument file name                </span>
            <span class="n">suffix</span><span class="p">:</span>             <span class="c1"># A suffix for this output argument file name                </span>
                                <span class="c1"># between prefix and suffix will be the sample name [in sample scope] or the project title [in project scope] </span>
            <span class="n">constant_file_name</span><span class="p">:</span> <span class="c1"># Use constant file name for this output argument [ignore prefix and suffix]</span>
                                <span class="c1"># If empty [&#39;&#39;] will enter the output directory location</span>
            <span class="n">use_base_name</span><span class="p">:</span>      <span class="c1"># use only the base name of the output file [ignored if constant_file_name is used]</span>
    <span class="n">copy_File_Types</span><span class="p">:</span>            <span class="c1"># Transferring information between File_Types</span>
        <span class="n">STR</span><span class="p">:</span>                    <span class="c1"># Unique name for the transfer</span>
            <span class="n">source</span><span class="p">:</span>
                <span class="n">File_Type</span><span class="p">:</span>      <span class="c1"># Copy the content of source File_Type to the target File_Type [copy from here]</span>
                <span class="n">scope</span><span class="p">:</span>          <span class="c1"># From the scope of scope_in [if scope_in is not specified the default is sample]</span>
            <span class="n">target</span><span class="p">:</span>
                <span class="n">File_Type</span><span class="p">:</span>      <span class="c1"># Copy the content of source File_Type to the target File_Type [copy to here]</span>
                <span class="n">scope</span><span class="p">:</span>          <span class="c1"># From the scope of scope_out [if scope_out is not specified the default is sample]</span>
    <span class="n">qsub_params</span><span class="p">:</span>                <span class="c1"># Parameters for qsub [number of cpus or memory to reserve etc ]</span>
        <span class="n">STR</span><span class="p">:</span> 
    <span class="n">redirects</span><span class="p">:</span>                  <span class="c1"># Parameters to pass directly to the command</span>
        <span class="n">STR</span><span class="p">:</span> 
</pre></div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Workflow_docs/Assembly_Indexing_mapping.html" class="btn btn-neutral float-right" title="Assembly and Mapping" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="NeatSeq_Flow’s Module and Workflow Repository" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Menachem Sklarz.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>